- Strandedness settings regarding different tools (also for two Illumina kits): https://github.com/igordot/genomics/blob/master/notes/rna-seq-strand.md
- Moreover, nf-core/rnaseq checks strandedness no matter what is specified in the "sample sheet" and would complain if the provided strandedness information is different from the inferred strandedness information. For strandedness inference, Salmon is employed ([relevant code here](https://github.com/drpatelh/nf-core-rnaseq/commit/0d760d26aeab37765e6b45720df17a6f3f92b01f)) and I believe strandedness will not be in any of the outputs (can be found within the `work` folder though) but the MultiQC output would also have a section regarding strandedness, "Infer experiment", generated by RSeQC (levels: `Sense`, `Antisense`, `Undetermined`). The code below shows how the `RSeQC infer_experiment.py` output is mapped to the levels above ([full code here](https://github.com/ewels/MultiQC/blob/master/multiqc/modules/rseqc/infer_experiment.py)):
```
    # Set up vars
    self.infer_exp = dict()
    regexes = {
        "pe_sense": r"\"1\+\+,1--,2\+-,2-\+\": (\d\.\d+)",
        "pe_antisense": r"\"1\+-,1-\+,2\+\+,2--\": (\d\.\d+)",
        "se_sense": r"\"\+\+,--\": (\d\.\d+)",
        "se_antisense": r"\+-,-\+\": (\d\.\d+)",
        "failed": r"Fraction of reads failed to determine: (\d\.\d+)",
    }
```
