<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.54">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Modern Statistics for Modern Biology - Introduction</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./imgs/msmb-favicon.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="msmb.css">
<meta property="og:title" content="Modern Statistics for Modern Biology - Introduction">
<meta property="og:description" content="If you are a biologist and want to get the best out of the powerful methods of modern computational statistics, this is your book.">
<meta property="og:image" content="https://www.huber.embl.de/msmb-quarto/imgs/MSFMB-Cover2-cropped.jpg">
<meta property="og:site-name" content="Modern Statistics for Modern Biology">
<meta name="twitter:title" content="Modern Statistics for Modern Biology - Introduction">
<meta name="twitter:description" content="If you are a biologist and want to get the best out of the powerful methods of modern computational statistics, this is your book.">
<meta name="twitter:image" content="https://www.huber.embl.de/msmb-quarto/imgs/MSFMB-Cover2-cropped.jpg">
<meta name="twitter:card" content="summary">
<script type="text/javascript">
  var _paq = _paq || [];
  
  // Call disableCookies before calling trackPageView 
  _paq.push(['disableCookies']);
  _paq.push(['trackPageView']);
  _paq.push(['enableLinkTracking']);
  (function() {
    var u='//tr-denbi.embl.de/heimdall/';
    _paq.push(['setTrackerUrl', u+'p.php']);
    _paq.push(['setSiteId', '26']);
    var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
    g.type='text/javascript'; g.async=true; g.defer=true; g.src=u+'p.js'; s.parentNode.insertBefore(g,s);
  })();
</script>

</head>

<body class="nav-sidebar floating nav-fixed slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Modern Statistics for Modern Biology</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-chapters" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Chapters</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-chapters">    
        <li>
    <a class="dropdown-item" href="./index.html" rel="" target="">
 <span class="dropdown-text">Home</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./00-chap.html" rel="" target="">
 <span class="dropdown-text">Introduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./01-chap.html" rel="" target="">
 <span class="dropdown-text">Generative Models for Discrete Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./02-chap.html" rel="" target="">
 <span class="dropdown-text">Statistical Modeling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./03-chap.html" rel="" target="">
 <span class="dropdown-text">High Quality Graphics in R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./04-chap.html" rel="" target="">
 <span class="dropdown-text">Mixture Models</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./05-chap.html" rel="" target="">
 <span class="dropdown-text">Clustering</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./06-chap.html" rel="" target="">
 <span class="dropdown-text">Testing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./07-chap.html" rel="" target="">
 <span class="dropdown-text">Multivariate Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./08-chap.html" rel="" target="">
 <span class="dropdown-text">High-Throughput Count Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./09-chap.html" rel="" target="">
 <span class="dropdown-text">Multivariate methods for heterogeneous data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./10-chap.html" rel="" target="">
 <span class="dropdown-text">Networks and Trees</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./11-chap.html" rel="" target="">
 <span class="dropdown-text">Image data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./12-chap.html" rel="" target="">
 <span class="dropdown-text">Supervised Learning</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./13-chap.html" rel="" target="">
 <span class="dropdown-text">Design of High Throughput Experiments and their Analyses</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./14-chap.html" rel="" target="">
 <span class="dropdown-text">Statistical Concordance</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./15-chap.html" rel="" target="">
 <span class="dropdown-text">Acknowledgements</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./16-chap.html" rel="" target="">
 <span class="dropdown-text">References</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Introduction</li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Sections</h2>
   
  <ul class="collapse">
  <li><a href="#the-challenge-heterogeneity" id="toc-the-challenge-heterogeneity" class="nav-link active" data-scroll-target="#the-challenge-heterogeneity">The challenge: heterogeneity</a></li>
  <li><a href="#whats-in-this-book" id="toc-whats-in-this-book" class="nav-link" data-scroll-target="#whats-in-this-book">What’s in this book?</a></li>
  <li><a href="#computational-tools-for-modern-biologists" id="toc-computational-tools-for-modern-biologists" class="nav-link" data-scroll-target="#computational-tools-for-modern-biologists">Computational tools for modern biologists</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Introduction</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<div class="cell page-columns page-full" data-layout-align="center" data-hash="00-chap_cache/html/WringingFlood-web_ca38f203c3cc549fe32f86ec87346e94">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/WringingFlood_web.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div></div></div>
<p>The two instances of <em>modern</em> in the title of this book reflect the two major recent revolutions in biological data analyses:</p>
<ul>
<li><p>Biology, formerly a science with sparse, often only qualitative data has turned into a field whose production of quantitative data is on par with high energy physics or astronomy, and whose data are wildly more heterogeneous and complex.</p></li>
<li><p>Statistics, a field that in the 20th century had become an application ground for probability theory and calculus, often taught loaded with notation and perceived with a heavy emphasis on hypothesis testing, has been transformed by the ubiquity of computers and of data in machine-readable form. Exploratory data analysis, visualization, resampling, simulations, pragmatic hybridizations of Bayesian ideas and methods with frequentist data analysis have become part of the toolset.</p></li>
</ul>
<p>The aim of this book is to enable scientists working in biological research to quickly learn many of the important ideas and methods that they need to make the best of their experiments and of other available data. The book takes a hands-on approach. The narrative is driven by classes of questions, or by certain data types. Methods and theory are introduced on a need-to-know basis. We don’t try to systematically deduce from first principles. The book will often throw readers into the pool and hope they can swim in spite of so many missing details.</p>
<p>By no means this book will replace systematic training in underlying theory: probability, linear algebra, calculus, computer science, databases, multivariate statistics. Such training takes many semesters of coursework. Perhaps the book will whet your appetite to engage more deeply with one of these fields.</p>
<section id="the-challenge-heterogeneity" class="level2">
<h2 class="anchored" data-anchor-id="the-challenge-heterogeneity">The challenge: heterogeneity</h2>
<p>Any biological system or organism is composed of tens of thousands of components, which can be in different states and interact in multiple ways. Modern biology aims to understand such systems by acquiring comprehensive –and this means high-dimensional– data in their temporal and spatial context, with multiple covariates and interactions. Dealing with this complexity will be our primary challenge. This includes real, biological complexity as well as the practical complexities and heterogeneities of the data we are able to acquire with our always imperfect instruments.</p>
<p>Biological data come in all sorts of shapes: nucleic acid and protein sequences, rectangular tables of counts, multiple tables, continuous variables, batch factors, phenotypic images, spatial coordinates. Besides data measured in lab experiments, there are clinical data, environmental observations, measurements in space a d time, networks and lineage trees, and heaps of previously accumulated knowledge in biological databases as free text or controlled vocabularies, <span class="math inline">\(...\)</span></p>
<blockquote class="blockquote">
<p>“Homogeneous data are all alike; all heterogeneous data are heterogeneous in their own way.” The Anna Karenina principle.</p>
</blockquote>
<p>It is such heterogeneity that motivates our choice of R and Bioconductor as the computational platform for this book – some more on this below.</p>
</section>
<section id="whats-in-this-book" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="whats-in-this-book">What’s in this book?</h2>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="00-chap_cache/html/fig-Fisher_c7dd19f5f5d0bad7cda0c7c18e695b66">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-Fisher" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="imgs/FisherParadigm.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;1: The hypothesis testing paradigm recommended by R.A.&nbsp;Fisher starts with the formulation of a null hypothesis and the design of an experiment before the collection of any data. We could think in a similarly schematic way about model fitting: just replace <em>Hypothesis H0</em> by <em>Parametric Model</em> and by <em>Fit Parameters</em>.</figcaption>
</figure>
</div>
</div></div></div>
<p><a href="#fig-Fisher" class="quarto-xref">Figure&nbsp;<span>1</span></a> outlines a sequential view of statistical data analysis. Motivated by the groundbreaking work on significance and hypothesis testing in the 1930s by <span class="citation" data-cites="fisher1935design">Fisher (<a href="16-chap.html#ref-fisher1935design" role="doc-biblioref">1935</a>)</span> and <span class="citation" data-cites="Neyman1936">Neyman and Pearson (<a href="16-chap.html#ref-Neyman1936" role="doc-biblioref">1936</a>)</span>, it is well amenable to mathematical formalism, especially the part where we compute the distribution of test statistics under a hypothesis (null or alternative), or where we need to set up distributional assumptions and can search for analytical approximations.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="00-chap_cache/html/fig-iterative_ee2579945808f5caaeae97b8bdbccc25">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-iterative" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="imgs/iterativeparadigm.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;2: J.W.&nbsp;Tukey recommended starting any analysis with the data and wrote: “No catalogue of techniques can convey a willingness to look for what can be seen, whether or not anticipated.” <span class="citation" data-cites="Holmes1985">(<a href="16-chap.html#ref-Holmes1985" role="doc-biblioref">Holmes - Junca 1985</a>)</span>.</figcaption>
</figure>
</div>
</div></div></div>
<p>Real scientific discovery rarely works in the caricature manner of <a href="#fig-Fisher" class="quarto-xref">Figure&nbsp;<span>1</span></a>. <span class="citation" data-cites="Tukey77">Tukey (<a href="16-chap.html#ref-Tukey77" role="doc-biblioref">1977</a>)</span> emphasized two separate approaches. The first he termed <strong>exploratory data analysis</strong> (<strong>EDA</strong>). EDA uses the data themselves to decide how to conduct the statistical analysis. EDA is built on data visualization, and then complemented by <strong>confirmatory data analyses</strong> (CDA): hypothesis-driven inferential methods, which ideally should be robust and not rely on complex assumptions. Tukey recommended an iterative approach schematized in <a href="#fig-iterative" class="quarto-xref">Figure&nbsp;<span>2</span></a> that enable us to see the data at different resolutions and from different perspectives. This enables successive refinement of our understanding of the data and the underlying natural phenomena.</p>
<p>Biology in the late 1990s raised the large-<span class="math inline">\(p\)</span>, small-<span class="math inline">\(n\)</span> problem: consider a gene expression dataset for <span class="math inline">\(n=200\)</span> patient-derived tissue samples on <span class="math inline">\(p=20000\)</span> genes. If we want to construct a regression or classification model that “predicts” a clinical variable, for instance the disease type or outcome, from the <span class="math inline">\(20000\)</span> genes, or features, we immediately run into problems, since the potential number of model parameters could be orders of magnitudes larger than the number of measurements. The problems derive from <em>non-identifiability</em> of the parameters or <em>overfitting</em>. At least, this is the case for common models, say, an ordinary multivariate linear model. Statisticians realized that they could remedy the situation by requiring sparsity through the use of regularization techniques <span class="citation" data-cites="HastieTibshiraniFriedman">(<a href="16-chap.html#ref-HastieTibshiraniFriedman" role="doc-biblioref">Hastie, Tibshirani, and Friedman 2008</a>)</span>, i.e., by requiring that many of the potential parameters are either zero or at least close to it.</p>
<p>A generalization of the sparsity principle is attained by invoking one of the most powerful recent ideas in high-dimensional statistics, which goes under the name of <strong>empirical Bayes</strong>: we don’t try to learn all the parameters from scratch, but rather use the fact that groups of them will be similar, or even the same. There are several important book long treatments <span class="citation" data-cites="Efron2010">(<a href="16-chap.html#ref-Efron2010" role="doc-biblioref">Efron 2010</a>)</span> of the subject of large scale inference so essential in modern estimation and hypotheses testing. In the 2010s, computer scientists also discovered ways of engineering “deep neural networks” that can provide sometimes amazing predictive qualities without worrying too much about parameter identifiability or uniqueness.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="00-chap_cache/html/roulette-chunk-1_723a510fe2dd63e5e07116f40c2b4242">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/roulette.png" class="img-fluid figure-img" width="113"></p>
</figure>
</div>
</div></div></div>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="">We’ll use the roulette icon to identify when we are using a Monte Carlo approximation method. This name for the method is used because it uses randomness, similar to the randomness of casino games. Ironically, for many casino games the probabilities of winning is not known analytically, and casinos use their own empirical data to evaluate the odds of winning.</span></div></div>
<p><em>Simulations</em> play an essential role in this book, as many of the results we need escape the reach of standard analytic approaches. In other words, simulations liberate us from being able to only consider methods that are tractable by “paper and pencil mathematics”, and from worrying about the appropriateness of simplifying assumptions or approximations.</p>
<p>In this book, we try to cover a wide spectrum of these developments and their applications to current biological research. We cover many different types of data that modern biologists have to deal with, including RNA-Seq, flow-cytometry, taxa abundances, imaging data and single cell measurements. We assume no prior formal training in statistics. However, you’ll need some familiarity with R and willingness to engage in mathematical and analytical thinking.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="00-chap_cache/html/unnamed-chunk-8_5c75ff56451de91044fc2987479c849b">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/devil.png" class="img-fluid figure-img" width="123"></p>
<figcaption class="figure-caption">We will put notes and extra information under the devil icon, this is the devil who looks after the details.</figcaption>
</figure>
</div>
</div></div></div>
<p>As outlined in the following, the chapters in the book build upon each other, but they are reasonably self-contained, so they can also be studied separately. Each chapter starts with a motivations and goals section. Questions in the text will help you check whether you are following along. The text contains complete R code examples throughout. You don’t need to scrape R code from the HTML or manually copy it from the book. Please use the R files (extension <code>.R</code>) on this website. Each chapter concludes with a summary of the main points and a set of exercises.</p>
<p><em>Generative models</em> are our basic building blocks. In order to draw conclusions about complicated data it tends to be useful to have simple models for the data generated under this or that situation. We do this through the use of probability theory and generative models, which we introduce in <a href="01-chap.html" class="quarto-xref"><span>Chapter&nbsp;1</span></a>. We will use examples from immunology and DNA analysis to describe useful generative models for biological data: binomial, multinomial and Poisson random variables.</p>
<p>Once we know how data would look like under a certain model, we can start working our way backwards: given some data, what model is most likely able to explain it? This <em>bottom up approach</em> is the core of statistical thinking, and we explain it in <a href="02-chap.html" class="quarto-xref"><span>Chapter&nbsp;2</span></a>.</p>
<p>We saw the primary role of <em>graphics</em> in Tukey’s scheme (<a href="#fig-iterative" class="quarto-xref">Figure&nbsp;<span>2</span></a>), and so we’ll learn how to visualize our data in <a href="03-chap.html" class="quarto-xref"><span>Chapter&nbsp;3</span></a>. We’ll use the grammar of graphics and <a href="https://cran.r-project.org/web/packages/ggplot2/">ggplot2</a>.</p>
<p>Real biological data often have more complex distributional properties than what we could cover in <a href="01-chap.html" class="quarto-xref"><span>Chapter&nbsp;1</span></a>. We’ll use <em>mixtures</em>, which we explore in <a href="04-chap.html" class="quarto-xref"><span>Chapter&nbsp;4</span></a>; these enable us to build realistic models for heterogeneous biological data and provide solid foundations for choosing appropriate variance stabilizing transformations.</p>
<p>The large, matrix-like datasets in biology naturally lend themselves to <em>clustering</em>: once we define a distance measure between matrix rows (the features), we can cluster and group the genes by similarity of their expression patterns, and similarly, for the columns (the patient samples). We’ll cover clustering in <a href="05-chap.html" class="quarto-xref"><span>Chapter&nbsp;5</span></a>. Since clustering only relies on distances, we can even apply it to data that are not matrix-shaped, as long as there are objects and distances defined between them.</p>
<p>Further following the path of EDA, we cover the most fundamental unsupervised analysis method for simple matrices –<em>principal component analysis</em>– in <a href="07-chap.html" class="quarto-xref"><span>Chapter&nbsp;7</span></a>. We turn to more heterogeneous data that combine multiple data types in <a href="09-chap.html" class="quarto-xref"><span>Chapter&nbsp;9</span></a>. There, we’ll see nonlinear unsupervised methods for counts from single cell data. We’ll also address how to use generalizations of the multivariate approaches covered in <a href="07-chap.html" class="quarto-xref"><span>Chapter&nbsp;7</span></a> to combinations of categorical variables and multiple assays recorded on the same observational units.</p>
<p>The basic hypothesis testing workflow outlined in <a href="#fig-Fisher" class="quarto-xref">Figure&nbsp;<span>1</span></a> is explained in <a href="06-chap.html" class="quarto-xref"><span>Chapter&nbsp;6</span></a>. We use the opportunity to apply it to one of the most common queries to <span class="math inline">\(n\times p\)</span>-datasets: which of the genes (features) are <em>associated with</em> a certain property of the samples, say, disease type or outcome? However, conventional significance thresholds would lead to lots of spurious associations: with a false positive rate of <span class="math inline">\(\alpha=0.05\)</span> we expect <span class="math inline">\(p\alpha=1000\)</span> false positives if none of the <span class="math inline">\(p=20000\)</span> features has a true association. Therefore we also need to deal with multiple testing.</p>
<p>One of the most fruitful ideas in statistics is that of variance decomposition, or analysis of variance (ANOVA). We’ll explore this, in the framework of linear models and generalized linear models, in <a href="08-chap.html" class="quarto-xref"><span>Chapter&nbsp;8</span></a>. Since we’ll draw our example data from an RNA-Seq experiment, this gives us also an opportunity to discuss models for such count data, and concepts of <em>robustness</em>.</p>
<div class="page-columns page-full"><p>Nothing in biology makes sense except in the light of evolution<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>, and evolutionary relationships are usefully encoded in phylogenetic trees. We’ll explore networks and trees in <a href="10-chap.html" class="quarto-xref"><span>Chapter&nbsp;10</span></a>.</p><div class="no-row-height column-margin column-container"><li id="fn1"><p><sup>1</sup>&nbsp;Theodosius Dobzhansky, <a href="https://en.wikipedia.org/wiki/Nothing_in_Biology_Makes_Sense_Except_in_the_Light_of_Evolution" class="uri">https://en.wikipedia.org/wiki/Nothing_in_Biology_Makes_Sense_Except_in_the_Light_of_Evolution</a></p></li></div></div>
<p>A rich source of data in biology is images, and in <a href="11-chap.html" class="quarto-xref"><span>Chapter&nbsp;11</span></a> we reinforce our willingness to do EDA on all sorts of heterogeneous data types by exploring feature extraction from images and spatial statistics.</p>
<p>In <a href="12-chap.html" class="quarto-xref"><span>Chapter&nbsp;12</span></a>, we look at supervised learning: train an algorithm to distinguish between different classes of objects, given a multivariate set of features for each object. We’ll start simple with low-dimensional feature vectors and linear methods, and then step forward to some of the issues of classification in high-dimensional settings. Here, we focus on ‘’classical’’ supervised learning, where (at least conceptually) a training set of ground truth classifications is available to the algorithm all at once, simultaneously, to learn from. We do not (yet…) cover reinforcement learning, a more flexible framework in which one or several agents learn by interacting with a so-called environment by successively performing actions and getting feedback on them; so there are notions of <em>time</em> and <em>state</em> in that framework.</p>
<p>We wrap up in <a href="13-chap.html" class="quarto-xref"><span>Chapter&nbsp;13</span></a> with considerations on good practices in the design of experiments and of data analyses. For this we’ll use and reflect what we have learned in the course of the preceding chapters.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="00-chap_cache/html/fig-structuralescal_d0eb19a3171d1cd5cb0cc63b22106b24">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-structuralescal" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="imgs/structuralescal.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;3: Analyzing data is not a one step process. Each step involves visualizing and decomposing some of the complexity in the data. Tukey’s iterative data structuration can be conceptualized as <span class="math inline">\(Total=V_1+V_2+V_3\)</span></figcaption>
</figure>
</div>
</div></div></div>
</section>
<section id="computational-tools-for-modern-biologists" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="computational-tools-for-modern-biologists">Computational tools for modern biologists</h2>
<p>As we’ll see over and over again, the analysis approaches, tools and choices to be made are manifold. Our work can only be validated by keeping careful records in a reproducible script format. R and Bioconductor provide such a platform.</p>
<p>Although we are tackling many different types of data, questions and statistical methods hands-on, we maintain a consistent computational approach by keeping all the computation under one roof: the R programming language and statistical environment, enhanced by the biological data infrastructure and specialized method packages from the Bioconductor project. The reader will have to start by acquiring some familiarity with R before using the book. There are many good books and online resources. One of them is by <span class="citation" data-cites="Wickham:r4ds">Grolemund and Wickham (<a href="16-chap.html#ref-Wickham:r4ds" role="doc-biblioref">2017</a>)</span>, online at <a href="http://r4ds.had.co.nz" class="uri">http://r4ds.had.co.nz</a>.</p>
<div class="page-columns page-full"><p>R code is a major component of this book. It is how we make the textual explanations explicit. Essentially every data visualization in the book is produced with code that is shown, and the reader should be able to replicate all of these figures, and any other result shown. </p><div class="no-row-height column-margin column-container"><span class="">Not all code is shown in the HTML pages, especially in cases where the code is lengthy and not directly conducive to the training aims. But all code is available in the .R files that we provide on this webpage.</span></div></div>
<p>Even if you have a some familiarity with R, don’t worry if you don’t immediately understand every line of code in the book. Although we have tried to keep the code explicit and give tips and hints at potentially challenging places, there will be instances where</p>
<ul>
<li><p>there is a function invoked that you have not seen before and that does something mysterious,</p></li>
<li><p>there is a complicated R expression that you don’t understand (perhaps involving <code>apply</code>-functions or data manipulations from the <strong><a href="https://cran.r-project.org/web/packages/dplyr/">dplyr</a></strong> package).</p></li>
</ul>
<p>Don’t panic. For the mysterious function, have a look at its manual page. Open up RStudio and use the object explorer to look at the variables that go into the expression, and those that come out. Split up the expression to look at intermediate values.</p>
<p>In Chapters <a href="01-chap.html" class="quarto-xref"><span>1</span></a> and <a href="02-chap.html" class="quarto-xref"><span>2</span></a>, we use <strong><a href="https://cran.r-project.org/web/packages/base/">base</a></strong> R functionality for light doses of plotting and data manipulation. As we successively need more sophisticated operations, we introduce the <strong><a href="https://cran.r-project.org/web/packages/ggplot2/">ggplot2</a></strong> way of making graphics in <a href="03-chap.html" class="quarto-xref"><span>Chapter&nbsp;3</span></a>. Besides the powerful grammar of graphics concepts that enable us to produce sophisticated plots using only a limited set of instructions, this implies using the <strong><a href="https://cran.r-project.org/web/packages/dplyr/">dplyr</a></strong> way of data manipulation.</p>
<section id="why-r-and-bioconductor" class="level4 page-columns page-full">
<h4 class="anchored" data-anchor-id="why-r-and-bioconductor">Why R and Bioconductor?</h4>
<div class="page-columns page-full"><p>There are many reasons why we have chosen to present all analyses on the R <span class="citation" data-cites="RlangB">(<a href="16-chap.html#ref-RlangB" role="doc-biblioref">Ihaka and Gentleman 1996</a>)</span> and Bioconductor <span class="citation" data-cites="Huber:2015">(<a href="16-chap.html#ref-Huber:2015" role="doc-biblioref">Huber et al. 2015</a>)</span> platforms: </p><div class="no-row-height column-margin column-container"><span class="">Download <a href="http://cran.us.r-project.org/">R</a> and <a href="http://www.rstudio.com/">Rstudio</a> to follow the code in the book.</span></div></div>
<p><strong>Cutting edge solutions.</strong> The availability of over <span class="math inline">\(10,000\)</span> packages ensures that almost all statistical methods are available, including the most recent developments. Moreover, there are implementations of or interfaces to many methods from computer science, mathematics, machine learning, data management, visualization and internet technologies. This puts thousands of person-years of work by experts at your finger tips.</p>
<p><strong>Open source and community-owned.</strong> R and Bioconductor have been built collaboratively by a large community of developers. They are constantly tried and tested by thousands of users.</p>
<p><strong>Data input and wrangling.</strong> Bioconductor packages support the reading of many of the data types and formats produced by measurement instruments used in modern biology, as well as the needed technology-specific “preprocessing” routines. The community is actively keeping these up-to-date with the rapid developments on the instrument market.</p>
<p><strong>Simulation.</strong> There are random number generators for every known statistical distribution and powerful numeric routines for linear algebra, optimization, etc.</p>
<p><strong>Visualization and presentation.</strong> R can make attractive, publication-quality graphics. We’ve dedicated <a href="03-chap.html" class="quarto-xref"><span>Chapter&nbsp;3</span></a> to this and practice data visualization extensively throughout the book.</p>
<p><strong>Easy to use interactive development environment.</strong> <strong>RStudio</strong> is easy and fun to use and helps with all aspects of programming in R. It is an essential piece in following the iterative approach to data analysis schematized in <a href="#fig-iterative" class="quarto-xref">Figure&nbsp;<span>2</span></a>.</p>
<p><strong>Reproducibility.</strong> As an equivalent to the laboratory notebook that is standard good practice in labwork, we advocate the use of a computational diary written in the R markdown or quarto formats. We use the <strong><a href="https://quarto.org/">quarto</a></strong> system to convert these files into easy-to-read and shareable HTML or PDF documents. These can even become full-fledged scientific articles or supplements. Together with a version control system, this approach also helps with tracking changes.</p>
<p><strong>Collaborative environment.</strong> Quarto enables the creation of websites containing code, text, figures and tables with a minimum of work.</p>
<p><strong>Rich data structures.</strong> The Bioconductor project has defined specialized data containers to represent complex biological datasets. These help to keep your data consistent, safe and easy to use.</p>
<p><strong>Interoperability and distributed development.</strong> Bioconductor in particular contains packages from diverse authors that cover a wide range of functionalities but still interoperate because of the common data containers.</p>
<p><strong>Documentation.</strong> Many R packages come with excellent documentation in their function manual pages and vignettes. The vignettes are usually the best starting point into a package, as they give you a high-level narrative on what the package does, whereas the manual pages give detailed information on input, output and inner workings of each function. There are online tutorials, fora and mailing lists for many aspects of working with R and Bioconductor.</p>
<div class="page-columns page-full"><p><strong>High-level language.</strong> R is an interpreted high-level language. Its roots in LISP and its functional programming features mean that code is data and can be computed on, which enables efficient programming and is fun. These features facilitate constructing powerful domain specific languages<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>. R is not a fixed language – throughout its history, it has been actively evolving and is constantly improving.</p><div class="no-row-height column-margin column-container"><li id="fn2"><p><sup>2</sup>&nbsp;Examples include R’s formula interface, the grammar of graphics in <strong><a href="https://cran.r-project.org/web/packages/ggplot2/">ggplot2</a></strong>, the data manipulation functionality of <strong><a href="https://cran.r-project.org/web/packages/dplyr/">dplyr</a></strong> and R markdown.</p></li></div></div>


<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-Efron2010" class="csl-entry" role="listitem">
Efron, Bradley. 2010. <em>Large-Scale Inference: Empirical <span>B</span>ayes Methods for Estimation, Testing, and Prediction</em>. Cambridge University Press.
</div>
<div id="ref-fisher1935design" class="csl-entry" role="listitem">
Fisher, Ronald Aylmer. 1935. <em>The Design of Experiments</em>. Oliver &amp; Boyd.
</div>
<div id="ref-Wickham:r4ds" class="csl-entry" role="listitem">
Grolemund, Garrett, and Hadley Wickham. 2017. <em><span>R</span> for Data Science</em>. O’Reilly.
</div>
<div id="ref-HastieTibshiraniFriedman" class="csl-entry" role="listitem">
Hastie, Trevor, Robert Tibshirani, and Jerome Friedman. 2008. <em>The Elements of Statistical Learning</em>. 2^{\text{nd}} ed. Springer.
</div>
<div id="ref-Holmes1985" class="csl-entry" role="listitem">
Holmes - Junca, Susan. 1985. <span>“Outils Informatiques Pour l’<span>é</span>valuation de La Pertinence d’un r<span>é</span>sultat En Analyse Des Donn<span>é</span>es.”</span> PhD thesis, Université Montpellier II, France.
</div>
<div id="ref-Huber:2015" class="csl-entry" role="listitem">
Huber, Wolfgang, Vincent J Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S Carvalho, Hector Corrada Bravo, et al. 2015. <span>“Orchestrating High-Throughput Genomic Analysis with <span>Bioconductor</span>.”</span> <em>Nature Methods</em> 12 (2): 115–21.
</div>
<div id="ref-RlangB" class="csl-entry" role="listitem">
Ihaka, Ross, and Robert Gentleman. 1996. <span>“R: A Language for Data Analysis and Graphics.”</span> <em>Journal of Computational and Graphical Statistics</em> 5 (3): 299–314.
</div>
<div id="ref-Neyman1936" class="csl-entry" role="listitem">
Neyman, Jerzy, and Egon S Pearson. 1936. <em>Sufficient Statistics and Uniformly Most Powerful Tests of Statistical Hypotheses</em>. University California Press.
</div>
<div id="ref-Tukey77" class="csl-entry" role="listitem">
Tukey, John W. 1977. <span>“Exploratory Data Analysis.”</span> <em>Massachusetts: Addison-Wesley</em>.
</div>
</div>
</section>
</section>
<aside id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr>
<ol>


</ol>
</aside>

<p>Page built on 2023-08-03 21:37:40.742167 using R version 4.3.0 (2023-04-21)</p></main> <!-- /main --></p><p class="build-date">
Support for maintaining the online version of this book is provided by <a href="http://www.denbi.de">de.NBI</a>
  <img src="imgs/denbi.png" style="vertical-align: text-top;height: 16px"><br>
For website support please contact msmith@embl.de
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS_eq = new window.AnchorJS();
  anchorJS_eq.options = {
    placement: 'left',
    class: 'eq-link',
    icon: icon
  };
  anchorJS_eq.add('.anchored-eq');
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>
