<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.54">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Modern Statistics for Modern Biology - 9&nbsp; Multivariate methods for heterogeneous data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./imgs/msmb-favicon.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="msmb.css">
<meta property="og:title" content="Modern Statistics for Modern Biology - 9&nbsp; Multivariate methods for heterogeneous data">
<meta property="og:description" content="If you are a biologist and want to get the best out of the powerful methods of modern computational statistics, this is your book.">
<meta property="og:image" content="https://www.huber.embl.de/msmb-quarto/imgs/MSFMB-Cover2-cropped.jpg">
<meta property="og:site-name" content="Modern Statistics for Modern Biology">
<meta name="twitter:title" content="Modern Statistics for Modern Biology - 9&nbsp; Multivariate methods for heterogeneous data">
<meta name="twitter:description" content="If you are a biologist and want to get the best out of the powerful methods of modern computational statistics, this is your book.">
<meta name="twitter:image" content="https://www.huber.embl.de/msmb-quarto/imgs/MSFMB-Cover2-cropped.jpg">
<meta name="twitter:card" content="summary">
<script type="text/javascript">
  var _paq = _paq || [];
  
  // Call disableCookies before calling trackPageView 
  _paq.push(['disableCookies']);
  _paq.push(['trackPageView']);
  _paq.push(['enableLinkTracking']);
  (function() {
    var u='//tr-denbi.embl.de/heimdall/';
    _paq.push(['setTrackerUrl', u+'p.php']);
    _paq.push(['setSiteId', '26']);
    var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
    g.type='text/javascript'; g.async=true; g.defer=true; g.src=u+'p.js'; s.parentNode.insertBefore(g,s);
  })();
</script>

</head>

<body class="nav-sidebar floating nav-fixed slimcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">Modern Statistics for Modern Biology</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-chapters" role="button" data-bs-toggle="dropdown" aria-expanded="false" rel="" target="">
 <span class="menu-text">Chapters</span>
    </a>
    <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="nav-menu-chapters">    
        <li>
    <a class="dropdown-item" href="./index.html" rel="" target="">
 <span class="dropdown-text">Home</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./00-chap.html" rel="" target="">
 <span class="dropdown-text">Introduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./01-chap.html" rel="" target="">
 <span class="dropdown-text">Generative Models for Discrete Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./02-chap.html" rel="" target="">
 <span class="dropdown-text">Statistical Modeling</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./03-chap.html" rel="" target="">
 <span class="dropdown-text">High Quality Graphics in R</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./04-chap.html" rel="" target="">
 <span class="dropdown-text">Mixture Models</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./05-chap.html" rel="" target="">
 <span class="dropdown-text">Clustering</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./06-chap.html" rel="" target="">
 <span class="dropdown-text">Testing</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./07-chap.html" rel="" target="">
 <span class="dropdown-text">Multivariate Analysis</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./08-chap.html" rel="" target="">
 <span class="dropdown-text">High-Throughput Count Data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./09-chap.html" rel="" target="">
 <span class="dropdown-text">Multivariate methods for heterogeneous data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./10-chap.html" rel="" target="">
 <span class="dropdown-text">Networks and Trees</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./11-chap.html" rel="" target="">
 <span class="dropdown-text">Image data</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./12-chap.html" rel="" target="">
 <span class="dropdown-text">Supervised Learning</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./13-chap.html" rel="" target="">
 <span class="dropdown-text">Design of High Throughput Experiments and their Analyses</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./14-chap.html" rel="" target="">
 <span class="dropdown-text">Statistical Concordance</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./15-chap.html" rel="" target="">
 <span class="dropdown-text">Acknowledgements</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="./16-chap.html" rel="" target="">
 <span class="dropdown-text">References</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Multivariate methods for heterogeneous data</span></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Sections</h2>
   
  <ul class="collapse">
  <li><a href="#goals-for-this-chapter" id="toc-goals-for-this-chapter" class="nav-link active" data-scroll-target="#goals-for-this-chapter"><span class="header-section-number">9.1</span> Goals for this chapter</a></li>
  <li><a href="#sec-hetero-multidimensionalscalingandordination" id="toc-sec-hetero-multidimensionalscalingandordination" class="nav-link" data-scroll-target="#sec-hetero-multidimensionalscalingandordination"><span class="header-section-number">9.2</span> Multidimensional scaling and ordination</a></li>
  <li><a href="#contiguous-or-supplementary-information" id="toc-contiguous-or-supplementary-information" class="nav-link" data-scroll-target="#contiguous-or-supplementary-information"><span class="header-section-number">9.3</span> Contiguous or supplementary information</a></li>
  <li><a href="#sec-hetero-CA" id="toc-sec-hetero-CA" class="nav-link" data-scroll-target="#sec-hetero-CA"><span class="header-section-number">9.4</span> Correspondence analysis for contingency tables</a></li>
  <li><a href="#sec-hetero-findingtime" id="toc-sec-hetero-findingtime" class="nav-link" data-scroll-target="#sec-hetero-findingtime"><span class="header-section-number">9.5</span> Finding time…and other important gradients.</a></li>
  <li><a href="#multitable-techniques" id="toc-multitable-techniques" class="nav-link" data-scroll-target="#multitable-techniques"><span class="header-section-number">9.6</span> Multitable techniques</a></li>
  <li><a href="#summary-of-this-chapter" id="toc-summary-of-this-chapter" class="nav-link" data-scroll-target="#summary-of-this-chapter"><span class="header-section-number">9.7</span> Summary of this chapter</a></li>
  <li><a href="#further-reading" id="toc-further-reading" class="nav-link" data-scroll-target="#further-reading"><span class="header-section-number">9.8</span> Further reading</a></li>
  <li><a href="#exercises" id="toc-exercises" class="nav-link" data-scroll-target="#exercises"><span class="header-section-number">9.9</span> Exercises</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
    </div>
<!-- main -->
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="sec-hetero" class="quarto-section-identifier"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Multivariate methods for heterogeneous data</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/Brighton-West-Pier-20090214-sunset_45de09b7f0703e279d68574e1f198555">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/Brighton_West_Pier_20090214_sunset.jpg" class="img-fluid figure-img"></p>
</figure>
</div>
</div></div></div>
<div class="page-columns page-full"><p></p><div class="no-row-height column-margin column-container"><span class="">Real situations often involve point clouds, gradients, graphs, attraction points, noise and different spatial milieux, a little like this picture where we have a rigid skeleton, waves, sun and starlings.</span></div></div>
<p>In <a href="07-chap.html" class="quarto-xref"><span>Chapter&nbsp;7</span></a>, we saw how to summarize rectangular matrices whose columns were continuous variables. The maps we made used unsupervised dimensionality reduction techniques such as principal component analysis aimed at isolating the most important <em>signal</em> component in a matrix <span class="math inline">\(X\)</span> when all the columns have meaningful variances.</p>
<p>Here we extend these ideas to more complex heterogeneous data where continuous and categorical variables are combined. Indeed, sometimes our observations cannot be easily described by sets of individual variables or coordinates – but it is possible to determine distances or (dis)similarities between them, or to describe relationships between them using a graph or a tree. Examples include species in a species tree or biological sequences. Outside of biology, examples include text documents or movie files, where we may have a reasonable method to determine (dis)similarity between them, but no obvious variables or coordinates.</p>
<p>This chapter contains more advanced techniques, for which we often omit technical details. Having come this far, we hope that by giving you some hands-on experience with examples, and extensive references, to enable you to understand and use some of the more `cutting edge’ techniques in nonlinear multivariate analysis.</p>
<section id="goals-for-this-chapter" class="level2" data-number="9.1">
<h2 data-number="9.1" class="anchored" data-anchor-id="goals-for-this-chapter"><span class="header-section-number">9.1</span> Goals for this chapter</h2>
<p>In this chapter, we will:</p>
<ul>
<li><p>Extend linear dimension reduction methods to cases when the distances between observations are available, known as <strong>m</strong>ulti<strong>d</strong>imensional <strong>s</strong>caling (MDS) or principal coordinates analysis.</p></li>
<li><p>Find modifications of MDS that are nonlinear and robust to outliers.</p></li>
<li><p>Encode combinations of categorical data and continuous data as well as so-called ‘supplementary’ information. We will see that this enables us to deal with <em>batch effects</em>.</p></li>
<li><p>Use chi-square distances and <strong>c</strong>orrespondence <strong>a</strong>nalysis (CA) to see where categorical data (contingency tables) contain notable dependencies.</p></li>
<li><p>Generalize clustering methods that can uncover latent variables that are not categorical. This will allow us to detect gradients, “pseudotime” and hidden nonlinear effects in our data.</p></li>
<li><p>Generalize the notion of variance and covariance to the study of tables of data from multiple different data domains.</p></li>
</ul>
</section>
<section id="sec-hetero-multidimensionalscalingandordination" class="level2 page-columns page-full" data-number="9.2">
<h2 data-number="9.2" class="anchored" data-anchor-id="sec-hetero-multidimensionalscalingandordination"><span class="header-section-number">9.2</span> Multidimensional scaling and ordination</h2>
<p>Sometimes, data are <em>not</em> represented as points in a feature space. This can occur when we are provided with (dis)similarity matrices between objects such as drugs, images, trees or other complex objects, which have no obvious coordinates in <span class="math inline">\({\mathbb R}^n\)</span>.</p>
<p>In <a href="05-chap.html" class="quarto-xref"><span>Chapter&nbsp;5</span></a> we saw how to produce <em>clusters</em> from distances. Here our goal is to visualize the data in maps in low dimensional spaces (e.g., planes) reminiscent of the ones we make from the first few principal axes in PCA.</p>
<div class="page-columns page-full"><p>We start with an intuitive example using geography data. In <a href="#fig-HeatDists" class="quarto-xref">Figure&nbsp;<span>9.1</span></a>, a heatmap and clustering of the distances between cities and places in Ukraine<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> are shown.</p><div class="no-row-height column-margin column-container"><li id="fn1"><p><sup>1</sup>&nbsp;The provenance of these data are described in the script ukraine-dists.R in the data folder.</p></li></div></div>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/UkrCities_d9e123562b183f29b8ed25e1af86c49b">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"pheatmap"</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"ukraine_dists"</span>, <span class="at">package =</span> <span class="st">"MSMB"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">as.matrix</span>(ukraine_dists)[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               Kyiv    Odesa Sevastopol Chernihiv
Kyiv         0.0000 441.2548   687.7551  128.1287
Odesa      441.2548   0.0000   301.7482  558.6483
Sevastopol 687.7551 301.7482     0.0000  783.6561
Chernihiv  128.1287 558.6483   783.6561    0.0000</code></pre>
</div>
</div>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-HeatDists_5adf35a507a8a614cef1bc24d5dafd6a">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(<span class="fu">as.matrix</span>(ukraine_dists), </span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">color =</span> <span class="fu">colorRampPalette</span>(<span class="fu">c</span>(<span class="st">"#0057b7"</span>, <span class="st">"#ffd700"</span>))(<span class="dv">50</span>),</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="fu">max</span>(ukraine_dists)<span class="sc">^</span>(<span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>), <span class="at">length.out =</span> <span class="dv">51</span>)<span class="sc">^</span><span class="dv">2</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">treeheight_row =</span> <span class="dv">10</span>, <span class="at">treeheight_col =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-HeatDists" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-HeatDists-1.png" class="img-fluid figure-img" width="552"></p>
<figcaption class="figure-caption">Figure&nbsp;9.1: A heatmap of the <code>ukraine_dists</code> distance matrix. Distances are measured in metres. The function has re-arranged the order of the cities, and grouped the closest ones.</figcaption>
</figure>
</div>
</div></div></div>
<p>Besides <code>ukraine_dists</code>, which contains the pairwise distances, the RData file that we loaded above also contains the dataframe <code>ukraine_coords</code> with the longitudes and latitudes; we will use this later as a ground truth. Given the distances, multidimensional scaling (MDS) provides a “map” of their relative locations. It will not be possible to arrange the cities such that their Euclidean distances on a 2D plane exactly reproduce the given distance matrix: the cities lie on the curved surface of the Earth rather than in a plane. Nevertheless, we can expect to find a two dimensional embedding that represents the data well. With biological data, our 2D embeddings are likely to be much less clearcut. We call the function with:</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/cmdscaleDistE_f3f9ed8e491fd302c0c8d0c7b475b358">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>ukraine_mds <span class="ot">=</span> <span class="fu">cmdscale</span>(ukraine_dists, <span class="at">eig =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We make a function that we will reuse several times in this chapter to make a screeplot from the result of a call to the <code>cmdscale</code> function:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"dplyr"</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggplot2"</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>plotscree <span class="ot">=</span> <span class="cf">function</span>(x, <span class="at">m =</span> <span class="fu">length</span>(x<span class="sc">$</span>eig)) {</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">tibble</span>(<span class="at">eig =</span> x<span class="sc">$</span>eig[<span class="fu">seq_len</span>(m)], <span class="at">k =</span> <span class="fu">seq</span>(<span class="at">along =</span> eig)),</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">aes</span>(<span class="at">x =</span> k, <span class="at">y =</span> eig)) <span class="sc">+</span> <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_discrete</span>(<span class="st">"k"</span>, <span class="at">limits =</span> <span class="fu">as.factor</span>(<span class="fu">seq_len</span>(m))) <span class="sc">+</span> </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">width =</span> <span class="fl">0.5</span>, <span class="at">fill =</span> <span class="st">"#ffd700"</span>, <span class="at">col =</span> <span class="st">"#0057b7"</span>)</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-plotscreeeig_32b01a879d15c534ea339ceb8f9ae4f5">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotscree</span>(ukraine_mds, <span class="at">m =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-plotscreeeig" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-plotscreeeig-1.png" class="img-fluid figure-img" width="320"></p>
<figcaption class="figure-caption">Figure&nbsp;9.2: Screeplot of the first four eigenvalues. There is a pronounced drop after the first two eigenvalues, which indicates that the data are well described by a two-dimensional embedding.</figcaption>
</figure>
</div>
</div></div></div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-barplotEurope" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.1 </strong></span>Look at all the eigenvalues output by the <code>cmdscale</code> function: what do you notice?</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>If you execute:</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/fig-plotscreeeigall_58232e6b82cd3f8bdf33bb7398c1a64b">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>ukraine_mds<span class="sc">$</span>eig <span class="sc">|&gt;</span> <span class="fu">signif</span>(<span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  3.91e+06  1.08e+06  3.42e+02  4.84e-01  2.13e-01  3.83e-05  5.90e-06
 [8]  5.83e-07  8.78e-08  4.95e-08  7.66e-10  2.24e-10  8.65e-11  6.10e-11
[15]  3.64e-11  4.67e-12 -5.63e-11 -8.58e-11 -8.91e-11 -9.36e-11 -1.39e-10
[22] -2.16e-10 -2.47e-10 -3.61e-10 -3.73e-10 -2.31e-09 -1.52e-08 -9.60e-05
[29] -2.51e-04 -1.41e-02 -1.19e-01 -3.58e+02 -8.85e+02</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotscree</span>(ukraine_mds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-plotscreeeigall" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-plotscreeeigall-1.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption">Figure&nbsp;9.3: Screeplot of all the eigenvalues.</figcaption>
</figure>
</div>
</div>
</div>
<p>you will note that unlike in PCA, there are some negative eigenvalues. These are due to the way <code>cmdscale</code> works.</p>
</div>
</div>
</div>
<p>The main output from the <code>cmdscale</code> function are the coordinates of the two-dimensional embedding, which we show in <a href="#fig-ukrainemds1" class="quarto-xref">Figure&nbsp;<span>9.4</span></a> (we will discuss how the algorithm works in the next section).</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-ukrainemds1_6ff28d4332717489b71e7d5d676fe4b9">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>ukraine_mds_df <span class="ot">=</span> <span class="fu">tibble</span>(</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">PCo1 =</span> ukraine_mds<span class="sc">$</span>points[, <span class="dv">1</span>],</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">PCo2 =</span> ukraine_mds<span class="sc">$</span>points[, <span class="dv">2</span>],</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">labs =</span> <span class="fu">rownames</span>(ukraine_mds<span class="sc">$</span>points)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ggrepel"</span>)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>g <span class="ot">=</span> <span class="fu">ggplot</span>(ukraine_mds_df, <span class="fu">aes</span>(<span class="at">x =</span> PCo1, <span class="at">y =</span> PCo2, <span class="at">label =</span> labs)) <span class="sc">+</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_text_repel</span>(<span class="at">col =</span> <span class="st">"#0057b7"</span>) <span class="sc">+</span> <span class="fu">coord_fixed</span>() </span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>g</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-ukrainemds1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-ukrainemds1-1.png" class="img-fluid figure-img" width="480"></p>
<figcaption class="figure-caption">Figure&nbsp;9.4: MDS map based on the distances.</figcaption>
</figure>
</div>
</div></div></div>
<p>Note that while relative positions are correct, the orientation of the map is unconventional: Crimea is at the top. This is a common phenomenon with methods that reconstruct planar embeddings from distances. Since the distances between the points are invariant under rotations and reflections (axis flips), any solution is as good as any other solution that relates to it via rotation or reflection. Functions like <code>cmdscale</code> will pick one of the equally optimal solutions, and the particular choice can depend on minute details of the data or the computing platform being used. Here, we can transform our result into a more conventional orientation by reversing the sign of the <span class="math inline">\(y\)</span>-axis. We redraw the map in <a href="#fig-ukrainemds2" class="quarto-xref">Figure&nbsp;<span>9.5</span></a> and compare this to the true longitudes and latitudes from the <code>ukraine_coords</code> dataframe (<a href="#fig-ukrainecoord1" class="quarto-xref">Figure&nbsp;<span>9.6</span></a>).</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-ukrainemds2_89ffd90b31e89a50ad868603ede00b5e">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>g <span class="sc">%+%</span> <span class="fu">mutate</span>(ukraine_mds_df, <span class="at">PCo1 =</span> PCo1, <span class="at">PCo2 =</span> <span class="sc">-</span>PCo2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-ukrainemds2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-ukrainemds2-1.png" class="img-fluid figure-img" width="480"></p>
<figcaption class="figure-caption">Figure&nbsp;9.5: Same as <a href="#fig-ukrainemds1" class="quarto-xref">Figure&nbsp;<span>9.4</span></a>, but with y-axis flipped.</figcaption>
</figure>
</div>
</div></div></div>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-ukrainecoord1_8e82374dd630f2cc5ea0788db9f95310">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"ukraine_coords"</span>, <span class="at">package =</span> <span class="st">"MSMB"</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print.data.frame</span>(ukraine_coords[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>,  <span class="fu">c</span>(<span class="st">"city"</span>, <span class="st">"lat"</span>, <span class="st">"lon"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        city      lat      lon
1       Kyiv 50.45003 30.52414
2      Odesa 46.48430 30.73229
3 Sevastopol 44.60544 33.52208
4  Chernihiv 51.49410 31.29433</code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ukraine_coords, <span class="fu">aes</span>(<span class="at">x =</span> lon, <span class="at">y =</span> lat, <span class="at">label =</span> city)) <span class="sc">+</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_text_repel</span>(<span class="at">col =</span> <span class="st">"#0057b7"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-ukrainecoord1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-ukrainecoord1-1.png" class="img-fluid figure-img" width="480"></p>
<figcaption class="figure-caption">Figure&nbsp;9.6: True latitudes and longitudes, taken from the <code>ukraine_coords</code> dataframe.</figcaption>
</figure>
</div>
</div></div></div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-mdsflip" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.2 </strong></span>We drew the longitudes and latitudes in the right panel of <a href="#fig-ukrainecoord1" class="quarto-xref">Figure&nbsp;<span>9.6</span></a> without attention to aspect ratio. What is the right aspect ratio for this plot?</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-4-contents" aria-controls="callout-4" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-4" class="callout-4-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>There is no simple relationship between the distances that correspond to 1 degree change in longitude and to 1 degree change in latitude, so the choice is difficult to make. Even under the simplifying assumption that our Earth is spherical and has a radius of 6371 km, it’s complicated: one degree in latitude always corresponds to a distance of 111 km (<span class="math inline">\(6371\times2\pi/360\)</span>), as does one degree of longitude on the equator. However, when you move away from the equator, a degree of longitude corresponds to shorter and shorter distances (and to no distance at all at the poles). Pragmatically, for displays such as in <a href="#fig-ukrainecoord1" class="quarto-xref">Figure&nbsp;<span>9.6</span></a>, we could choose a value for the aspect ratio that’s somewhere in the middle between the Northern and Southern most points, say, the cosine for 48 degrees.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-ukrainemap" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.3 </strong></span>Add international borders and geographic features such as rivers to <a href="#fig-ukrainecoord1" class="quarto-xref">Figure&nbsp;<span>9.6</span></a>.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-6-contents" aria-controls="callout-6" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-6" class="callout-6-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>A start point is provided by the code below, which adds international borders (<a href="#fig-ukrainecoord2" class="quarto-xref">Figure&nbsp;<span>9.7</span></a>).</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/fig-ukrainecoord3_c85aec9e114541008ee3bdbe4738968c">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"sf"</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"rnaturalearth"</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"rnaturalearthdata"</span>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>world <span class="ot">=</span> <span class="fu">ne_countries</span>(<span class="at">scale =</span> <span class="st">"medium"</span>, <span class="at">returnclass =</span> <span class="st">"sf"</span>)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span> </span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data =</span> world, </span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>          <span class="at">fill =</span> <span class="fu">ifelse</span>(world<span class="sc">$</span>geounit <span class="sc">==</span> <span class="st">"Ukraine"</span>, <span class="st">"#ffd700"</span>, <span class="st">"#f0f0f0"</span>)) <span class="sc">+</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_sf</span>(<span class="at">xlim =</span> <span class="fu">range</span>(ukraine_coords<span class="sc">$</span>lon) <span class="sc">+</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.5</span>, <span class="sc">+</span> <span class="dv">2</span>), </span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>           <span class="at">ylim =</span> <span class="fu">range</span>(ukraine_coords<span class="sc">$</span>lat) <span class="sc">+</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.5</span>, <span class="sc">+</span> <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> lon, <span class="at">y =</span> lat), <span class="at">data =</span> ukraine_coords) <span class="sc">+</span> </span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="fu">aes</span>(<span class="at">x =</span> lon, <span class="at">y =</span> lat, <span class="at">label =</span> city), </span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>                  <span class="at">color =</span> <span class="st">"#0057b7"</span>, <span class="at">data =</span> ukraine_coords)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/fig-ukrainecoord2_2ca5997aad0df44b6ee7487ac154ca13">
<div class="cell-output-display">
<div id="fig-ukrainecoord2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-ukrainecoord2-1.png" class="img-fluid figure-img" style="width:75.0%"></p>
<figcaption class="figure-caption">Figure&nbsp;9.7: International borders added to <a href="#fig-ukrainecoord1" class="quarto-xref">Figure&nbsp;<span>9.6</span></a>.</figcaption>
</figure>
</div>
</div>
</div>
<p><strong>Note:</strong> MDS creates similar output as PCA, however there is only one ‘dimension’ to the data (the sample points). There is no ‘dual’ dimension, there are no biplots and no loading vectors. This is a drawback when coming to interpreting the maps. Interpretation can be facilitated by examining carefully the extreme points and their differences.</p>
<section id="how-does-the-method-work" class="level3 page-columns page-full" data-number="9.2.1">
<h3 data-number="9.2.1" class="anchored" data-anchor-id="how-does-the-method-work"><span class="header-section-number">9.2.1</span> How does the method work?</h3>
<div class="page-columns page-full"><p>Let’s take a look at what would happen if we really started with points whose coordinates were known<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>. We put these coordinates into the two columns of a matrix with as many rows as there are points. Now we compute the distances between points based on these coordinates. To go from the coordinates <span class="math inline">\(X\)</span> to distances, we write <span class="math display">\[d^2_{i,j} = (x_i^1 - x_j^1)^2 + \dots + (x_i^p - x_j^p)^2.\]</span> We will call the matrix of squared distances <code>DdotD</code> in R and <span class="math inline">\(D\bullet D\)</span> in the text . We want to find points such that the square of their distances is as close as possible to the <span class="math inline">\(D\bullet D\)</span> observed.<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a></p><div class="no-row-height column-margin column-container"><li id="fn2"><p><sup>2</sup>&nbsp;Here we commit a slight ‘abuse’ by using the longitude and longitude of our cities as Cartesian coordinates and ignoring the curvature of the earth’s surface. Check out the internet for information on the Haversine formula.</p></li><span class="">This is different from <span class="math inline">\(DD\)</span> or <span class="math inline">\(D^2\)</span>, the matrix-multiplication of <span class="math inline">\(D\)</span> with itself.</span><li id="fn3"><p><sup>3</sup>&nbsp;Here we commit a slight ‘abuse’ by using the longitudes and latitudes of our cities as Cartesian coordinates and ignoring the fact that they are curvilinear coordinates on a sphere-like surface.</p></li></div></div>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/D2_a5d1d458ffc36bc6fa0460ccfd564cf9">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>X <span class="ot">=</span> <span class="fu">with</span>(ukraine_coords, <span class="fu">cbind</span>(lon, lat <span class="sc">*</span> <span class="fu">cos</span>(<span class="dv">48</span>)))</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>DdotD <span class="ot">=</span> <span class="fu">as.matrix</span>(<span class="fu">dist</span>(X)<span class="sc">^</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The relative distances do not depend on the point of origin of the data. We center the data by using the centering matrix <span class="math inline">\(H\)</span> defined as <span class="math inline">\(H=I-\frac{1}{n}{\mathbf{11}}^t\)</span>. Let’s check the <em>centering</em> property of <span class="math inline">\(H\)</span> using:</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/CheckMDS_c70c666d078d8fa23b1b4f96a064af06">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>n <span class="ot">=</span> <span class="fu">nrow</span>(X)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>H <span class="ot">=</span> <span class="fu">diag</span>(<span class="fu">rep</span>(<span class="dv">1</span>,n))<span class="sc">-</span>(<span class="dv">1</span><span class="sc">/</span>n) <span class="sc">*</span> <span class="fu">matrix</span>(<span class="dv">1</span>, <span class="at">nrow =</span> n, <span class="at">ncol =</span> n)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>Xc <span class="ot">=</span> <span class="fu">sweep</span>(X,<span class="dv">2</span>,<span class="fu">apply</span>(X,<span class="dv">2</span>,mean))</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>Xc[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            lon          
[1,] -1.1722946 -1.184705
[2,] -0.9641429  1.353935</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>HX <span class="ot">=</span> H <span class="sc">%*%</span> X</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>HX[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>            lon          
[1,] -1.1722946 -1.184705
[2,] -0.9641429  1.353935</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">apply</span>(HX, <span class="dv">2</span>, mean)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          lon               
-1.618057e-15  1.747077e-16 </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-centering" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.4 </strong></span>Call <code>B0</code> the matrix obtained by applying the centering matrix both to the right and to the left of <code>DdotD</code> Consider the points centered at the origin given by the <span class="math inline">\(HX\)</span> matrix and compute its cross product, we’ll call this <code>B2</code>. What do you have to do to <code>B0</code> to make it equal to <code>B2</code>?</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-8-contents" aria-controls="callout-8" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-8" class="callout-8-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/unnamed-chunk-19_e0c7a8e2fd4a6a8c2bb15261c9d4a3a7">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>B0 <span class="ot">=</span> H  <span class="sc">%*%</span> DdotD <span class="sc">%*%</span> H</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>B2 <span class="ot">=</span> HX <span class="sc">%*%</span> <span class="fu">t</span>(HX)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>B2[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>] <span class="sc">/</span> B0[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2] [,3]
[1,] -0.5 -0.5 -0.5
[2,] -0.5 -0.5 -0.5
[3,] -0.5 -0.5 -0.5</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">max</span>(<span class="fu">abs</span>(<span class="sc">-</span><span class="fl">0.5</span> <span class="sc">*</span> B0 <span class="sc">-</span> B2))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 9.237056e-14</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Therefore, given the squared distances between rows (<span class="math inline">\(D\bullet D\)</span>) and the cross product of the centered matrix <span class="math inline">\(B=(HX)(HX)^t\)</span>, we have shown:</p>
<p><span class="anchored-eq" data-anchor-id="eq-91" id="eq-91"><span class="math display">\[
-\frac{1}{2} H(D\bullet D) H=B
\tag{9.1}\]</span></span></p>
<p>This is always true, and we use it to reverse-engineer an <span class="anchored-eq" data-anchor-id="eq-91" class="math inline">\(X\)</span> which satisfies <a href="#eq-91" class="quarto-xref">Equation&nbsp;<span>9.1</span></a> when we are given <span class="math inline">\(D\bullet D\)</span> to start with.</p>
<section id="from-dbullet-d-to-x-using-singular-vectors." class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="from-dbullet-d-to-x-using-singular-vectors.">From <span class="math inline">\(D\bullet D\)</span> to <span class="math inline">\(X\)</span> using singular vectors.</h4>
<p>We can go backwards from a matrix <span class="anchored-eq" data-anchor-id="eq-91" class="math inline">\(D\bullet D\)</span> to <span class="math inline">\(X\)</span> by taking the eigen-decomposition of <span class="math inline">\(B\)</span> as defined in <a href="#eq-91" class="quarto-xref">Equation&nbsp;<span>9.1</span></a>. This also enables us to choose how many coordinates, or columns, we want for the <span class="math inline">\(X\)</span> matrix. This is very similar to how PCA provides the best rank <span class="math inline">\(r\)</span> approximation.<br>
<strong>Note</strong>: As in PCA, we can write this using the singular value decomposition of <span class="math inline">\(HX\)</span> (or the eigen decomposition of <span class="math inline">\(HX(HX)^t\)</span>):</p>
<p><span class="math display">\[
HX^{(r)} = US^{(r)}V^t \mbox{ with }
S^{(r)} \mbox{ the diagonal matrix of the first } r \mbox{ singular values},
\]</span></p>
<div class="page-columns page-full"><p> This provides the best approximate representation in an Euclidean space of dimension <span class="math inline">\(r\)</span>.  The algorithm gives us the coordinates of points that have approximately the same distances as those provided by the <span class="math inline">\(D\)</span> matrix.</p><div class="no-row-height column-margin column-container"><span class=""><span class="math display">\[S^{(r)} =
\begin{pmatrix}
s_1 &amp;  0  &amp;  0  &amp;  0  &amp; ...\\
0  &amp; s_2 &amp;  0  &amp;  0  &amp; ...\\
0  &amp;  0  &amp; ... &amp; ... &amp; ...\\
0  &amp;  0  &amp; ... &amp; s_r &amp; ...\\
...&amp; ...&amp; ...&amp; 0 &amp; 0 \\
\end{pmatrix}\]</span></span><span class="">The method is often called Principal Coordinates Analysis, or PCoA which stresses the connection to PCA.</span></div></div>
</section>
<section id="classical-mds-algorithm." class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="classical-mds-algorithm.">Classical MDS Algorithm.</h4>
<p>In summary, given an <span class="math inline">\(n \times n\)</span> matrix of squared interpoint distances <span class="math inline">\(D\bullet D\)</span>, we can find points and their coordinates <span class="math inline">\(\tilde{X}\)</span> by the following operations:</p>
<ol type="1">
<li><p>Double center the interpoint distance squared and multiply it by <span class="math inline">\(-\frac{1}{2}\)</span>:<br>
<span class="math inline">\(B = -\frac{1}{2}H D\bullet D H\)</span>.</p></li>
<li><p>Diagonalize <span class="math inline">\(B\)</span>: <span class="math inline">\(\quad B = U \Lambda U^t\)</span>.</p></li>
<li><p>Extract <span class="math inline">\(\tilde{X}\)</span>: <span class="math inline">\(\quad \tilde{X} = U \Lambda^{1/2}\)</span>.</p></li>
</ol>
</section>
<section id="finding-the-right-underlying-dimensionality." class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="finding-the-right-underlying-dimensionality.">Finding the right underlying dimensionality.</h4>
<p>As an example, let’s take objects for which we have similarities (surrogrates for distances) but for which there is no natural underlying Euclidean space.</p>
<p>In a psychology experiment from the 1950s, <span class="citation" data-cites="Ekman:1954">Ekman (<a href="16-chap.html#ref-Ekman:1954" role="doc-biblioref">1954</a>)</span> asked 31 subjects to rank the similarities of 14 different colors. His goal was to understand the underlying dimensionality of color perception. The similarity or confusion matrix was scaled to have values between 0 and 1. The colors that were often confused had similarities close to 1. We transform the data into a dissimilarity by subtracting the values from 1:</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/ekmandis_3af2bfdefa90fada3d0a3d8a9f233761">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>ekm <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">"../data/ekman.txt"</span>, <span class="at">header=</span><span class="cn">TRUE</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(ekm) <span class="ot">=</span> <span class="fu">colnames</span>(ekm)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>disekm <span class="ot">=</span> <span class="dv">1</span> <span class="sc">-</span> ekm <span class="sc">-</span> <span class="fu">diag</span>(<span class="dv">1</span>, <span class="fu">ncol</span>(ekm))</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>disekm[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     w434 w445 w465 w472 w490
w434 0.00 0.14 0.58 0.58 0.82
w445 0.14 0.00 0.50 0.56 0.78
w465 0.58 0.50 0.00 0.19 0.53
w472 0.58 0.56 0.19 0.00 0.46
w490 0.82 0.78 0.53 0.46 0.00</code></pre>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>disekm <span class="ot">=</span> <span class="fu">as.dist</span>(disekm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We compute the MDS coordinates and eigenvalues. We combine the eigenvalues in the screeplot shown in <a href="#fig-ekmanMDSeig" class="quarto-xref">Figure&nbsp;<span>9.8</span></a>:</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-ekmanMDSeig_cb547db90a8023862a07371eeb8a3b18">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>mdsekm <span class="ot">=</span> <span class="fu">cmdscale</span>(disekm, <span class="at">eig =</span> <span class="cn">TRUE</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotscree</span>(mdsekm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-ekmanMDSeig" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-ekmanMDSeig-1.png" class="img-fluid figure-img" width="240"></p>
<figcaption class="figure-caption">Figure&nbsp;9.8: The screeplot shows us that the phenomenon is largely two dimensional.</figcaption>
</figure>
</div>
</div></div></div>
<p>We plot the different colors using the first two principal coordinates as follows:</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-ekmanMDS-1_15529b60055658a7c38188e64f661770">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>dfekm <span class="ot">=</span> mdsekm<span class="sc">$</span>points[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="sc">|&gt;</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">colnames&lt;-</span><span class="st">`</span>(<span class="fu">paste0</span>(<span class="st">"MDS"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)) <span class="sc">|&gt;</span></span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">|&gt;</span></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">name =</span> <span class="fu">rownames</span>(ekm),</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">rgb =</span> photobiology<span class="sc">::</span><span class="fu">w_length2rgb</span>(<span class="fu">as.numeric</span>(<span class="fu">sub</span>(<span class="st">"w"</span>, <span class="st">""</span>, name))))</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(dfekm, <span class="fu">aes</span>(<span class="at">x =</span> MDS1, <span class="at">y =</span> MDS2)) <span class="sc">+</span></span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">col =</span> dfekm<span class="sc">$</span>rgb, <span class="at">size =</span> <span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="fu">aes</span>(<span class="at">label =</span> name)) <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-ekmanMDS-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-ekmanMDS-1-1.png" class="img-fluid figure-img" width="368"></p>
<figcaption class="figure-caption">Figure&nbsp;9.9: The layout of the scatterpoints in the first two dimensions has a horseshoe shape. The labels and colors show that the arch corresponds to the wavelengths.</figcaption>
</figure>
</div>
</div></div></div>
<p><a href="#fig-ekmanMDS-1" class="quarto-xref">Figure&nbsp;<span>9.9</span></a> shows the Ekman data in the new coordinates. There is a striking pattern that calls for explanation. This horseshoe or arch structure in the points is often an indicator of a sequential latent ordering or gradient in the data <span class="citation" data-cites="Goel:2007aa">(<a href="16-chap.html#ref-Goel:2007aa" role="doc-biblioref">Diaconis, Goel, and Holmes 2008</a>)</span>. We will revisit this in <a href="#sec-hetero-findingtime" class="quarto-xref"><span>Section&nbsp;9.5</span></a>.</p>
</section>
</section>
<section id="robust-versions-of-mds" class="level3 page-columns page-full" data-number="9.2.2">
<h3 data-number="9.2.2" class="anchored" data-anchor-id="robust-versions-of-mds"><span class="header-section-number">9.2.2</span> Robust versions of MDS</h3>
<div class="page-columns page-full"><p> Multidimensional scaling aims to minimize the difference between the squared distances as given by <span class="math inline">\(D\bullet D\)</span> and the squared distances between the points with their new coordinates. Unfortunately, this objective tends to be sensitive to outliers: one single data point with large distances to everyone else can dominate, and thus skew, the whole analysis. Often, we like to use something that is more robust, and one way to achieve this is to disregard the actual values of the distances and only ask that the relative rankings of the original and the new distances are as similar as possible. Such a rank based approach is robust: its sensitivity to outliers is reduced.</p><div class="no-row-height column-margin column-container"><span class=""><strong>Robustness:</strong> A method is robust if it is not too influenced by a few outliers. For example, the median of a set of <span class="math inline">\(n\)</span> numbers does not change by a lot even if we change 20 the numbers by arbitrarily large amounts; to drastically shift the median, we need to change more than half of the numbers. In contrast, we can change the mean by a large amount by just manipulating one of the numbers. We say that the <em>breakdown point</em> of the median is 1/2, while that of the mean is only <span class="math inline">\(1/n\)</span>. Both mean and median are estimators of the <em>location</em> of a distribution (i.e., what is a "typical" value of the numbers), but the median is more robust. The median is based on the ranks; more generally, methods based on ranks are often more robust than those based on the actual values. Many nonparametric tests are based on reductions of data to their ranks.</span></div></div>
<p>We will use the Ekman data to show how useful robust methods are when we are not quite sure about the ‘scale’ of our measurements. Robust ordination, called non metric multidimensional scaling (NMDS for short) only attempts to embed the points in a new space such that the <strong>order</strong> of the reconstructed distances in the new map is the same as the ordering of the original distance matrix.</p>
<p>Non metric MDS looks for a transformation <span class="math inline">\(f\)</span> of the given dissimilarities in the matrix <span class="math inline">\(d\)</span> and a set of coordinates in a low dimensional space (<em>the map</em>) such that the distance in this new map is <span class="math inline">\(\tilde{d}\)</span> and <span class="math inline">\(f(d)\thickapprox \tilde{d}\)</span>. The quality of the approximation can be measured by the standardized residual sum of squares (<span class="smallcaps">stress</span>) function:</p>
<p><span class="math display">\[
\text{stress}^2=\frac{\sum(f(d)-\tilde{d})^2}{\sum d^2}.
\]</span></p>
<p>NMDS is not sequential in the sense that we have to specify the underlying dimensionality at the outset and the optimization is run to maximize the reconstruction of the distances according to that number. There is no notion of percentage of variation explained by individual axes as provided in PCA. However, we can make a simili-screeplot by running the program for all the successive values of <span class="math inline">\(k\)</span> (<span class="math inline">\(k=1, 2, 3, ...\)</span>) and looking at how well the <span class="smallcaps">stress</span> drops. Here is an example of looking at these successive approximations and their goodness of fit. As in the case of diagnostics for clustering, we will take the number of axes <strong>after</strong> the <span class="smallcaps">stress</span> has a steep drop.</p>
<p>Because each calculation of a NMDS result requires a new optimization that is both random and dependent on the <span class="math inline">\(k\)</span> value, we use a similar procedure to what we did for clustering in <a href="04-chap.html" class="quarto-xref"><span>Chapter&nbsp;4</span></a>. We execute the <code>metaMDS</code> function, say, 100 times for each of the four possible values of <span class="math inline">\(k\)</span> and record the <span class="smallcaps">stress</span> values.</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/simstress_edb8103440cfba9314a132411ef3bf20">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"vegan"</span>)</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>nmds.stress <span class="ot">=</span> <span class="cf">function</span>(x, <span class="at">sim =</span> <span class="dv">100</span>, <span class="at">kmax =</span> <span class="dv">4</span>) {</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sapply</span>(<span class="fu">seq_len</span>(kmax), <span class="cf">function</span>(k)</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">replicate</span>(sim, <span class="fu">metaMDS</span>(x, <span class="at">k =</span> k, <span class="at">autotransform =</span> <span class="cn">FALSE</span>)<span class="sc">$</span>stress))</span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>stress <span class="ot">=</span> <span class="fu">nmds.stress</span>(disekm, <span class="at">sim =</span> <span class="dv">100</span>)</span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(stress)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s look at the boxplots of the results. This can be a useful diagnostic plot for choosing <span class="math inline">\(k\)</span> (<a href="#fig-NMDSscreeplot-1" class="quarto-xref">Figure&nbsp;<span>9.10</span></a>).</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-NMDSscreeplot-1_f17755d253d0d84c3465bb9a040277fa">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>dfstr <span class="ot">=</span> reshape2<span class="sc">::</span><span class="fu">melt</span>(stress, <span class="at">varnames =</span> <span class="fu">c</span>(<span class="st">"replicate"</span>,<span class="st">"dimensions"</span>))</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(dfstr, <span class="fu">aes</span>(<span class="at">y =</span> value, <span class="at">x =</span> dimensions, <span class="at">group =</span> dimensions)) <span class="sc">+</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_boxplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-NMDSscreeplot-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-NMDSscreeplot-1-1.png" class="img-fluid figure-img" width="320"></p>
<figcaption class="figure-caption">Figure&nbsp;9.10: Several replicates at each dimension were run to evaluate the stability of the <span class="smallcaps">stress</span>. We see that the <span class="smallcaps">stress</span> drops dramatically with two or more dimensions, thus indicating that a two dimensional solution is appropriate here.</figcaption>
</figure>
</div>
</div></div></div>
<p>We can also compare the distances and their approximations using what is known as a Shepard plot for <span class="math inline">\(k=2\)</span> for instance, computed with:</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-Shepardsplot-1_3d9d59c3e3f3ee20be005821d05a3d49">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>nmdsk2 <span class="ot">=</span> <span class="fu">metaMDS</span>(disekm, <span class="at">k =</span> <span class="dv">2</span>, <span class="at">autotransform =</span> <span class="cn">FALSE</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a><span class="fu">stressplot</span>(nmdsk2, <span class="at">pch =</span> <span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-Shepardsplot-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-Shepardsplot-1-1.png" class="img-fluid figure-img" width="320"></p>
<figcaption class="figure-caption">Figure&nbsp;9.11: The Shepard’s plot compares the original distances or dissimilarities (along the horizonal axis) to the reconstructed distances, in this case for <span class="math inline">\(k=2\)</span> (vertical axis).</figcaption>
</figure>
</div>
</div></div></div>
<p>Both the Shepard’s plot in <a href="#fig-Shepardsplot-1" class="quarto-xref">Figure&nbsp;<span>9.11</span></a> and the screeplot in <a href="#fig-NMDSscreeplot-1" class="quarto-xref">Figure&nbsp;<span>9.10</span></a> point to a two-dimensional solution for Ekman’s color confusion study. Let us compare the output of the two different MDS programs, the classical metric least squares approximation and the nonmetric rank approximation method. The right panel of <a href="#fig-ekmannonMDS-1" class="quarto-xref">Figure&nbsp;<span>9.12</span></a> shows the result from the nonmetric rank approximation, the left panel is the same as <a href="#fig-ekmanMDS-1" class="quarto-xref">Figure&nbsp;<span>9.9</span></a>. The projections are almost identical in both cases. For these data, it makes little difference whether we use a Euclidean or nonmetric multidimensional scaling method.</p>
<div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>nmdsk2<span class="sc">$</span>points[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="sc">|&gt;</span> </span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">`</span><span class="at">colnames&lt;-</span><span class="st">`</span>(<span class="fu">paste0</span>(<span class="st">"NmMDS"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)) <span class="sc">|&gt;</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">|&gt;</span> </span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(dplyr<span class="sc">::</span><span class="fu">select</span>(dfekm, rgb, name)) <span class="sc">|&gt;</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> NmMDS1, <span class="at">y =</span> NmMDS2)) <span class="sc">+</span></span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">col =</span> dfekm<span class="sc">$</span>rgb, <span class="at">size =</span> <span class="dv">4</span>) <span class="sc">+</span></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_text_repel</span>(<span class="fu">aes</span>(<span class="at">label =</span> name))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-ekmannonMDS-1" class="cell quarto-layout-panel" data-hash="09-chap_cache/html/fig-ekmannonMDS-1_0fc3389ee127ce64b22e39e629842ea7">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-ekmannonMDS-1-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-ekmannonMDS-1-1.png" class="img-fluid figure-img" data-ref-parent="fig-ekmannonMDS-1" width="368"></p>
<figcaption class="figure-caption">(a)</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-ekmannonMDS-1-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-ekmannonMDS-1-2.png" class="img-fluid figure-img" data-ref-parent="fig-ekmannonMDS-1" width="368"></p>
<figcaption class="figure-caption">(b)</figcaption>
</figure>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;9.12: Comparison of the output from (a) the classical multidimensional scaling (same as <a href="#fig-ekmanMDS-1" class="quarto-xref">Figure&nbsp;<span>9.9</span></a>) and (b) the nonmetric version.</figcaption><p></p>
</figure>
</div>
</div>
</section>
</section>
<section id="contiguous-or-supplementary-information" class="level2 page-columns page-full" data-number="9.3">
<h2 data-number="9.3" class="anchored" data-anchor-id="contiguous-or-supplementary-information"><span class="header-section-number">9.3</span> Contiguous or supplementary information</h2>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/unnamed-chunk-117_2467365defe90de81c0802da50b6e7f8">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/devil.png" class="img-fluid figure-img" width="123"></p>
<figcaption class="figure-caption"><strong>Metadata:</strong> Many programs and workflows for biological sequence analysis or assays separate the environmental and contextual information, which they call <strong>metadata</strong>, from the assay data or sequence reads. We discourage such practice as the exact connections between the samples and covariates are important. A lost connection between the assays and covariates makes later analyses impossible. Covariates such as clinical history, time, batch or location are important and should be considered components of the data.</figcaption>
</figure>
</div>
</div></div></div>
<p>In <a href="03-chap.html" class="quarto-xref"><span>Chapter&nbsp;3</span></a> we introduced the R <em>data.frame</em> class that enables us to combine heterogeneous data types: categorical factors, text and continuous values. Each row of a dataframe corresponds to an object, or a record, and the columns are the different variables, or features.</p>
<p>Extra information about sample batches, dates of measurement, different protocols are often named <em>metadata</em>; this can be misnomer if it is implied that metadata are somehow less important. Such information is <em>real data</em> that need to be integrated into the analyses. We typically store it in a <em>data.frame</em> or a similar R class and tightly link it to the primary assay data.</p>
<section id="known-batches-in-data" class="level3 page-columns page-full" data-number="9.3.1">
<h3 data-number="9.3.1" class="anchored" data-anchor-id="known-batches-in-data"><span class="header-section-number">9.3.1</span> Known batches in data</h3>
<div class="page-columns page-full"><p>Here we show an example of an analysis that was done by <span class="citation" data-cites="holmes2011psb">Holmes et al. (<a href="16-chap.html#ref-holmes2011psb" role="doc-biblioref">2011</a>)</span> on bacterial abundance data from <em>Phylochip</em> <span class="citation" data-cites="Brodie:2006">(<a href="16-chap.html#ref-Brodie:2006" role="doc-biblioref">Brodie et al. 2006</a>)</span> microarrays. The experiment was designed to detect differences between a group of healthy rats and a group who had Irritable Bowel Disease <span class="citation" data-cites="Nelson:2011">(<a href="16-chap.html#ref-Nelson:2011" role="doc-biblioref">Nelson et al. 2010</a>)</span>. This example shows a case where the nuisance batch effects become apparent in the analysis of experimental data. It is an illustration of the fact that best practices in data analyses are sequential and that it is better to analyse data as they are collected to adjust for severe problems in the experimental design <em>as they occur</em>, instead of having to deal with them <em>post mortem</em><a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a>.</p><div class="no-row-height column-margin column-container"><li id="fn4"><p><sup>4</sup>&nbsp;Fisher’s terminology, see <a href="13-chap.html" class="quarto-xref"><span>Chapter&nbsp;13</span></a>.</p></li></div></div>
<p>When data collection started on this project, days 1 and 2 were delivered and we made the plot that appears in <a href="#fig-rankthreshPCA" class="quarto-xref">Figure&nbsp;<span>9.14</span></a>. This showed a definite <em>day</em> effect. When investigating the source of this effect, we found that both the protocol and the array were different in days 1 and 2. This leads to uncertainty in the source of variation, we call this <strong>confounding</strong> of effects.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/unnamed-chunk-118_d2407f1ec3f1394152660d5be749534e">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/devil.png" class="img-fluid figure-img" width="123"></p>
<figcaption class="figure-caption"><strong>Bioconductor container:</strong> These data are an example of an awkward way of combining batch information with the actual data. The <code>day</code> information has been combined with the array data and encoded as a number and could be confused with a continuous variable. We will see in the next section a better practice for storing and manipulating heterogeneous data using a Bioconductor container called <em>SummarizedExperiment</em>.</figcaption>
</figure>
</div>
</div></div></div>
<p>We load the data and the packages we use for this section:</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/vsn28Exprs_a26e5710079550b80279c3a55e433c47">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>IBDchip <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">"../data/vsn28Exprd.rds"</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"ade4"</span>)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"factoextra"</span>)</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"sva"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-matrixIBDchip" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.5 </strong></span>What class is the <code>IBDchip</code> ? Look at the last row of the matrix, what do you notice?</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-10-contents" aria-controls="callout-10" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-10" class="callout-10-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/IBDchip_8267aca00f5d5c214025fa8499f7dc32">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(IBDchip)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "matrix" "array" </code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(IBDchip)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8635   28</code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tail</span>(IBDchip[,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                 20CF     20DF     20MF
bm-026.1.sig_st              7.299308 7.275802 7.383103
bm-125.1.sig_st              8.538857 8.998562 9.296096
bru.tab.d.HIII.Con32.sig_st  6.802736 6.777566 6.859950
bru.tab.d.HIII.Con323.sig_st 6.463604 6.501139 6.611851
bru.tab.d.HIII.Con5.sig_st   5.739235 5.666060 5.831079
day                          2.000000 2.000000 2.000000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(IBDchip[<span class="fu">nrow</span>(IBDchip), ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
 1  2  3 
 8 16  4 </code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="page-columns page-full"><p>The data are normalized abundance measurements of 8634 taxa measured on 28 samples. We use a rank-threshold transformation, giving the top 3000 most abundant taxa scores from 3000 to 1, and letting the remaining (low abundant) ones all have a score of 1. We also separate out the proper assay data from the (awkwardly placed) day variable, which should be considered a factor<a href="#fn5" class="footnote-ref" id="fnref5" role="doc-noteref"><sup>5</sup></a>:</p><div class="no-row-height column-margin column-container"><li id="fn5"><p><sup>5</sup>&nbsp;Below, we show how to arrange these data into a Bioconductor <code>SummarizedExperiment</code>, which is a much more sane way of storing such data.</p></li></div></div>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/assayIBD_3785b995cc241248142b12d5265ea7f6">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>assayIBD <span class="ot">=</span> IBDchip[<span class="sc">-</span><span class="fu">nrow</span>(IBDchip), ]</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>day      <span class="ot">=</span> <span class="fu">factor</span>(IBDchip[<span class="fu">nrow</span>(IBDchip), ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Instead of using the continuous, somehow normalized data, we use a robust analysis replacing the values by their ranks. The lower values are considered ties encoded as a threshold chosen to reflect the number of expected taxa thought to be present:</p>
<!-- TODO: these plots aren't shown online? -->
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/fig-screepc12_d57838ba0b5ad77e381be67fdfcd7479">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>rankthreshPCA <span class="ot">=</span> <span class="cf">function</span>(x, <span class="at">threshold =</span> <span class="dv">3000</span>) {</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>  ranksM <span class="ot">=</span> <span class="fu">apply</span>(x, <span class="dv">2</span>, rank)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>  ranksM[ranksM <span class="sc">&lt;</span> threshold] <span class="ot">=</span> threshold</span>
<span id="cb46-4"><a href="#cb46-4" aria-hidden="true" tabindex="-1"></a>  ranksM <span class="ot">=</span> threshold <span class="sc">-</span> ranksM</span>
<span id="cb46-5"><a href="#cb46-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dudi.pca</span>(<span class="fu">t</span>(ranksM), <span class="at">scannf =</span> <span class="cn">FALSE</span>, <span class="at">nf =</span> <span class="dv">2</span>)</span>
<span id="cb46-6"><a href="#cb46-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb46-7"><a href="#cb46-7" aria-hidden="true" tabindex="-1"></a>pcaDay12 <span class="ot">=</span> <span class="fu">rankthreshPCA</span>(assayIBD[, day <span class="sc">!=</span> <span class="dv">3</span>])</span>
<span id="cb46-8"><a href="#cb46-8" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_eig</span>(pcaDay12, <span class="at">bar_width =</span> <span class="fl">0.6</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-screepc12" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-screepc12-1.png" class="img-fluid figure-img" width="320"></p>
<figcaption class="figure-caption">Figure&nbsp;9.13: The screeplot shows us that the samples can be usefully represented in a two dimensional embedding.</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/fig-rankthreshPCA_a0cd3c103e8ff84acf47ad4f570e6eb3">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>day12 <span class="ot">=</span> day[ day<span class="sc">!=</span><span class="dv">3</span> ]</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>rtPCA1 <span class="ot">=</span> <span class="fu">fviz</span>(pcaDay12, <span class="at">element =</span> <span class="st">"ind"</span>, <span class="at">axes =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">geom =</span> <span class="fu">c</span>(<span class="st">"point"</span>, <span class="st">"text"</span>),</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">habillage =</span> day12, <span class="at">repel =</span> <span class="cn">TRUE</span>, <span class="at">palette =</span> <span class="st">"Dark2"</span>,</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">addEllipses =</span> <span class="cn">TRUE</span>, <span class="at">ellipse.type =</span> <span class="st">"convex"</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">""</span>) <span class="sc">+</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>rtPCA1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-rankthreshPCA" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-rankthreshPCA-1.png" class="img-fluid figure-img" width="416"></p>
<figcaption class="figure-caption">Figure&nbsp;9.14: We have used colors to identify the different days and have kept the sample labels as well. We have also added convex hulls for each day. The group mean is identified as the point with the larger symbol (circle, triangle or square).</figcaption>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-rankthreshold" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.6 </strong></span>Why do we use a threshold for the ranks?</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-12-contents" aria-controls="callout-12" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-12" class="callout-12-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>Low abundances, at noise level occur for species that are not really present, of which there are more than half. A large jump in rank for these observations could easily occur without any meaningful reason. Thus we create a large number of ties for low abundance.</p>
</div>
</div>
</div>
<p><a href="#fig-rankthreshPCA" class="quarto-xref">Figure&nbsp;<span>9.14</span></a> shows that the sample arrange themselves naturally into two different groups according to the day of the samples. After discovering this effect, we delved into the differences that could explain these distinct clusters. There were two different protocols used (protocol 1 on day 1, protocol 2 on day 2) <em>and</em> unfortunately two different provenances for the arrays used on those two days (array 1 on day 1, array 2 on day 2).</p>
<p>A third set of data of four samples had to be collected to deconvolve the confounding effect. Array 2 was used with protocol 2 on Day 3, <a href="#fig-Threesetspca123" class="quarto-xref">Figure&nbsp;<span>9.15</span></a> shows the new PCA plot with all the samples created by the following:</p>
<div class="cell" data-layout-align="center" data-warning.known="unlabeled data points" data-hash="09-chap_cache/html/Threesetspca123-code_6d1b5b02cb9cc030e0b72557894a47c3">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>pcaDay123 <span class="ot">=</span> <span class="fu">rankthreshPCA</span>(assayIBD)</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz</span>(pcaDay123, <span class="at">element =</span> <span class="st">"ind"</span>, <span class="at">axes =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">geom =</span> <span class="fu">c</span>(<span class="st">"point"</span>, <span class="st">"text"</span>),</span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">habillage =</span> day, <span class="at">repel =</span> <span class="cn">TRUE</span>, <span class="at">palette =</span> <span class="st">"Dark2"</span>,</span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">addEllipses =</span> <span class="cn">TRUE</span>, <span class="at">ellipse.type =</span> <span class="st">"convex"</span>) <span class="sc">+</span> </span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">""</span>) <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="fig-Threesetspca123" class="cell quarto-layout-panel page-columns page-full" data-warning.known="unlabeled data points" data-hash="09-chap_cache/html/fig-Threesetspca123_a5c516cc6ca7ae54e7ff28ed152d6bb4">
<figure class="figure page-columns page-full">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-Threesetspca123-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-Threesetspca123-1.png" class="img-fluid figure-img" data-ref-parent="fig-Threesetspca123" width="360"></p>
<figcaption class="figure-caption">(a)</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-Threesetspca123-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-Threesetspca123-2.png" class="img-fluid figure-img" data-ref-parent="fig-Threesetspca123" width="360"></p>
<figcaption class="figure-caption">(b)</figcaption>
</figure>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption margin-caption">Figure&nbsp;9.15: When comparing the three day analysis to that of the first two days, we notice the inversion of signs in the coordinates on the second axis: this has no biological relevance. The important finding is that group 3 overlaps heavily with group 1 indicating that it was the protocol change on Day 2 which created the variability.</figcaption><p></p>
</figure>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-ellipseIBD" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.7 </strong></span>In which situation would it be preferable to make confidence ellipses around the group means using the following code?</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/exCodeEllipse_68c4f7b988a72c76b2616fd480045b32">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_pca_ind</span>(pcaDay123, <span class="at">habillage =</span> day, <span class="at">labelsize =</span> <span class="dv">3</span>,</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">palette =</span> <span class="st">"Dark2"</span>, <span class="at">addEllipses =</span> <span class="cn">TRUE</span>, <span class="at">ellipse.level =</span> <span class="fl">0.69</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-screepc123_2976cd559d4e04a049a8b1f73207c29c">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-screepc123" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-screepc123-1.png" class="img-fluid figure-img" width="320"></p>
<figcaption class="figure-caption">Figure&nbsp;9.16: The eigenvalue screeplot the case of 3 groups is extremely similar to that with two groups shown in <a href="#fig-screepc12" class="quarto-xref">Figure&nbsp;<span>9.13</span></a>.</figcaption>
</figure>
</div>
</div></div></div>
<p>Through this visualization we were able to uncover a flaw in the original experimental design. The first two batches shown in green and brown were both balanced with regards to IBS and healthy rats. They do show very different levels of variability and overall multivariate coordinates. In fact, there are two <strong>confounded</strong> effects. Both the arrays and protocols were different on those two days. We had to run a third batch of experiments on day 3, represented in purple, this used protocol from day 1 and the arrays from day 2. The third group faithfully overlaps with batch 1, telling us that the change in protocol was responsible for the variability.</p>
</section>
<section id="removing-batch-effects" class="level3 page-columns page-full" data-number="9.3.2">
<h3 data-number="9.3.2" class="anchored" data-anchor-id="removing-batch-effects"><span class="header-section-number">9.3.2</span> Removing batch effects</h3>
<div class="page-columns page-full"><p>Through the combination of the continuous measurements from <code>assayIBD</code> and the <strong>supplementary</strong> batch number as a factor, the PCA map has provided an invaluable investigation tool. This is a good example of the use of <strong>supplementary points</strong><a href="#fn6" class="footnote-ref" id="fnref6" role="doc-noteref"><sup>6</sup></a>. The mean-barycenter points are created by using the group-means of points in each of the three groups and serve as extra markers on the plot.</p><div class="no-row-height column-margin column-container"><li id="fn6"><p><sup>6</sup>&nbsp;This is called a supplementary point because the new observation-point is not used in the matrix decomposition.</p></li></div></div>
<p>We can decide to re-align the three groups by subtracting the group means so that all the batches are centered on the origin. A slightly more effective way is to use the <code>ComBat</code> function available in the <strong><a href="https://bioconductor.org/packages/sva/">sva</a></strong> package. This function uses a similar, but slightly more sophisticated method (Empirical Bayes mixture approach <span class="citation" data-cites="Leek:2010:batch">(<a href="16-chap.html#ref-Leek:2010:batch" role="doc-biblioref">Leek et al. 2010</a>)</span>). We can see its effect on the data by redoing our robust PCA (see the result in <a href="#fig-CombatIBD" class="quarto-xref">Figure&nbsp;<span>9.17</span></a>):</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-CombatIBD_d14f8388b8f6607a7ee8ff73a1823ef9">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>model0 <span class="ot">=</span> <span class="fu">model.matrix</span>(<span class="sc">~</span><span class="dv">1</span>, day)</span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>combatIBD <span class="ot">=</span> <span class="fu">ComBat</span>(<span class="at">dat =</span> assayIBD, <span class="at">batch =</span> day, <span class="at">mod =</span> model0)</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>pcaDayBatRM <span class="ot">=</span> <span class="fu">rankthreshPCA</span>(combatIBD)</span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz</span>(pcaDayBatRM, <span class="at">element =</span> <span class="st">"ind"</span>, <span class="at">geom =</span> <span class="fu">c</span>(<span class="st">"point"</span>, <span class="st">"text"</span>),</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">habillage =</span> day, <span class="at">repel=</span><span class="cn">TRUE</span>, <span class="at">palette =</span> <span class="st">"Dark2"</span>, <span class="at">addEllipses =</span> <span class="cn">TRUE</span>,</span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">ellipse.type =</span> <span class="st">"convex"</span>, <span class="at">axes =</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>)) <span class="sc">+</span> <span class="fu">coord_fixed</span>() <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-CombatIBD" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-CombatIBD-1.png" class="img-fluid figure-img" width="416"></p>
<figcaption class="figure-caption">Figure&nbsp;9.17: The modified data with the batch effects removed now show three batch-groups heavily overlapping and centered almost at the origin.</figcaption>
</figure>
</div>
</div></div></div>
</section>
<section id="hybrid-data-and-bioconductor-containers" class="level3 page-columns page-full" data-number="9.3.3">
<h3 data-number="9.3.3" class="anchored" data-anchor-id="hybrid-data-and-bioconductor-containers"><span class="header-section-number">9.3.3</span> Hybrid data and Bioconductor containers</h3>
<p>A more rational way of combining the batch and treatment information into compartments of a composite object is to use the <em>SummarizedExperiment</em> class. It includes special <em>slots</em> for the assay(s) where rows represent features of interest (e.g., genes, transcripts, exons, etc.) and columns represent samples. Supplementary information about the features can be stored in a <em>DataFrame</em> object, accessible using the function <code>rowData</code>. Each row of the <em>DataFrame</em> provides information on the feature in the corresponding row of the <em>SummarizedExperiment</em> object.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/unnamed-chunk-119_dbcb4672ff55f102a404bc86c437d3e7">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/devil.png" class="img-fluid figure-img" width="123"></p>
<figcaption class="figure-caption">A confusing notational similarity occurs here, in the SummarizedExperiment framework a <code>DataFrame</code> is not the same as a <em>data.frame.</em></figcaption>
</figure>
</div>
</div></div></div>
<p>Here we insert the two covariates day and treatment in the <code>colData</code> object and combine it with assay data in a new <em>SummarizedExperiment</em> object.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"SummarizedExperiment"</span>)</span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>treatment  <span class="ot">=</span> <span class="fu">factor</span>(<span class="fu">ifelse</span>(<span class="fu">grepl</span>(<span class="st">"Cntr|^C"</span>, <span class="fu">colnames</span>(IBDchip)), <span class="st">"CTL"</span>, <span class="st">"IBS"</span>))</span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a>sampledata <span class="ot">=</span> <span class="fu">DataFrame</span>(<span class="at">day =</span> day, <span class="at">treatment =</span> treatment)</span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a>chipse <span class="ot">=</span> <span class="fu">SummarizedExperiment</span>(<span class="at">assays  =</span> <span class="fu">list</span>(<span class="at">abundance =</span> assayIBD),</span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a>                              <span class="at">colData =</span> sampledata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This is the best way to keep all the relevant data together, it will also enable you to quickly filter the data while keeping all the information aligned properly.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/unnamed-chunk-120_8b3ec29318af435fd25a04e15a676a53">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/devil.png" class="img-fluid figure-img" width="123"></p>
<figcaption class="figure-caption">You can explore composite objects using the Environment pane in RStudio. You will see that in <code>chipse</code>, some of the slots are empty.</figcaption>
</figure>
</div>
</div></div></div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-SEfilter" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.8 </strong></span>Make a new <em>SummarizedExperiment</em> object by choosing the subset of the samples that were created on day 2.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-15-contents" aria-controls="callout-15" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-15" class="callout-15-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/SEfilter_b1c7590f22cc9d8724f85964f9b16157">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>chipse[, day <span class="sc">==</span> <span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: SummarizedExperiment 
dim: 8634 16 
metadata(0):
assays(1): abundance
rownames(8634): 01010101000000.2104_gPM_GC 01010101000000.2141_gPM_GC
  ... bru.tab.d.HIII.Con323.sig_st bru.tab.d.HIII.Con5.sig_st
rowData names(0):
colnames(16): 20CF 20DF ... IBSM IBSP
colData names(2): day treatment</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Columns of the <em>DataFrame</em> represent different attributes of the features of interest, e.g., gene or transcript IDs, etc. Here is an example of hybrid data container from single cell experiments (see Bioconductor workflow in <span class="citation" data-cites="Perraudeau:2017">Perraudeau et al. (<a href="16-chap.html#ref-Perraudeau:2017" role="doc-biblioref">2017</a>)</span> for more details).</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>corese <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">"../data/normse.rds"</span>)</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>norm <span class="ot">=</span> <span class="fu">assays</span>(corese)<span class="sc">$</span>normalizedValues</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>After the pre-processing and normalization steps prescribed in the workflow, we retain the 1000 most variable genes measured on 747 cells.</p>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-coldatacore" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.9 </strong></span>How many different batches do the cells belong to ?</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-17-contents" aria-controls="callout-17" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-17" class="callout-17-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/coldatacore_9db0339524d37dd4c0c206e893d72565">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(<span class="fu">unique</span>(<span class="fu">colData</span>(corese)<span class="sc">$</span>Batch))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 18</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>We can look at a PCA of the normalized values and check graphically that the batch effect has been removed:</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-screeplotnorm_d936f7f47bf97b2c1c1903815ae5629d">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>respca <span class="ot">=</span> <span class="fu">dudi.pca</span>(<span class="fu">t</span>(norm), <span class="at">nf =</span> <span class="dv">3</span>, <span class="at">scannf =</span> <span class="cn">FALSE</span>)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotscree</span>(respca, <span class="dv">15</span>)</span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>PCS <span class="ot">=</span> respca<span class="sc">$</span>li[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-screeplotnorm" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-screeplotnorm-1.png" class="img-fluid figure-img" width="320"></p>
<figcaption class="figure-caption">Figure&nbsp;9.18: Screeplot of the PCA of the normalized data.</figcaption>
</figure>
</div>
</div></div></div>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/unnamed-chunk-121_88d6f1ffa1728753e02966587f78c728">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/devil.png" class="img-fluid figure-img" width="123"></p>
<figcaption class="figure-caption">We have set up colors for the clusters as in the workflow, (the code is not shown here).</figcaption>
</figure>
</div>
</div></div></div>
<p>Since the screeplot in <a href="#fig-screeplotnorm" class="quarto-xref">Figure&nbsp;<span>9.18</span></a> shows us that we must not dissociate axes 2 and 3, we will make a three dimensional plot with the <strong><a href="https://cran.r-project.org/web/packages/rgl/">rgl</a></strong> package. We use the following interactive code:</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/screeplotnorm-2_1c039c62c16bbb6c2397d968ced6c46a">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"rgl"</span>)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>batch <span class="ot">=</span> <span class="fu">colData</span>(corese)<span class="sc">$</span>Batch</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot3d</span>(PCS,<span class="at">aspect=</span><span class="fu">sqrt</span>(<span class="fu">c</span>(<span class="dv">84</span>,<span class="dv">24</span>,<span class="dv">20</span>)),<span class="at">col=</span>col_clus[batch])</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot3d</span>(PCS,<span class="at">aspect=</span><span class="fu">sqrt</span>(<span class="fu">c</span>(<span class="dv">84</span>,<span class="dv">24</span>,<span class="dv">20</span>)),</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a><span class="at">col =</span> col_clus[<span class="fu">as.character</span>(publishedClusters)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="fig-3dplotsnorm" class="cell quarto-layout-panel page-columns page-full" data-hash="09-chap_cache/html/fig-3dplotsnorm_a466b29fe7e6b4b511dd91176eaddbd9">
<figure class="figure page-columns page-full">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-3dplotsnorm-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="imgs/plotnormpcabatch1.png" class="img-fluid figure-img" data-ref-parent="fig-3dplotsnorm"></p>
<figcaption class="figure-caption">(a)</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-3dplotsnorm-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="imgs/plotnormpclust1.png" class="img-fluid figure-img" data-ref-parent="fig-3dplotsnorm"></p>
<figcaption class="figure-caption">(b)</figcaption>
</figure>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption margin-caption">Figure&nbsp;9.19: Two-dimensional screenshots of three-dimensional <strong><a href="https://cran.r-project.org/web/packages/rgl/">rgl</a></strong> plots. The points are colored according to batch numbers in (a), and according to the original clustering in (b). We can see that the batch effect has been effectively removed and that the cells show the original clustering.</figcaption><p></p>
</figure>
</div>
<p><strong>Note:</strong> Of course, the book medium is limiting here, as we are showing two static projections that do not do justice to the depth available when looking at the interactive dynamic plots as they appear using the <code>plot3d</code> function. We encourage the reader to experiment extensively with these and other interactive packages and they provide a much more intuitive experience of the data.</p>
</section>
</section>
<section id="sec-hetero-CA" class="level2 page-columns page-full" data-number="9.4">
<h2 data-number="9.4" class="anchored" data-anchor-id="sec-hetero-CA"><span class="header-section-number">9.4</span> Correspondence analysis for contingency tables</h2>
<section id="cross-tabulation-and-contingency-tables" class="level3 page-columns page-full" data-number="9.4.1">
<h3 data-number="9.4.1" class="anchored" data-anchor-id="cross-tabulation-and-contingency-tables"><span class="header-section-number">9.4.1</span> Cross-tabulation and contingency tables</h3>
<p>Categorical data abound in biological settings: sequence status (CpG/non-CpG), phenotypes, taxa are often coded as factors as we saw in Chapter 2. Cross-tabulation of two such variables gives us a <strong>contingency table</strong>; the result of counting the co-occurrence of two phenotypes (sex and colorblindness was such an example). We saw that the first step is to look at the independence of the two categorical variables; the standard statistical measure of independence uses the <strong>chisquare distance</strong>. This quantity will replace the variance we used for continuous measurements.</p>
<p>The columns and rows of the table have the same `status’ and we are not in supervised/regression type setting. We won’t see a sample/variable divide; as a consequence the rows and columns will have the same status and we will ‘center’ both the rows and the columns. This symmetry will also translate in our use of <strong>biplots</strong> where both dimensions appear on the same plot.</p>
<div class="cell tbl-cap-location-top page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/tbl-HIV_80919b0803a1530426abdd77680d7373">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="tbl-HIV" class="anchored">
<table data-quarto-postprocess="true" class="table table-sm table-striped small">
<caption>Table&nbsp;9.1: Sample by mutation matrix.</caption>
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th">Patient</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Mut1</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Mut2</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Mut3</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">...</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">AHX112</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">AHX717</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">AHX543</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;"></td>
</tr>
</tbody>
</table>
</div>


</div></div></div>
<section id="transforming-the-data-to-tabular-form." class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="transforming-the-data-to-tabular-form.">Transforming the data to tabular form.</h4>
<p>If the data are collected as long lists with each subject (or sample) associated to its levels of the categorical variables, we may want to transform them into a contingency table. Here is an example. In <a href="#tbl-HIV" class="quarto-xref">Table&nbsp;<span>9.1</span></a> HIV mutations are tabulated as indicator (0/1) binary variables. These data are then transformed into a <strong>mutation co-occurrence matrix</strong> shown in <a href="#tbl-crossHIV" class="quarto-xref">Table&nbsp;<span>9.2</span></a>.</p>
<div class="cell tbl-cap-location-top page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/tbl-crossHIV_84a2219edaf49d8218129df7de8ea450">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="tbl-crossHIV" class="anchored">
<table data-quarto-postprocess="true" class="table table-sm table-striped small">
<caption>Table&nbsp;9.2: Cross-tabulation of the HIV mutations showing two-way co-occurrences.</caption>
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th">Patient</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Mut1</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Mut2</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Mut3</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">...</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Mut1</td>
<td style="text-align: right;">853</td>
<td style="text-align: right;">29</td>
<td style="text-align: right;">10</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">Mut2</td>
<td style="text-align: right;">29</td>
<td style="text-align: right;">853</td>
<td style="text-align: right;">52</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Mut3</td>
<td style="text-align: right;">10</td>
<td style="text-align: right;">52</td>
<td style="text-align: right;">853</td>
<td style="text-align: left;"></td>
</tr>
</tbody>
</table>
</div>


</div></div></div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-suffHIV" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.10 </strong></span>What information is lost in this cross-tabulation ?<br>
When will this matter?</p>
</div>
</div>
</div>
<p>Here are some co-occurrence data from the HIV database <span class="citation" data-cites="HIVdb">(<a href="16-chap.html#ref-HIVdb" role="doc-biblioref">Rhee et al. 2003</a>)</span>. Some of these mutations have a tendency to co-occur.</p>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-IndHIV" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.11 </strong></span>Test the hypothesis of independence of the mutations.</p>
</div>
</div>
</div>
<p>Before explaining the details of how correspondence analysis works, let’s look at the output of one of many correspondence analysis functions. We use <code>dudi.coa</code> from the <strong><a href="https://cran.r-project.org/web/packages/ade4/">ade4</a></strong> package to plot the mutations in a lower dimensional projection; the procedure follows what we did for PCA.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-HIVnnrti_c07b262fe576b9b04e32119f1c7b61e7">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>cooc <span class="ot">=</span> <span class="fu">read.delim2</span>(<span class="st">"../data/coccurHIV.txt"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">sep =</span> <span class="st">","</span>)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>cooc[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">11</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    X4S X6D X6K X11R X20R X21I X35I X35L X35M X35T X39A
4S    0  28   8    0   99    0   22    5   15    3   45
6D   26   0   0   34  131    0  108    4   30   13   84
6K    7   0   0    6   45    0    5   13   38   35   12
11R   0  35   7    0  127   12   60   17   15    6   42</code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>HIVca <span class="ot">=</span> <span class="fu">dudi.coa</span>(cooc, <span class="at">nf =</span> <span class="dv">4</span>, <span class="at">scannf =</span> <span class="cn">FALSE</span>)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_eig</span>(HIVca, <span class="at">geom =</span> <span class="st">"bar"</span>, <span class="at">bar_width =</span> <span class="fl">0.6</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-HIVnnrti" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-HIVnnrti-1.png" class="img-fluid figure-img" width="320"></p>
<figcaption class="figure-caption">Figure&nbsp;9.20: The dependencies between HIV mutations is clearly a three dimensional phenomenon, the three first eigenvalues show a clear signal in the data.</figcaption>
</figure>
</div>
</div></div></div>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-HIV3d_c300cc725276b3be78a4ec8ada12b0aa">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-HIV3d" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="imgs/scatter3d-HIV.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;9.21: A screenshot of the output from an interactive 3d plotting function (<code>plot3d</code>).</figcaption>
</figure>
</div>
</div></div></div>
<p>After looking at a screeplot, we see that dimensionality of the underlying variation is definitely three dimensional, we plot these three dimensions. Ideally this would be done with an interactive three-dimensional plotting function such as that provided through the package <strong><a href="https://cran.r-project.org/web/packages/rgl/">rgl</a></strong> as shown in <a href="#fig-HIV3d" class="quarto-xref">Figure&nbsp;<span>9.21</span></a>.</p>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-3DonHIV" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.12 </strong></span>Using the <strong><a href="https://cran.r-project.org/web/packages/car/">car</a></strong> and <strong><a href="https://cran.r-project.org/web/packages/rgl/">rgl</a></strong> packages make 3d scatterplot similar to <a href="#fig-HIV3d" class="quarto-xref">Figure&nbsp;<span>9.21</span></a>.<br>
Compare to the plot obtained using aspect=FALSE with the plot3d function from <strong><a href="https://cran.r-project.org/web/packages/rgl/">rgl</a></strong>.<br>
What structure do you notice by rotating the cloud of points?</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-21-contents" aria-controls="callout-21" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-21" class="callout-21-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/CA123-HIV_77561f3dbbfc69ab8ddcc5b11fa6912c">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"rgl"</span>)</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>CA1<span class="ot">=</span>HIVca<span class="sc">$</span>li[,<span class="dv">1</span>];CA2<span class="ot">=</span>HIVca<span class="sc">$</span>li[,<span class="dv">2</span>];CA3<span class="ot">=</span>HIVca<span class="sc">$</span>li[,<span class="dv">3</span>]</span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot3d</span>(CA1,CA2,CA3,<span class="at">aspect=</span><span class="cn">FALSE</span>,<span class="at">col=</span><span class="st">"purple"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="page-columns page-full">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_ca_row</span>(HIVca,<span class="at">axes =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>),<span class="at">geom=</span><span class="st">"text"</span>, <span class="at">col.row=</span><span class="st">"purple"</span>,</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">labelsize=</span><span class="dv">3</span>)<span class="sc">+</span><span class="fu">ggtitle</span>(<span class="st">""</span>) <span class="sc">+</span> <span class="fu">xlim</span>(<span class="sc">-</span><span class="fl">0.55</span>, <span class="fl">1.7</span>) <span class="sc">+</span> <span class="fu">ylim</span>(<span class="sc">-</span><span class="fl">0.53</span>,<span class="fl">1.1</span>) <span class="sc">+</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span>  <span class="fu">coord_fixed</span>()</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_ca_row</span>(HIVca,<span class="at">axes =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>), <span class="at">geom=</span><span class="st">"text"</span>,<span class="at">col.row=</span><span class="st">"purple"</span>,</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">labelsize=</span><span class="dv">3</span>)<span class="sc">+</span><span class="fu">ggtitle</span>(<span class="st">""</span>)<span class="sc">+</span> <span class="fu">xlim</span>(<span class="sc">-</span><span class="fl">0.55</span>, <span class="fl">1.7</span>)<span class="sc">+</span><span class="fu">ylim</span>(<span class="sc">-</span><span class="fl">0.5</span>,<span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-HIVca" class="cell quarto-layout-panel page-columns page-full" data-hash="09-chap_cache/html/fig-HIVca_e69b9502a0dbc6b8d253ef9d1ddeb13b">
<figure class="figure page-columns page-full">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-HIVca-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-HIVca-1.png" class="img-fluid figure-img" data-ref-parent="fig-HIVca" width="360"></p>
<figcaption class="figure-caption">(a)</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-HIVca-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-HIVca-2.png" class="img-fluid figure-img" data-ref-parent="fig-HIVca" width="360"></p>
<figcaption class="figure-caption">(b)</figcaption>
</figure>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption margin-caption">Figure&nbsp;9.22: Two planar maps of the mutations defined with the horizontal axis corresponding to the first eigenvector of the CA and the vertical axis being the second axis in (a), and the third in (b); notice the difference in heights.</figcaption><p></p>
</figure>
</div>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-plane13" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.13 </strong></span>Show the code for plotting the plane defined by axes 1 and 3 of the correspondence analysis respecting the scaling of the vertical axis as shown in the bottom figure of <a href="#fig-HIVca" class="quarto-xref">Figure&nbsp;<span>9.22</span></a>.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-23-contents" aria-controls="callout-23" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-23" class="callout-23-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/HIVnnrtiMut13a_ea709bebd8f16e3e8f2ba4a2ed831558">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_ca_row</span>(HIVca, <span class="at">axes=</span><span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>), <span class="at">geom=</span><span class="st">"text"</span>, <span class="at">col.row=</span><span class="st">"purple"</span>, <span class="at">labelsize=</span><span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">""</span>) <span class="sc">+</span> <span class="fu">theme_minimal</span>() <span class="sc">+</span> <span class="fu">coord_fixed</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<p>This first example showed how to map all the different levels of one categorical variable (the mutations) in a similar way to how PCA projects continuous variables. We will now explore how this can be extended to two or more categorical variables.</p>
</section>
</section>
<section id="hair-color-eye-color-and-phenotype-co-occurrence" class="level3 page-columns page-full" data-number="9.4.2">
<h3 data-number="9.4.2" class="anchored" data-anchor-id="hair-color-eye-color-and-phenotype-co-occurrence"><span class="header-section-number">9.4.2</span> Hair color, eye color and phenotype co-occurrence</h3>
<p>We will consider a small table, so we can follow the analysis in detail. The data are a contingency table of hair-color and eye-color phenotypic co-occurrence from students as shown in <a href="#tbl-HairEye" class="quarto-xref">Table&nbsp;<span>9.3</span></a>. In <a href="02-chap.html" class="quarto-xref"><span>Chapter&nbsp;2</span></a>, we used a <span class="math inline">\(\chi^2\)</span> test to detect possible dependencies:</p>
<div class="cell" data-layout-align="center" data-warning.known="approximation may be incorrect" data-hash="09-chap_cache/html/chisquaredtest_cc0c37e864b09f088541cfe7db1a76ab">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>HairColor <span class="ot">=</span> HairEyeColor[,,<span class="dv">2</span>]</span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a><span class="fu">chisq.test</span>(HairColor)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Pearson's Chi-squared test

data:  HairColor
X-squared = 106.66, df = 9, p-value &lt; 2.2e-16</code></pre>
</div>
</div>
<div class="cell tbl-cap-location-top page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/tbl-HairEye_2ac4a956a2b639488e5a579719f4b871">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="tbl-HairEye" class="anchored">
<table data-quarto-postprocess="true" class="table table-sm table-striped small">
<caption>Table&nbsp;9.3: Cross tabulation of students hair and eye color.</caption>
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th"></th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Brown</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Blue</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Hazel</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Green</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Black</td>
<td style="text-align: right;">36</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Brown</td>
<td style="text-align: right;">66</td>
<td style="text-align: right;">34</td>
<td style="text-align: right;">29</td>
<td style="text-align: right;">14</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Red</td>
<td style="text-align: right;">16</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="even">
<td style="text-align: left;">Blond</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">64</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">8</td>
</tr>
</tbody>
</table>
</div>


</div></div></div>
<p>However, stating <em>non independence</em> between hair and eye color is not enough. We need a more detailed explanation of where the dependencies occur: which hair color occurs more often with green eyes ? Are some of the variable levels independent? In fact we can study the departure from independence using a special weighted version of SVD. This method can be understood as a simple extension of PCA and MDS to contingency tables.</p>
<section id="independence-computationally-and-visually." class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="independence-computationally-and-visually.">Independence: computationally and visually.</h4>
<p>We start by computing the row and column sums; we use these to build the table that would be expected if the two phenotypes were independent. We call this expected table <code>HCexp</code>.</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/ExpectedEyes_be5dc0d5b4a6970b8aa8daac57db20df">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>rowsums <span class="ot">=</span> <span class="fu">as.matrix</span>(<span class="fu">apply</span>(HairColor, <span class="dv">1</span>, sum))</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>rowsums</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      [,1]
Black   52
Brown  143
Red     37
Blond   81</code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a>colsums <span class="ot">=</span> <span class="fu">as.matrix</span>(<span class="fu">apply</span>(HairColor, <span class="dv">2</span>, sum))</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a><span class="fu">t</span>(colsums)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Brown Blue Hazel Green
[1,]   122  114    46    31</code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>HCexp <span class="ot">=</span> rowsums <span class="sc">%*%</span><span class="fu">t</span> (colsums) <span class="sc">/</span> <span class="fu">sum</span>(colsums)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<div class="no-row-height column-margin column-container"><div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/fig-HCexp_f3b17f3974fc470a5a6d9810167c665e">

</div></div><p>Now we compute the <span class="math inline">\(\chi^2\)</span> (chi-squared) statistic, which is the sum of the scaled residuals for each of the cells of the table:</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/unnamed-chunk-62_424bdbf3795676268bc4928417ff6be1">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>((HairColor  <span class="sc">-</span> HCexp)<span class="sc">^</span><span class="dv">2</span><span class="sc">/</span>HCexp)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 106.6637</code></pre>
</div>
</div>
<p>We can study these residuals from the expected table, first numerically then in <a href="#fig-MosaicHair" class="quarto-xref">Figure&nbsp;<span>9.23</span></a>.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-MosaicHair_24524af793db6a96e74145acf805f8f2">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(<span class="fu">t</span>(HairColor<span class="sc">-</span>HCexp))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Hair
Eye     Black Brown Red Blond
  Brown    16    10   2   -28
  Blue    -10   -18  -6    34
  Hazel    -3     8   2    -7
  Green    -3     0   3     0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"vcd"</span>)</span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mosaicplot</span>(HairColor, <span class="at">shade=</span><span class="cn">TRUE</span>, <span class="at">las=</span><span class="dv">1</span>, <span class="at">type=</span><span class="st">"pearson"</span>, <span class="at">cex.axis=</span><span class="fl">0.7</span>, <span class="at">main=</span><span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div id="fig-MosaicHair" class="quarto-figure quarto-figure-center anchored page-columns page-full">
<figure class="figure page-columns page-full">
<p><img src="09-chap_files/figure-html/fig-MosaicHair-1.png" class="img-fluid figure-img" style="width:75.0%"></p>
<figcaption class="figure-caption margin-caption">Figure&nbsp;9.23: Visualization of the departure from independence. Now, the boxes are proportional in size to the actual observed counts and we no longer have a ‘rectangular’ property. The departure from independence is measured in Chisquared distance for each of the boxes and colored according to whether the residuals are large and positive. Dark blue indicates a positive association, for instance between blue eyes and blonde hair, red indicates a negative association such as in the case of blond hair and brown eyes.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="mathematical-formulation." class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="mathematical-formulation.">Mathematical Formulation.</h4>
<p>Here are the computations we just did in R in a more mathematical form. For a general contingency table <span class="math inline">\({\mathbf N}\)</span> with <span class="math inline">\(I\)</span> rows and <span class="math inline">\(J\)</span> columns and a total sample size of <span class="math inline">\(n=\sum_{i=1}^I \sum_{j=1}^J n_{ij}= n_{\cdot \cdot}\)</span>. If the two categorical variables were independent, each cell frequency would be approximately equal to</p>
<p><span class="math display">\[
n_{ij} = \frac{n_{i \cdot}}{n} \frac{n_{\cdot j}}{n} \times n
\]</span></p>
<p>can also be written:</p>
<p><span class="math display">\[
{\mathbf N} = {\mathbf c r'} \times n, \qquad
\mbox{ where } c= \frac{1}{n} {{\mathbf N}} {\mathbb 1}_m
\;\mbox{ and }\; r'=\frac{1}{n} {\mathbf N}' {\mathbb 1}_p
\]</span></p>
<p>The departure from independence is measured by the <span class="math inline">\(\chi^2\)</span> statistic</p>
<p><span class="math display">\[
{\cal X}^2=\sum_{i,j}
\frac{\left(n_{ij}-\frac{n_{i\cdot}}{n}\frac{n_{\cdot j}}{n}n\right)^2}
{\frac{n_{i\cdot}n_{\cdot j}}{n^2}n}
\]</span></p>
<div class="page-columns page-full"><p> Once we have ascertained that the two variables are not independent, we use a weighted multidimensional scaling using <span class="math inline">\(\chi^2\)</span> distances to visualize the associations.</p><div class="no-row-height column-margin column-container"><span class=""><strong>Correspondece Analysis functions</strong> <code>CCA</code> in <strong><a href="https://cran.r-project.org/web/packages/vegan/">vegan</a></strong>, <code>CA</code> in <strong><a href="https://cran.r-project.org/web/packages/FactoMineR/">FactoMineR</a></strong>, <code>ordinate</code> in <strong><a href="https://bioconductor.org/packages/phyloseq/">phyloseq</a></strong>, <code>dudi.coa</code> in <strong><a href="https://cran.r-project.org/web/packages/ade4/">ade4</a></strong>.</span></div></div>
<p>The method is called <strong>Correspondence Analysis (CA)</strong> or <strong>Dual Scaling</strong> and there are multiple R packages that implement it.</p>
<p>Here we make a simple biplot of the Hair and Eye colors.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-HCscatter_1772065a9c43f347981a171e69c45e98">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a>HC <span class="ot">=</span> <span class="fu">as.data.frame.matrix</span>(HairColor)</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>coaHC <span class="ot">=</span> <span class="fu">dudi.coa</span>(HC,<span class="at">scannf=</span><span class="cn">FALSE</span>,<span class="at">nf=</span><span class="dv">2</span>)</span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(coaHC<span class="sc">$</span>eig[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]<span class="sc">/</span><span class="fu">sum</span>(coaHC<span class="sc">$</span>eig)<span class="sc">*</span><span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 89 10  2</code></pre>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_ca_biplot</span>(coaHC, <span class="at">repel=</span><span class="cn">TRUE</span>, <span class="at">col.col=</span><span class="st">"brown"</span>, <span class="at">col.row=</span><span class="st">"purple"</span>) <span class="sc">+</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">""</span>) <span class="sc">+</span> <span class="fu">ylim</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.5</span>,<span class="fl">0.5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display page-columns page-full">
<div id="fig-HCscatter" class="quarto-figure quarto-figure-center anchored page-columns page-full">
<figure class="figure page-columns page-full">
<p><img src="09-chap_files/figure-html/fig-HCscatter-1.png" class="img-fluid figure-img" style="width:75.0%"></p>
<figcaption class="figure-caption margin-caption">Figure&nbsp;9.24: The CA plot gives a representation of a large proportion of the chisquare distance between the data and the values expected under independence. The first axis shows a contrast between black haired and blonde haired students, mirrored by the brown eye, blue eye opposition. In CA the two categories play symmetric roles and we can interpret the proximity of Blue eyes and Blond hair has meaning that there is strong co-occurence of these categories.</figcaption>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-percentagechi" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.14 </strong></span>What percentage of the Chisquare statistic is explained by the first two axes of the Correspondence Analysis?</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-compareCCA" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.15 </strong></span>Compare the results with those obtained by using <code>CCA</code> in the <strong><a href="https://cran.r-project.org/web/packages/vegan/">vegan</a></strong> package with the appropriate value for the <code>scaling</code> parameter.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-26-contents" aria-controls="callout-26" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-26" class="callout-26-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/VeganCCA_f3d5da5341a672604022723ed2ddcef3">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"vegan"</span>)</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>res.ca <span class="ot">=</span> vegan<span class="sc">::</span><span class="fu">cca</span>(HairColor)</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(res.ca, <span class="at">scaling=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
<section id="interpreting-the-biplots" class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="interpreting-the-biplots">Interpreting the biplots</h4>
<p>CA has a special barycentric property: the biplot scaling is chosen so that the row points are placed at the center of gravity of the column levels with their respective weights. For instance, the Blue eyes column point is at the center gravity of the (Black, Brown, Red, Blond) with weights proportional to (9, 34, 7, 64). The Blond row point is very heavily weighted, this is why <a href="#fig-HCscatter" class="quarto-xref">Figure&nbsp;<span>9.24</span></a> shows Blond and Blue quite close together.</p>
</section>
</section>
</section>
<section id="sec-hetero-findingtime" class="level2 page-columns page-full" data-number="9.5">
<h2 data-number="9.5" class="anchored" data-anchor-id="sec-hetero-findingtime"><span class="header-section-number">9.5</span> Finding time…and other important gradients.</h2>
<p>All the methods we have studied in the last sections are commonly known as <strong>ordination</strong> methods. In the same way <strong>clustering</strong> allowed us to detect and interpret a hidden factor/categorical variable, ordination enables us to detect and interpret a hidden ordering, gradient or latent variable in the data.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/ProustProxy_fb4abd0473802cb24d87060fd8ff2315">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/ProustProxy.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div></div></div>
<div class="page-columns page-full"><p> Ecologists have a long history of interpreting the arches formed by observations points in correspondence analysis and principal components as ecological gradients <span class="citation" data-cites="Prentice:1977">(<a href="16-chap.html#ref-Prentice:1977" role="doc-biblioref">Prentice 1977</a>)</span>. Let’s illustrate this first with a very simple data set on which we perform a correspondence analysis.</p><div class="no-row-height column-margin column-container"><span class="">The first examples of seriation or chronology detection was that of archaelogical artifacts by <span class="citation" data-cites="Kendall:1969">Kendall (<a href="16-chap.html#ref-Kendall:1969" role="doc-biblioref">1969</a>)</span>, who used presence/absence of features on pottery to date them. These so-called seriation methods are still relevant today as we follow developmental trajectories in single cell data for instance.</span></div></div>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/LakeCAr_ec738147d6fc7bd27c8ef8edb3b20ab1">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../data/lakes.RData"</span>)</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>lakelike[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     plant1 plant2 plant3 plant4 plant5 plant6 plant7 plant8
loc1      6      4      0      3      0      0      0      0
loc2      4      5      5      3      4      2      0      0
loc3      3      4      7      4      5      2      1      1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>reslake<span class="ot">=</span><span class="fu">dudi.coa</span>(lakelike,<span class="at">scannf=</span><span class="cn">FALSE</span>,<span class="at">nf=</span><span class="dv">2</span>)</span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a><span class="fu">round</span>(reslake<span class="sc">$</span>eig[<span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]<span class="sc">/</span><span class="fu">sum</span>(reslake<span class="sc">$</span>eig),<span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.56 0.25 0.09 0.03 0.03 0.02 0.01 0.00</code></pre>
</div>
</div>
<p>We plot both the row-location points (<a href="#fig-LakeCAr" class="quarto-xref">Figure&nbsp;<span>9.25</span></a> (a)) and the biplot of both location and plant species in the lower part of <a href="#fig-LakeCAr" class="quarto-xref">Figure&nbsp;<span>9.25</span></a> (b); this plot was made with:</p>
<div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_ca_row</span>(reslake,<span class="at">repel=</span><span class="cn">TRUE</span>)<span class="sc">+</span><span class="fu">ggtitle</span>(<span class="st">""</span>)<span class="sc">+</span><span class="fu">ylim</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.55</span>,<span class="fl">1.7</span>))</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_ca_biplot</span>(reslake,<span class="at">repel=</span><span class="cn">TRUE</span>)<span class="sc">+</span><span class="fu">ggtitle</span>(<span class="st">""</span>)<span class="sc">+</span><span class="fu">ylim</span>(<span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.55</span>,<span class="fl">1.7</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-LakeCAr" class="cell quarto-layout-panel" data-hash="09-chap_cache/html/fig-LakeCAr_e8a0877935b1048b1d5cf7a87c33b96d">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-LakeCAr-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-LakeCAr-1.png" class="img-fluid figure-img" data-ref-parent="fig-LakeCAr" width="320"></p>
<figcaption class="figure-caption">(a)</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-LakeCAr-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-LakeCAr-2.png" class="img-fluid figure-img" data-ref-parent="fig-LakeCAr" width="320"></p>
<figcaption class="figure-caption">(b)</figcaption>
</figure>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;9.25: The locations near the lake are ordered along an arch as shown in (a). In the biplot (b), we can see which plants are most frequent at which locations by looking at the red triangles closest to the blue points.</figcaption><p></p>
</figure>
</div>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-matrixpattern" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.16 </strong></span>Looking back at the raw matrix <code>lakes</code> as it appears, do you see a pattern in its entries?<br>
What would happen if the plants had been ordered by actual taxa names for instance?</p>
</div>
</div>
</div>
<section id="sec-hetero-celldynamics" class="level3 page-columns page-full" data-number="9.5.1">
<h3 data-number="9.5.1" class="anchored" data-anchor-id="sec-hetero-celldynamics"><span class="header-section-number">9.5.1</span> Dynamics of cell development</h3>
<p>We will now analyse a more interesting data set that was published by <span class="citation" data-cites="Moignard:2015">Moignard et al. (<a href="16-chap.html#ref-Moignard:2015" role="doc-biblioref">2015</a>)</span>. This paper describes the dynamics of blood cell development. The data are single cell gene expression measurements of 3,934 cells with blood and endothelial potential from five populations from between embryonic days E7.0 and E8.25.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-CellTree_682b5163ff0934dcff9c1b834e5cbaec">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-CellTree" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="imgs/CellTree.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;9.26: The four cell populations studied here are representative of three sequential states (PS,NP,HF) and two possible final branches (4SG and 4SFG<span class="math inline">\(^{-}\)</span>).</figcaption>
</figure>
</div>
</div></div></div>
<p>Remember from <a href="04-chap.html" class="quarto-xref"><span>Chapter&nbsp;4</span></a> that several different distances are available for comparing our cells. Here, we start by computing both an <span class="math inline">\(L_2\)</span> distance and the <span class="math inline">\(\ell_1\)</span> distance between the 3,934 cells.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>Moignard <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">"../data/Moignard.rds"</span>)</span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a>cellt <span class="ot">=</span> <span class="fu">rowData</span>(Moignard)<span class="sc">$</span>celltypes</span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>colsn <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"purple"</span>, <span class="st">"orange"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>)</span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>blom <span class="ot">=</span> <span class="fu">assay</span>(Moignard)</span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>dist2n.euclid <span class="ot">=</span> <span class="fu">dist</span>(blom)</span>
<span id="cb86-6"><a href="#cb86-6" aria-hidden="true" tabindex="-1"></a>dist1n.l1     <span class="ot">=</span> <span class="fu">dist</span>(blom, <span class="st">"manhattan"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The classical multidimensional scaling on these two distances matrices can be carried out using:</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/CMDS_7bbe628170dbd3d1ef06a9dc259eb49b">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>ce1Mds <span class="ot">=</span> <span class="fu">cmdscale</span>(dist1n.l1,     <span class="at">k =</span> <span class="dv">20</span>, <span class="at">eig =</span> <span class="cn">TRUE</span>)</span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>ce2Mds <span class="ot">=</span> <span class="fu">cmdscale</span>(dist2n.euclid, <span class="at">k =</span> <span class="dv">20</span>, <span class="at">eig =</span> <span class="cn">TRUE</span>)</span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>perc1  <span class="ot">=</span> <span class="fu">round</span>(<span class="dv">100</span><span class="sc">*</span><span class="fu">sum</span>(ce1Mds<span class="sc">$</span>eig[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>])<span class="sc">/</span><span class="fu">sum</span>(ce1Mds<span class="sc">$</span>eig))</span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a>perc2  <span class="ot">=</span> <span class="fu">round</span>(<span class="dv">100</span><span class="sc">*</span><span class="fu">sum</span>(ce2Mds<span class="sc">$</span>eig[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>])<span class="sc">/</span><span class="fu">sum</span>(ce2Mds<span class="sc">$</span>eig))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We look at the underlying dimension and see in <a href="#fig-CMDSplotscree" class="quarto-xref">Figure&nbsp;<span>9.27</span></a> that two dimensions can provide a substantial fraction of the variance.</p>
<div class="page-columns page-full">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotscree</span>(ce1Mds, <span class="at">m =</span> <span class="dv">4</span>)</span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotscree</span>(ce2Mds, <span class="at">m =</span> <span class="dv">4</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div id="fig-CMDSplotscree" class="cell quarto-layout-panel" data-hash="09-chap_cache/html/fig-CMDSplotscree_4158a801d7639720b5e6af8f3321a13f">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 100.0%;justify-content: center;">
<div id="fig-CMDSplotscree-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-CMDSplotscree-1.png" class="img-fluid figure-img" data-ref-parent="fig-CMDSplotscree" width="200"></p>
<figcaption class="figure-caption">(a)</figcaption>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 100.0%;justify-content: center;">
<div id="fig-CMDSplotscree-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-CMDSplotscree-2.png" class="img-fluid figure-img" data-ref-parent="fig-CMDSplotscree" width="200"></p>
<figcaption class="figure-caption">(b)</figcaption>
</figure>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;9.27: Screeplots from MDS on <span class="math inline">\(\ell_1\)</span> (a) and <span class="math inline">\(L_2\)</span> (b) distances. We see that the eigenvalues are extremely similar and both point to a <span class="math inline">\(2\)</span> dimensional phenomenon.</figcaption><p></p>
</figure>
</div></div></div>
<p>The first 2 coordinates account for 78 % of the variability when the <span class="math inline">\(\ell_1\)</span> distance is used between cells, and 57% when the <span class="math inline">\(L^2\)</span> distance is used. We see in <a href="#fig-CMDSplotL2-1" class="quarto-xref">Figure&nbsp;<span>9.28 (a)</span></a> the first plane for the MDS on the <span class="math inline">\(\ell_1\)</span> distances between cells:</p>
<div class="page-columns page-full">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>c1mds <span class="ot">=</span> ce1Mds<span class="sc">$</span>points[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="sc">|&gt;</span></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>        <span class="st">`</span><span class="at">colnames&lt;-</span><span class="st">`</span>(<span class="fu">paste0</span>(<span class="st">"L1_PCo"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)) <span class="sc">|&gt;</span></span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">as_tibble</span>()</span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(c1mds, <span class="fu">aes</span>(<span class="at">x =</span> L1_PCo1, <span class="at">y =</span> L1_PCo2, <span class="at">color =</span> cellt)) <span class="sc">+</span></span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> cellt), <span class="at">alpha =</span> <span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> colsn) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>)</span>
<span id="cb89-7"><a href="#cb89-7" aria-hidden="true" tabindex="-1"></a>c2mds <span class="ot">=</span> ce2Mds<span class="sc">$</span>points[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="sc">|&gt;</span></span>
<span id="cb89-8"><a href="#cb89-8" aria-hidden="true" tabindex="-1"></a>        <span class="st">`</span><span class="at">colnames&lt;-</span><span class="st">`</span>(<span class="fu">paste0</span>(<span class="st">"L2_PCo"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)) <span class="sc">|&gt;</span></span>
<span id="cb89-9"><a href="#cb89-9" aria-hidden="true" tabindex="-1"></a>        <span class="fu">as_tibble</span>()</span>
<span id="cb89-10"><a href="#cb89-10" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(c2mds, <span class="fu">aes</span>(<span class="at">x =</span> L2_PCo1, <span class="at">y =</span> L2_PCo2, <span class="at">color =</span> cellt)) <span class="sc">+</span></span>
<span id="cb89-11"><a href="#cb89-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> cellt), <span class="at">alpha =</span> <span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb89-12"><a href="#cb89-12" aria-hidden="true" tabindex="-1"></a>   <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> colsn) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-CMDSplotL2" class="cell quarto-layout-panel page-columns page-full" data-hash="09-chap_cache/html/fig-CMDSplotL2_0cce05bfa4908f6effe4da110b4117b4">
<figure class="figure page-columns page-full">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-CMDSplotL2-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-CMDSplotL2-1.png" class="img-fluid figure-img" data-ref-parent="fig-CMDSplotL2" width="400"></p>
<figcaption class="figure-caption">(a)</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-CMDSplotL2-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-CMDSplotL2-2.png" class="img-fluid figure-img" data-ref-parent="fig-CMDSplotL2" width="400"></p>
<figcaption class="figure-caption">(b)</figcaption>
</figure>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption margin-caption">Figure&nbsp;9.28: Moignard cell data colored according to the cell types (blue: PS, green: NP, yellow: HF, red: 4SG, purple: 4SFG<span class="math inline">\(^-\)</span>) in the two dimensional MDS plots created. In (a) using <span class="math inline">\(\ell_1\)</span> distances and in (b) using the L2 distances.</figcaption><p></p>
</figure>
</div>
</div>
<!-- TODO: this isn't showing in the book online? 

::: {.cell .column-margin layout-align="center" hash='09-chap_cache/html/colorlegend_8268e52e061a1733465775e2ce2bb09c'}
::: {.cell-output-display}
![](09-chap_files/figure-html/colorlegend-1.png){fig-align='center' width=160}
:::
:::

-->
<p><a href="#fig-CMDSplotL2-2" class="quarto-xref">Figure&nbsp;<span>9.28 (b)</span></a> is created in the same way and shows the two-dimensional projection created by using MDS on the L2 distances.</p>
<p><a href="#fig-CMDSplotL2" class="quarto-xref">Figure&nbsp;<span>9.28</span></a> shows that both distances (L1 and L2) give the same first plane for the MDS with very similar representations of the underlying gradient followed by the cells.</p>
<p>We can see from <a href="#fig-CMDSplotL2" class="quarto-xref">Figure&nbsp;<span>9.28</span></a> that the cells are not distributed uniformly in the lower dimensions we have been looking at, we see a definite organization of the points. All the cells of type 4SG represented in red form an elongated cluster who are much less mixed with the other cell types.</p>
</section>
<section id="local-nonlinear-methods" class="level3 page-columns page-full" data-number="9.5.2">
<h3 data-number="9.5.2" class="anchored" data-anchor-id="local-nonlinear-methods"><span class="header-section-number">9.5.2</span> Local, nonlinear methods</h3>
<p>Multidimensional scaling and non metric multidimensional scaling aims to represent <strong>all</strong> distances as precisely as possible and the large distances between far away points skew the representations. It can be beneficial when looking for gradients or low dimensional manifolds to restrict ourselves to approximations of points that are close together. This calls for methods that try to represent local (small) distances well and do not try to approximate distances between faraway points with too much accuracy.</p>
<p>There has been substantial progress in such methods in recent years. The use of <strong>kernels</strong> computed using the calculated interpoint distances allows us to decrease the importance of points that are far apart. A radial basis kernel is of the form</p>
<p><span class="math display">\[
1-\exp\left(-\frac{d(x,y)^2}{\sigma^2}\right),
\quad\mbox{where } \sigma^2 \mbox{ is fixed.}
\]</span></p>
<p>It has the effect of heavily discounting large distances. This can be very useful as the precision of interpoint distances is often better at smaller ranges; several examples of such methods are covered in <a href="#exr-hetero-KernelMethods" class="quarto-xref">Exercise&nbsp;<span>9.6</span></a> at the end of this chapter.</p>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-exp" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.17 </strong></span>Why do we take the difference between the 1 and the exponential?<br>
What happens when the distance between <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> is very big?</p>
</div>
</div>
</div>
<section id="t-sne." class="level4 unnumbered page-columns page-full">
<h4 class="unnumbered anchored" data-anchor-id="t-sne.">t-SNE.</h4>
<p>This widely used method adds flexibility to the kernel defined above and allows the <span class="math inline">\(\sigma^2\)</span> parameter to vary locally (there is a normalization step so that it averages to one). The t-SNE method starts out from the positions of the points in the high dimensional space and derives a probability distribution on the set of pairs of points, such that the probabilities are proportional to the points’ proximities or similarities. It then uses this distribution to construct a representation of the dataset in low dimensions. The method is not robust and has the property of separating clusters of points artificially; however, this property can also help clarify a complex situation. One can think of it as a method akin to graph (or network) layout algorithms. They stretch the data to clarify relations between the very close (in the network: connected) points, but the distances between more distal (in the network: unconnected) points cannot be interpreted as being on the same scales in different regions of the plot. In particular, these distances will depend on the local point densities. Here is an example of the output of t-SNE on the cell data:</p>
<div class="page-columns page-full">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"Rtsne"</span>)</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>restsne <span class="ot">=</span> <span class="fu">Rtsne</span>(blom, <span class="at">dims =</span> <span class="dv">2</span>, <span class="at">perplexity =</span> <span class="dv">30</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>,</span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">max_iter =</span> <span class="dv">900</span>)</span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>dftsne <span class="ot">=</span> restsne<span class="sc">$</span>Y[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="sc">|&gt;</span></span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>         <span class="st">`</span><span class="at">colnames&lt;-</span><span class="st">`</span>(<span class="fu">paste0</span>(<span class="st">"axis"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)) <span class="sc">|&gt;</span></span>
<span id="cb90-6"><a href="#cb90-6" aria-hidden="true" tabindex="-1"></a>         <span class="fu">as_tibble</span>()</span>
<span id="cb90-7"><a href="#cb90-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(dftsne,<span class="fu">aes</span>(<span class="at">x =</span> axis1, <span class="at">y =</span> axis2, <span class="at">color =</span> cellt)) <span class="sc">+</span></span>
<span id="cb90-8"><a href="#cb90-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> cellt), <span class="at">alpha =</span> <span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb90-9"><a href="#cb90-9" aria-hidden="true" tabindex="-1"></a>   <span class="fu">scale_color_manual</span>(<span class="at">values =</span> colsn) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>)</span>
<span id="cb90-10"><a href="#cb90-10" aria-hidden="true" tabindex="-1"></a>restsne3 <span class="ot">=</span> <span class="fu">Rtsne</span>(blom, <span class="at">dims =</span> <span class="dv">3</span>, <span class="at">perplexity =</span> <span class="dv">30</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>,</span>
<span id="cb90-11"><a href="#cb90-11" aria-hidden="true" tabindex="-1"></a>                 <span class="at">max_iter =</span> <span class="dv">900</span>)</span>
<span id="cb90-12"><a href="#cb90-12" aria-hidden="true" tabindex="-1"></a>dftsne3 <span class="ot">=</span> restsne3<span class="sc">$</span>Y[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>] <span class="sc">|&gt;</span></span>
<span id="cb90-13"><a href="#cb90-13" aria-hidden="true" tabindex="-1"></a>          <span class="st">`</span><span class="at">colnames&lt;-</span><span class="st">`</span>(<span class="fu">paste0</span>(<span class="st">"axis"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>)) <span class="sc">|&gt;</span> </span>
<span id="cb90-14"><a href="#cb90-14" aria-hidden="true" tabindex="-1"></a>          <span class="fu">as_tibble</span>()</span>
<span id="cb90-15"><a href="#cb90-15" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(dftsne3,<span class="fu">aes</span>(<span class="at">x =</span> axis3, <span class="at">y =</span> axis2, <span class="at">group =</span> cellt)) <span class="sc">+</span></span>
<span id="cb90-16"><a href="#cb90-16" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">color =</span> cellt), <span class="at">alpha =</span> <span class="fl">0.6</span>) <span class="sc">+</span></span>
<span id="cb90-17"><a href="#cb90-17" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scale_colour_manual</span>(<span class="at">values =</span> colsn) <span class="sc">+</span> <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-tsnecells" class="cell quarto-layout-panel page-columns page-full" data-hash="09-chap_cache/html/fig-tsnecells_0ed0eb4061abe537b47d40ab17b2e66c">
<figure class="figure page-columns page-full">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-tsnecells-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-tsnecells-1.jpeg" class="img-fluid figure-img" data-ref-parent="fig-tsnecells" width="400"></p>
<figcaption class="figure-caption">(a)</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-tsnecells-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-tsnecells-2.jpeg" class="img-fluid figure-img" data-ref-parent="fig-tsnecells" width="400"></p>
<figcaption class="figure-caption">(b)</figcaption>
</figure>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption margin-caption">Figure&nbsp;9.29: The four cell populations studied here are representative of three sequential states (PS,NP,HF) and two possible final branches (4SG and 4SFG<span class="math inline">\(^{-}\)</span>). The plot on the left was obtained by choosing 2 dimensions for t-sne at a perplexity of 30. The lower plot has obtained by choosing 3 dimensions, we can see that this third t-SNE axis represented here as the horizontal axis.</figcaption><p></p>
</figure>
</div>
</div>
<p>In this case in order to see the subtle differences between MDS and t-SNE, it is really necessary to use 3d plotting.</p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<p>Use the <strong><a href="https://cran.r-project.org/web/packages/rgl/">rgl</a></strong> package to look at the three t-SNE dimensions and add the correct cell type colors to the display.</p>
</div>
</div>
<p>Two of these 3d snapshots are shown in <a href="#fig-tsne3d" class="quarto-xref">Figure&nbsp;<span>9.30</span></a>, we see a much stronger grouping of the purple points than in the MDS plots.</p>
<p><strong>Note:</strong> A site worth visiting in order to appreciate more about the sensitivity of the t-SNE method to the complexity and <span class="math inline">\(\sigma\)</span> parameters can be found at <a href="http://distill.pub/2016/misread-tsne" class="uri">http://distill.pub/2016/misread-tsne</a>.</p>
<div id="fig-tsne3d" class="cell quarto-layout-panel page-columns page-full" data-hash="09-chap_cache/html/fig-tsne3d_faf4fa3bf0938914c7fc2ccbf1347aa9">
<figure class="figure page-columns page-full">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-tsne3d-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="imgs/tsnemoignard3scrop.png" class="img-fluid figure-img" data-ref-parent="fig-tsne3d"></p>
<figcaption class="figure-caption">(a)</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-tsne3d-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="imgs/tsnemoignard3crop.png" class="img-fluid figure-img" data-ref-parent="fig-tsne3d"></p>
<figcaption class="figure-caption">(b)</figcaption>
</figure>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption margin-caption">Figure&nbsp;9.30: Moignard cell data colored according to the cell types (blue: PS, green: NP, yellow: HF, red: 4SG, purple: 4SFG<span class="math inline">\(^-\)</span>) in the three-dimensional t-SNE layouts. We can see that the purple cells (4SFG<span class="math inline">\(^-\)</span>) segregate at the outer shell on the top of the point cloud.</figcaption><p></p>
</figure>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-ukrainetsne" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.18 </strong></span>Visualize a two-dimensional t-SNE embedding of the Ukraine distances from <a href="#sec-hetero-multidimensionalscalingandordination" class="quarto-xref"><span>Section&nbsp;9.2</span></a>.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-31-contents" aria-controls="callout-31" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-31" class="callout-31-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/fig-ukrainetsne_26ad882ce0c800a1b098b11f13cd49ea">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>ukraine_tsne <span class="ot">=</span> <span class="fu">Rtsne</span>(ukraine_dists, <span class="at">is_distance =</span> <span class="cn">TRUE</span>, <span class="at">perplexity =</span> <span class="dv">8</span>)</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>ukraine_tsne_df <span class="ot">=</span> <span class="fu">tibble</span>(</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">PCo1 =</span> ukraine_tsne<span class="sc">$</span>Y[, <span class="dv">1</span>],</span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">PCo2 =</span> ukraine_tsne<span class="sc">$</span>Y[, <span class="dv">2</span>],</span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">labs =</span> <span class="fu">attr</span>(ukraine_dists, <span class="st">"Labels"</span>)</span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(ukraine_tsne_df, <span class="fu">aes</span>(<span class="at">x =</span> PCo1, <span class="at">y =</span> PCo2, <span class="at">label =</span> labs)) <span class="sc">+</span></span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span> <span class="fu">geom_text_repel</span>(<span class="at">col =</span> <span class="st">"#0057b7"</span>) <span class="sc">+</span> <span class="fu">coord_fixed</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div id="fig-ukrainetsne" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-ukrainetsne-1.png" class="img-fluid figure-img" style="width:70.0%"></p>
<figcaption class="figure-caption">Figure&nbsp;9.31: t-SNE map based of Ukraine.</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>There are several other nonlinear methods for estimating nonlinear trajectories followed by points in the relevant state spaces. Here are a few examples.</p>
<p><strong><a href="https://cran.r-project.org/web/packages/RDRToolbox/">RDRToolbox</a></strong> Local linear embedding (<strong>LLE</strong>) and <strong>isomap</strong></p>
<p><strong><a href="https://cran.r-project.org/web/packages/diffusionMap/">diffusionMap</a></strong> This package models connections between points as a Markovian kernel.</p>
<p><strong><a href="https://cran.r-project.org/web/packages/kernlab/">kernlab</a></strong> Kernel methods</p>
<p><strong><a href="https://cran.r-project.org/web/packages/LPCM-package/">LPCM-package</a></strong> Local principal curves</p>
</section>
</section>
</section>
<section id="multitable-techniques" class="level2 page-columns page-full" data-number="9.6">
<h2 data-number="9.6" class="anchored" data-anchor-id="multitable-techniques"><span class="header-section-number">9.6</span> Multitable techniques</h2>
<p>Current studies often attempt to quantify variation in the microbial, genomic, and metabolic measurements across different experimental conditions. As a result, it is common to perform multiple assays on the same biological samples and ask what features – microbes, genes, or metabolites, for example – are associated with different sample conditions. There are many ways to approach these questions. Which to apply depends on the study’s focus.</p>
<section id="co-variation-inertia-co-inertia-and-the-rv-coefficient" class="level3 page-columns page-full" data-number="9.6.1">
<h3 data-number="9.6.1" class="anchored" data-anchor-id="co-variation-inertia-co-inertia-and-the-rv-coefficient"><span class="header-section-number">9.6.1</span> Co-variation, inertia, co-inertia and the RV coefficient</h3>
<div class="page-columns page-full"><p>As in physics, we define inertia as a sum of distances with ‘weighted’ points. This enables us to compute the inertia of counts in a contingency table as the weighted sum of the squares of distances between observed and expected frequencies (as in the chisquare statistic). </p><div class="no-row-height column-margin column-container"><span class="">Another generalization of variance-inertia is the useful Phylogenetic diversity index. (computing the sum of distances between a subset of taxa through the tree). Other useful generalizations include using variability of points on a graph taken from standard spatial statistics.</span></div></div>
<p>If we want to study two standardized variables measured at the same 10 locations together, we use their <strong>covariance</strong>. If <span class="math inline">\(x\)</span> represents the standardized PH, and and <span class="math inline">\(y\)</span> the standardized humidity, we measure their covariation using the mean</p>
<p><span class="anchored-eq" data-anchor-id="eq-npkuhvloxdzr" id="eq-npkuhvloxdzr"><span class="math display">\[
\text{cov}(x,y) = \text{mean}(x1*y1 + x2*y2 + x3*y3 + \cdots + x10*y10).
\tag{9.2}\]</span></span></p>
<p>If <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span> co-vary in the same direction, this will be big. We saw how useful the correlation coefficient we defined in <a href="08-chap.html" class="quarto-xref"><span>Chapter&nbsp;8</span></a> was to our multivariate analyses. Multitable generalizations will be just as useful.</p>
</section>
<section id="mantel-coefficient-and-a-test-of-distance-correlation" class="level3 page-columns page-full" data-number="9.6.2">
<h3 data-number="9.6.2" class="anchored" data-anchor-id="mantel-coefficient-and-a-test-of-distance-correlation"><span class="header-section-number">9.6.2</span> Mantel coefficient and a test of distance correlation</h3>
<p>The Mantel coefficient, one of the earliest version of matrix association, developed and used by Henry Daniels, FN David and co-authors <span class="citation" data-cites="Josse:2016">(<a href="16-chap.html#ref-Josse:2016" role="doc-biblioref">Josse and Holmes 2016</a>)</span> is very popular, especially in ecology.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/unnamed-chunk-122_d47881ba770f8c8e38e73e042d75903c">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/devil.png" class="img-fluid figure-img" width="123"></p>
<figcaption class="figure-caption">There are some precautions to be taken when using the Mantel coefficient, see a critical review in <span class="citation" data-cites="Guillot:2013">Guillot and Rousset (<a href="16-chap.html#ref-Guillot:2013" role="doc-biblioref">2013</a>)</span>.</figcaption>
</figure>
</div>
</div></div></div>
<p>Given two dissimilarity matrices <span class="math inline">\({\mathbf D^X}\)</span> and <span class="math inline">\({\mathbf D^Y}\)</span>, make these matrices into vectors the way the R <code>dist</code> function does, and compute their correlation. This is defined mathematically as:</p>
<!-- TODO: this is too wide -->
<p><span class="math display">\[
\begin{aligned}
\mbox{r}_{\rm m}(\bf X,\bf Y)= \frac{\sum_{i=1}^n \sum_{j=1,j \neq i}^n (d_{ij}^**X**-
\bar{d}^**X**)(d_{ij}^**Y**-\bar d^**Y**)}{\sqrt{\sum_{i,j,j \neq i}(d_{ij}^**X**-
\bar{d}^**X**)^2 \sum_{i,j,j \neq i}(d_{ij}^**Y**-\bar{d}^**Y**)^2}},
\end{aligned}
\]</span></p>
<p>with <span class="math inline">\(\bar{d}^**X**\)</span> (resp <span class="math inline">\(\bar{d}^**Y**\)</span>) the mean of the lower diagonal terms of the dissimilarity matrix associated to <span class="math inline">\(\bf X\)</span> (resp. to <span class="math inline">\(\bf Y\)</span>). This formulation shows us that it is a measure of linear correlation between distances. It has been widely used for testing two sets of distances, for instance one distance <span class="math inline">\({\mathbf D^X}\)</span> could be computed using the soil chemistry at 17 different locations. The other distance <span class="math inline">\({\mathbf D^Y}\)</span> could record plant abundance dissimilarities using a Jaccard index between the 17 same locations.</p>
<p>The correlation’s significance is often assessed via a simple permutation test, see <span class="citation" data-cites="Josse:2016">Josse and Holmes (<a href="16-chap.html#ref-Josse:2016" role="doc-biblioref">2016</a>)</span> for a review with its historical background and modern incarnations. The coefficient and associated tests are implemented in several R packages such as <strong><a href="https://cran.r-project.org/web/packages/ade4/">ade4</a></strong> <span class="citation" data-cites="ade4">(<a href="16-chap.html#ref-ade4" role="doc-biblioref">Chessel, Dufour, and Thioulouse 2004</a>)</span>, <strong><a href="https://cran.r-project.org/web/packages/vegan/">vegan</a></strong> and <strong><a href="https://cran.r-project.org/web/packages/ecodist/">ecodist</a></strong> <span class="citation" data-cites="Goslee:2007">(<a href="16-chap.html#ref-Goslee:2007" role="doc-biblioref">Goslee, Urban, et al. 2007</a>)</span>.</p>
<p>RV coefficient The global measure of similarity of two data tables as opposed to two vectors can be done by a generalization of covariance provided by an inner product between tables that gives the RV coefficient, a number between 0 and 1, like a correlation coefficient, but for tables.</p>
<p><span class="math display">\[
RV(A,B)=\frac{Tr(A'BAB')}{\sqrt{Tr(A'A)}\sqrt{Tr(B'B)}}
\]</span></p>
<p>There are several other measures of matrix correlation available in the package <strong><a href="https://cran.r-project.org/web/packages/MatrixCorrelation/">MatrixCorrelation</a></strong>.</p>
<p>If we do ascertain a link between two matrices, we then need to find a way to understand that link. One such method is explained in the next section.</p>
</section>
<section id="canonical-correlation-analysis-cca" class="level3" data-number="9.6.3">
<h3 data-number="9.6.3" class="anchored" data-anchor-id="canonical-correlation-analysis-cca"><span class="header-section-number">9.6.3</span> Canonical correlation analysis (CCA)</h3>
<p>CCA is a method similar to PCA as it was developed by Hotelling in the 1930s to search for associations between two sets of continuous variables <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span>. Its goal is to find a linear projection of the first set of variables that maximally correlates with a linear projection of the second set of variables.</p>
<p>Finding correlated functions (covariates) of the two views of the same phenomenon by discarding the representation-specific details (noise) is expected to reveal the underlying hidden yet influential factors responsible for the correlation.</p>
<section id="canonical-correlation-algorithm." class="level4 unnumbered">
<h4 class="unnumbered anchored" data-anchor-id="canonical-correlation-algorithm.">Canonical Correlation Algorithm.</h4>
<p>Let us consider two matrices <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span> of order <span class="math inline">\(n &gt; p\)</span> and <span class="math inline">\(n &gt; q\)</span> respectively. The columns of <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span> correspond to variables and the rows correspond to the same <span class="math inline">\(n\)</span> experimental units. The jth column of the matrix <span class="math inline">\(X\)</span> is denoted by <span class="math inline">\(X_j\)</span>, likewise the kth column of <span class="math inline">\(Y\)</span> is denoted by <span class="math inline">\(Y_k\)</span>. Without loss of generality it will be assumed that the columns of <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span> are standardized (mean 0 and variance 1).</p>
<p>We denote by <span class="math inline">\(S_{XX}\)</span> and <span class="math inline">\(S_{YY}\)</span> the sample covariance matrices for variable sets <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span> respectively, and by <span class="math inline">\(S_{XY} = S_{YX}^t\)</span> the sample cross-covariance matrix between <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span> .</p>
<p>Classical CCA assumes first <span class="math inline">\(p \leq n\)</span> and <span class="math inline">\(q \leq n\)</span>, then that matrices <span class="math inline">\(X\)</span> and <span class="math inline">\(Y\)</span> are of full column rank p and q respectively. In the following, the principle of CCA is presented as a problem solved through an iterative algorithm. The first stage of CCA consists in finding two vectors <span class="math inline">\(a =(a_1,...,a_p)^t\)</span> and <span class="math inline">\(b =(b_1,...,b_q)^t\)</span> that maximize the correlation between the linear combinations <span class="math inline">\(U\)</span> and <span class="math inline">\(V\)</span> defined as</p>
<p><span class="math display">\[
\begin{aligned}
U=Xa&amp;=&amp;a_1 X_1 +a_2 X_2 +\cdots a_p X_p\\
V=Yb&amp;=&amp;b_1 Y_1 +b_2 Y_2 +\cdots a_q Y_q
\end{aligned}
\]</span></p>
<p>and assuming that vectors <span class="math inline">\(a\)</span> and $ b$ are normalized so that <span class="math inline">\(var(U) = var(V) = 1\)</span> In other words, the problem consists in finding <span class="math inline">\(a\)</span> and <span class="math inline">\(b\)</span> such that</p>
<p><span class="anchored-eq" data-anchor-id="eq-hwdsxrvnmzkg" id="eq-hwdsxrvnmzkg"><span class="math display">\[
\rho_1 = \text{cor}(U, V) = \max_{a,b} \text{cor} (Xa, Yb)\quad
\text{subject to}\quad \text{var}(Xa)=\text{var}(Yb) = 1.
\tag{9.3}\]</span></span></p>
<p>The resulting variables <span class="math inline">\(U\)</span> and <span class="math inline">\(V\)</span> are called the first canonical variates and <span class="math inline">\(\rho_1\)</span> is referred to as the first canonical correlation.</p>
<p><strong>Note:</strong> Higher order canonical variates and canonical correlations can be found as a stepwise problem. For <span class="math inline">\(s = 1,...,p\)</span>, we can successively find positive correlations <span class="math inline">\(\rho_1 \geq \rho_2 \geq ... \geq \rho_p\)</span> with corresponding vectors <span class="math inline">\((a^1, b^1), ..., (a^p, b^p)\)</span>, by maximizing</p>
<p><span class="anchored-eq" data-anchor-id="eq-kgarhpstjydq" id="eq-kgarhpstjydq"><span class="math display">\[
\rho_s = \text{cor}(U^s,V^s) = \max_{a^s,b^s} \text{cor} (Xa^s,Yb^s)\quad
\text{subject to}\quad \text{var}(Xa^s) = \text{var}(Yb^s)=1
\tag{9.4}\]</span></span></p>
<p>under the additional restrictions</p>
<p><span class="anchored-eq" data-anchor-id="eq-fvtikozmsagl" id="eq-fvtikozmsagl"><span class="math display">\[
\text{cor}(U^s,U^t) = \text{cor}(V^s, V^t)=0 \quad\text{for}\quad
1 \leq t &lt; s \leq p.
\tag{9.5}\]</span></span></p>
<p>We can think of CCA as a generalization of PCA where the variance we maximize is the `covariance’ between the two matrices (see <span class="citation" data-cites="frenchway">Holmes (<a href="16-chap.html#ref-frenchway" role="doc-biblioref">2006</a>)</span> for more details).</p>
</section>
</section>
<section id="sparse-canonical-correlation-analysis-scca" class="level3 page-columns page-full" data-number="9.6.4">
<h3 data-number="9.6.4" class="anchored" data-anchor-id="sparse-canonical-correlation-analysis-scca"><span class="header-section-number">9.6.4</span> Sparse canonical correlation analysis (sCCA)</h3>
<p>When the number of variables in each table is very large finding two very correlated vectors can be too easy and unstable: we have too many degrees of freedom.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/unnamed-chunk-123_ea8e5c0c299bd52ae29facc8a96d2531">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/devil.png" class="img-fluid figure-img" width="123"></p>
<figcaption class="figure-caption">We will see many examples of regularization and danger of overfitting in <a href="12-chap.html" class="quarto-xref"><span>Chapter&nbsp;12</span></a>.</figcaption>
</figure>
</div>
</div></div></div>
<p>Then it is beneficial to add a penalty maintains the number of non-zero coefficients to a minimum. This approach is called sparse canonical correlation analysis (sparse CCA or sCCA), a method well-suited to both exploratory comparisons between samples and the identification of features with interesting <strong>co</strong>variation. We will use an implementation from the <strong><a href="https://cran.r-project.org/web/packages/PMA/">PMA</a></strong> package.</p>
<p>Here we study a dataset collected by <span class="citation" data-cites="Kashyap:2013">Kashyap et al. (<a href="16-chap.html#ref-Kashyap:2013" role="doc-biblioref">2013</a>)</span> with two tables. One is a contingency table of bacterial abundances and another an abundance table of metabolites. There are 12 samples, so <span class="math inline">\(n = 12\)</span>. The metabolite table has measurements on <span class="math inline">\(p = 637\)</span> feature and the bacterial abundances had a total of $ q = 20,609$ OTUs, which we will filter down to around 200. We start by loading the data.</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/multitable-setup_ae21aee640178ed45a24fe31e5e6ee6d">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"genefilter"</span>)</span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../data/microbe.rda"</span>)</span>
<span id="cb92-3"><a href="#cb92-3" aria-hidden="true" tabindex="-1"></a>metab <span class="ot">=</span> <span class="fu">read.csv</span>(<span class="st">"../data/metabolites.csv"</span>, <span class="at">row.names =</span> <span class="dv">1</span>) <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We first filter down to bacteria and metabolites of interest, removing (“by hand”) those that are zero across many samples and giving an upper threshold of 50 to the large values. We transform the data to weaken the heavy tails.</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/multitable-filtering_6d9fdbed71d1087f4b7f80d3b2cf4000">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"phyloseq"</span>)</span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>metab   <span class="ot">=</span> metab[<span class="fu">rowSums</span>(metab <span class="sc">==</span> <span class="dv">0</span>) <span class="sc">&lt;=</span> <span class="dv">3</span>, ]</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>microbe <span class="ot">=</span> <span class="fu">prune_taxa</span>(<span class="fu">taxa_sums</span>(microbe) <span class="sc">&gt;</span> <span class="dv">4</span>, microbe)</span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>microbe <span class="ot">=</span> <span class="fu">filter_taxa</span>(microbe, <span class="fu">filterfun</span>(<span class="fu">kOverA</span>(<span class="dv">3</span>, <span class="dv">2</span>)), <span class="cn">TRUE</span>)</span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>metab   <span class="ot">=</span> <span class="fu">log</span>(<span class="dv">1</span> <span class="sc">+</span> metab, <span class="at">base =</span> <span class="dv">10</span>)</span>
<span id="cb93-6"><a href="#cb93-6" aria-hidden="true" tabindex="-1"></a>X       <span class="ot">=</span> <span class="fu">log</span>(<span class="dv">1</span> <span class="sc">+</span> <span class="fu">as.matrix</span>(<span class="fu">otu_table</span>(microbe)), <span class="at">base =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>A second step in our preliminary analysis is to look if there is any association between the two matrices using the <code>RV.test</code> from the <strong><a href="https://cran.r-project.org/web/packages/ade4/">ade4</a></strong> package:</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/RVtest_b22f1b93aa7f7ad4ffefe5c11e07c06c">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(metab) <span class="ot">=</span> <span class="fu">colnames</span>(X)</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>pca1 <span class="ot">=</span> <span class="fu">dudi.pca</span>(<span class="fu">t</span>(metab), <span class="at">scal =</span> <span class="cn">TRUE</span>, <span class="at">scann =</span> <span class="cn">FALSE</span>)</span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>pca2 <span class="ot">=</span> <span class="fu">dudi.pca</span>(<span class="fu">t</span>(X), <span class="at">scal =</span> <span class="cn">TRUE</span>, <span class="at">scann =</span> <span class="cn">FALSE</span>)</span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a>rv1 <span class="ot">=</span> <span class="fu">RV.rtest</span>(pca1<span class="sc">$</span>tab, pca2<span class="sc">$</span>tab, <span class="dv">999</span>)</span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a>rv1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Monte-Carlo test
Call: RV.rtest(df1 = pca1$tab, df2 = pca2$tab, nrepet = 999)

Observation: 0.8400429 

Based on 999 replicates
Simulated p-value: 0.003 
Alternative hypothesis: greater 

    Std.Obs Expectation    Variance 
5.755704546 0.315262467 0.008313011 </code></pre>
</div>
</div>
<p>We can now apply sparse CCA. This method compares sets of features across high-dimensional data tables, where there may be more measured features than samples. In the process, it chooses a subset of available features that capture the most covariance – these are the features that reflect signals present across multiple tables. We then apply PCA to this selected subset of features. In this sense, we use sparse CCA as a screening procedure, rather than as an ordination method.</p>
<p>The implementation is below. The parameters <code>penaltyx</code> and <code>penaltyz</code> are sparsity penalties. Smaller values of <code>penaltyx</code> will result in fewer selected microbes, similarly <code>penaltyz</code> modulates the number of selected metabolites. We tune them manually to facilitate subsequent interpretation – we generally prefer more sparsity than the default parameters would provide.</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/multitable-sparse-cca_e9ead961b8f7a7a2c5b2acd3a223a7cb">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"PMA"</span>)</span>
<span id="cb96-2"><a href="#cb96-2" aria-hidden="true" tabindex="-1"></a>ccaRes <span class="ot">=</span> <span class="fu">CCA</span>(<span class="fu">t</span>(X), <span class="fu">t</span>(metab), <span class="at">penaltyx =</span> <span class="fl">0.15</span>, <span class="at">penaltyz =</span> <span class="fl">0.15</span>, </span>
<span id="cb96-3"><a href="#cb96-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">typex =</span> <span class="st">"standard"</span>, <span class="at">typez =</span> <span class="st">"standard"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>123456789</code></pre>
</div>
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>ccaRes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Call: CCA(x = t(X), z = t(metab), typex = "standard", typez = "standard", 
    penaltyx = 0.15, penaltyz = 0.15)


Num non-zeros u's:  5 
Num non-zeros v's:  16 
Type of x:  standard 
Type of z:  standard 
Penalty for x: L1 bound is  0.15 
Penalty for z: L1 bound is  0.15 
Cor(Xu,Zv):  0.9904707</code></pre>
</div>
</div>
<p>With these parameters, 5 bacteria and 16 metabolites were selected based on their ability to explain covariation between tables. Further, these features result in a correlation of 0.99 between the two tables. We interpret this to mean that the microbial and metabolomic data reflect similar underlying signals, and that these signals can be approximated well by the selected features. Be wary of the correlation value, however, since the scores are far from the usual bivariate normal cloud. Further, note that it is possible that other subsets of features could explain the data just as well – sparse CCA has minimized redundancy across features, but makes no guarantee that these are the “true” features in any sense.</p>
<p>Nonetheless, we can still use these 21 features to compress information from the two tables without much loss. To relate the recovered metabolites and OTUs to characteristics of the samples on which they were measured, we use them as input to an ordinary PCA. We have omitted the code we used to generate Figure <a href="#fig-multitableinterpretpca" class="quarto-xref"><span>9.32</span></a>, we refer the reader to the online material accompanying the book or the workflow published in <span class="citation" data-cites="Callahan2016Bioc">Callahan et al. (<a href="16-chap.html#ref-Callahan2016Bioc" role="doc-biblioref">2016</a>)</span>.</p>
<p><a href="#fig-multitableinterpretpca" class="quarto-xref">Figure&nbsp;<span>9.32</span></a> displays the PCA <em>triplot</em>, where we show different types of samples and the multidomain features (Metabolites and OTUs). This allows comparison across the measured samples – triangles for knockout and circles for wild type –and characterizes the influence the different features – diamonds with text labels. For example, we see that the main variation in the data is across PD and ST samples, which correspond to the different diets. Further, large values of 15 of the features are associated with ST status, while small values for 5 of them indicate PD status.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-warning.known="unlabeled data points" data-hash="09-chap_cache/html/fig-multitableinterpretpca_910200f4b5aa76ea890bba7f9339894a">
<div class="cell-output-display page-columns page-full">
<div id="fig-multitableinterpretpca" class="quarto-figure quarto-figure-center anchored page-columns page-full">
<figure class="figure page-columns page-full">
<p><img src="09-chap_files/figure-html/fig-multitableinterpretpca-1.png" class="img-fluid figure-img" style="width:75.0%"></p>
<figcaption class="figure-caption margin-caption">Figure&nbsp;9.32: A PCA triplot produced from the CCA selected features from muliple data types (metabolites and OTUs).</figcaption>
</figure>
</div>
</div>
</div>
<p>The advantage of the sparse CCA screening is now clear – we can display most of the variation across samples using a relatively simple plot, and can avoid plotting the hundreds of additional points that would be needed to display all of the features.</p>
</section>
<section id="canonical-or-constrained-correspondence-analysis-ccpna" class="level3 page-columns page-full" data-number="9.6.5">
<h3 data-number="9.6.5" class="anchored" data-anchor-id="canonical-or-constrained-correspondence-analysis-ccpna"><span class="header-section-number">9.6.5</span> Canonical (or constrained) correspondence analysis (CCpnA)</h3>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/unnamed-chunk-124_206b35cbcbb07a78eb984068cc8a4b81">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/devil.png" class="img-fluid figure-img" width="123"></p>
<figcaption class="figure-caption"><strong>Notational overload for CCA</strong>: Originally invented by <span class="citation" data-cites="terBraak:1985">Braak (<a href="16-chap.html#ref-terBraak:1985" role="doc-biblioref">1985</a>)</span> and called Canonical Correspondence analysis, we will call this method Constrained Correspondence Analysis and abbreviate it CCpnA to avoid confusion with Canonical Correlation Analysis (CCA). However several R packages, such as <strong><a href="https://cran.r-project.org/web/packages/ade4/">ade4</a></strong> and <strong><a href="https://cran.r-project.org/web/packages/vegan/">vegan</a></strong> use the name <code>cca</code> for their correspondence analyses function.</figcaption>
</figure>
</div>
</div></div></div>
<p>The term constrained correspondence analysis translates the fact that this method is similar to a constrained regression. The method attempts to force the latent variables to be correlated with the environmental variables provided as `explanatory’.</p>
<p>CCpnA creates biplots where the positions of samples are determined by similarity in both species signatures and environmental characteristics. In contrast, principal components analysis or correspondence analysis only look at species signatures. More formally, it ensures that the resulting CCpnA directions lie in the span of the environmental variables. For thorough explanations see <span class="citation" data-cites="terBraak:1985 greenacre2007">Braak (<a href="16-chap.html#ref-terBraak:1985" role="doc-biblioref">1985</a>; <a href="16-chap.html#ref-greenacre2007" role="doc-biblioref">Greenacre 2007</a>)</span>.</p>
<p>This method can be run using the function <code>ordinate</code> in <strong><a href="https://bioconductor.org/packages/phyloseq/">phyloseq</a></strong>. In order to use the covariates from the sample data, we provide an extra argument, specifying which of the features to consider.</p>
<p>Here, we take the data we denoised using <strong><a href="https://bioconductor.org/packages/dada2/">dada2</a></strong> in <a href="04-chap.html" class="quarto-xref"><span>Chapter&nbsp;4</span></a>. We will see more details about creating the <em>phyloseq</em> object in <a href="10-chap.html" class="quarto-xref"><span>Chapter&nbsp;10</span></a>. For the time being, we use the <code>otu_table</code> component containing a contingency table of counts for different taxa. We would like to compute the constrained correspondence analyses that explain the taxa abundances by the age and family relationship (both variables are contained in the <code>sample_data</code> slot of the <code>ps1</code> object).</p>
<p>We would like to make two dimensional plots showing only using the four most abundant taxa (making the biplot easier to read):</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/ccpna-correspondence-analysis_db4c44b710e230918320aef97189e369">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>ps1<span class="ot">=</span><span class="fu">readRDS</span>(<span class="st">"../data/ps1.rds"</span>)</span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>ps1p<span class="ot">=</span><span class="fu">filter_taxa</span>(ps1, <span class="cf">function</span>(x) <span class="fu">sum</span>(x) <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="cn">TRUE</span>)</span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>psCCpnA <span class="ot">=</span> <span class="fu">ordinate</span>(ps1p, <span class="st">"CCA"</span>,</span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>                 <span class="at">formula =</span> ps1p <span class="sc">~</span> ageBin <span class="sc">+</span> family_relationship)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To access the positions for the biplot, we can use the <code>scores</code> function in the <strong><a href="https://cran.r-project.org/web/packages/vegan/">vegan</a></strong>. Further, to facilitate figure annotation, we also join the site scores with the environmental data in the <code>sample_data</code> slot. Of the 23 total taxonomic orders, we only explicitly annotate the four most abundant – this makes the biplot easier to read.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-warning.known="unlabeled data points" data-hash="09-chap_cache/html/fig-ccpnaplotage_09ffa15a1fc04c99c5afaac6ce7e8257">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>evalProp <span class="ot">=</span> <span class="dv">100</span> <span class="sc">*</span> psCCpnA<span class="sc">$</span>CCA<span class="sc">$</span>eig[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>] <span class="sc">/</span> <span class="fu">sum</span>(psCCpnA<span class="sc">$</span>CA<span class="sc">$</span>eig)</span>
<span id="cb101-2"><a href="#cb101-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb101-3"><a href="#cb101-3" aria-hidden="true" tabindex="-1"></a> <span class="fu">geom_point</span>(<span class="at">data =</span> sites,<span class="fu">aes</span>(<span class="at">x =</span>CCA2, <span class="at">y =</span>CCA1),<span class="at">shape =</span><span class="dv">2</span>,<span class="at">alpha=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb101-4"><a href="#cb101-4" aria-hidden="true" tabindex="-1"></a> <span class="fu">geom_point</span>(<span class="at">data =</span> species,<span class="fu">aes</span>(<span class="at">x =</span>CCA2,<span class="at">y =</span>CCA1,<span class="at">col =</span> Order),<span class="at">size=</span><span class="dv">1</span>)<span class="sc">+</span></span>
<span id="cb101-5"><a href="#cb101-5" aria-hidden="true" tabindex="-1"></a> <span class="fu">geom_text_repel</span>(<span class="at">data =</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(species, CCA2 <span class="sc">&lt;</span> (<span class="sc">-</span><span class="dv">2</span>)),</span>
<span id="cb101-6"><a href="#cb101-6" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">aes</span>(<span class="at">x =</span> CCA2, <span class="at">y =</span> CCA1, <span class="at">label =</span> otu_id),</span>
<span id="cb101-7"><a href="#cb101-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">size =</span> <span class="dv">2</span>, <span class="at">segment.size =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb101-8"><a href="#cb101-8" aria-hidden="true" tabindex="-1"></a> <span class="fu">facet_grid</span>(. <span class="sc">~</span> ageBin) <span class="sc">+</span></span>
<span id="cb101-9"><a href="#cb101-9" aria-hidden="true" tabindex="-1"></a> <span class="fu">guides</span>(<span class="at">col =</span> <span class="fu">guide_legend</span>(<span class="at">override.aes =</span> <span class="fu">list</span>(<span class="at">size =</span> <span class="dv">2</span>))) <span class="sc">+</span></span>
<span id="cb101-10"><a href="#cb101-10" aria-hidden="true" tabindex="-1"></a> <span class="fu">labs</span>(<span class="at">x =</span> <span class="fu">sprintf</span>(<span class="st">"Axis2 [%s%% variance]"</span>, <span class="fu">round</span>(evalProp[<span class="dv">2</span>])),</span>
<span id="cb101-11"><a href="#cb101-11" aria-hidden="true" tabindex="-1"></a>      <span class="at">y =</span> <span class="fu">sprintf</span>(<span class="st">"Axis1 [%s%% variance]"</span>, <span class="fu">round</span>(evalProp[<span class="dv">1</span>]))) <span class="sc">+</span></span>
<span id="cb101-12"><a href="#cb101-12" aria-hidden="true" tabindex="-1"></a> <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Set1"</span>) <span class="sc">+</span> <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"bottom"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-ccpnaplotage" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-ccpnaplotage-1.png" class="img-fluid figure-img" width="320"></p>
<figcaption class="figure-caption">Figure&nbsp;9.33: The mouse and taxa scores generated by CCpnA. The sites (mice samples) are triangles; species are circles, respectively. The separate panels indicate different age groups.</figcaption>
</figure>
</div>
</div></div></div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="prp-hetero-ccaplotage" class="theorem proposition">
<p><span class="theorem-title"><strong>Question 9.19 </strong></span>Look up the extra code for creating the <code>tax</code> and <code>species</code> objects in the online resources accompanying the book. Then make the analogue of <a href="#fig-ccpnaplotage" class="quarto-xref">Figure&nbsp;<span>9.33</span></a> but using litter as the faceting variable.</p>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-33-contents" aria-controls="callout-33" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-33" class="callout-33-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center" data-warning.known="unlabeled data points" data-hash="09-chap_cache/html/fig-ccpnaplotlitter_8dec4d34b3060524547e097548b5842b">
<div class="cell-output-display">
<div id="fig-ccpnaplotlitter" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-ccpnaplotlitter-1.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption">Figure&nbsp;9.34: The analogue to <a href="#fig-ccpnaplotage" class="quarto-xref">Figure&nbsp;<span>9.33</span></a>, faceting by litter membership rather than age bin.</figcaption>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<p>Figures <a href="#fig-ccpnaplotage" class="quarto-xref"><span>9.33</span></a> and <a href="#fig-ccpnaplotlitter" class="quarto-xref"><span>9.34</span></a> show the plots of these annotated scores, splitting sites by their age bin and litter membership, respectively. Note that to keep the appropriate aspect ratio in the presence of faceting, we have taken the vertical axis as our first canonical component. We have labeled individual bacteria that are outliers along the second CCpnA direction.</p>
<p>Evidently, the first CCpnA direction distinguishes between mice in the two main age bins. Circles on the left and right of the biplot represent bacteria that are characteristic of younger and older mice, respectively. The second CCpnA direction splits off the few mice in the oldest age group; it also partially distinguishes between the two litters. These samples low in the second CCpnA direction have more of the outlier bacteria than the others.</p>
<p>This CCpnA analysis supports the conclusion that the main difference between the microbiome communities of the different mice lies along the age axis. However, in situations where the influence of environmental variables is not so strong, CCA can have more power in detecting such associations. In general, it can be applied whenever it is desirable to incorporate supplemental data, but in a way that (1) is less aggressive than supervised methods, and (2) can use several environmental variables at once.</p>
</section>
</section>
<section id="summary-of-this-chapter" class="level2" data-number="9.7">
<h2 data-number="9.7" class="anchored" data-anchor-id="summary-of-this-chapter"><span class="header-section-number">9.7</span> Summary of this chapter</h2>
<p><strong>Heterogeneous data</strong> A mixture of many continuous and a few categorical variables can be handled by adding the categorical variables as supplementary information to the PCA. This is done by projecting the mean of all points in a group onto the map.</p>
<p><strong>Using distances</strong> Relations between data objects can often be summarized as interpoint distances (whether distances between trees, images, graphs, or other complex objects).</p>
<p><strong>Ordination</strong> A useful representation of these distances is available through a method similar to PCA called multidimensional scaling (MDS), otherwise known as PCoA (principal coordinate analysis). It can be helpful to think of the outcome of these analyses as uncovering latent variable. In the case of clustering the latent variables are categorical, in ordination they are latent variables like time or environmental gradients like distance to the water. This is why these methods are often called ordination.</p>
<p><strong>Robust versions</strong> can be used when interpoint distances are wildly different. NMDS (nonmetric multidimensional scaling) aims to produce coordinates such that the order of the interpoint distances is respected as closely as possible.</p>
<p><strong>Correspondence analysis</strong>: a method for computing low dimensional projections that explain dependencies in categorical data. It decomposes chisquare distance in much the same way that PCA decomposes variance. Correspondence analysis is usually the best way to follow up on a significant chisquare test. Once we have ascertained there are significant dependencies between different levels of categories, we can map them and interpret proximities on this map using plots and biplots.</p>
<p><strong>Permutation test for distances</strong> Given two sets of distances between the same points, we can measure whether they are related using the Mantel permutation test.</p>
<p><strong>Generalizations of variance and covariance</strong> When dealing with more than one matrix of measurements on the same data, we can generalize the notion of covariance and correlations to vectorial measurements of co-inertia.</p>
<p><strong>Canonical correlation</strong> is a method for finding a few linear combinations of variables from each table that are as correlated as possible. When using this method on matrices with large numbers of variables, we use a regularized version with an L1 penalty that reduces the number of non-zero coefficients.</p>
</section>
<section id="further-reading" class="level2" data-number="9.8">
<h2 data-number="9.8" class="anchored" data-anchor-id="further-reading"><span class="header-section-number">9.8</span> Further reading</h2>
<p>Interpretation of PCoA maps and nonlinear embeddings can also be enhanced the way we did for PCA using generalizations of the supplementary point method, see <span class="citation" data-cites="Trosset2008">Trosset and Priebe (<a href="16-chap.html#ref-Trosset2008" role="doc-biblioref">2008</a>)</span> or <span class="citation" data-cites="Bengio2004">Bengio et al. (<a href="16-chap.html#ref-Bengio2004" role="doc-biblioref">2004</a>)</span>. We saw in <a href="07-chap.html" class="quarto-xref"><span>Chapter&nbsp;7</span></a> how we can project one categorical variable onto a PCA. The correspondence analysis framework actually allows us to mix several categorical variables in with any number of continuous variables. This is done through an extension called multiple correspondence analysis (MCA) whereby we can do the same analysis on a large number of binary categorical variables and obtain useful maps. The trick here will be to turn the continuous variables into categorical variables first. For extensive examples using R see for instance the book by <span class="citation" data-cites="Pages:2016">Pagès (<a href="16-chap.html#ref-Pages:2016" role="doc-biblioref">2016</a>)</span>.</p>
<p>A simple extension to PCA that allows for <strong>nonlinear</strong> principal curve estimates instead of principal directions defined by eigenvectors was proposed in <span class="citation" data-cites="Hastie1989">Hastie and Stuetzle (<a href="16-chap.html#ref-Hastie1989" role="doc-biblioref">1989</a>)</span> and is available in the package <strong><a href="https://cran.r-project.org/web/packages/princurve/">princurve</a></strong>.</p>
<p>Finding curved subspaces containing a high density data for dimensions higher than <span class="math inline">\(1\)</span> is now called manifold embedding and can be done through Laplacian eigenmaps <span class="citation" data-cites="Belkin2003">(<a href="16-chap.html#ref-Belkin2003" role="doc-biblioref">Belkin and Niyogi 2003</a>)</span>, local linear embedding as in <span class="citation" data-cites="Roweis2000">Roweis and Saul (<a href="16-chap.html#ref-Roweis2000" role="doc-biblioref">2000</a>)</span> or using the isomap method <span class="citation" data-cites="Tenenbaum2000">(<a href="16-chap.html#ref-Tenenbaum2000" role="doc-biblioref">Tenenbaum, De Silva, and Langford 2000</a>)</span>. For textbooks covering nonlinear unsupervised learning methods see <span class="citation" data-cites="HastieTibshiraniFriedman">Hastie, Tibshirani, and Friedman (<a href="16-chap.html#ref-HastieTibshiraniFriedman" role="doc-biblioref">2008, chap. 14</a>)</span> or <span class="citation" data-cites="Izenman2008">Izenman (<a href="16-chap.html#ref-Izenman2008" role="doc-biblioref">2008</a>)</span>.</p>
<p>A review of many multitable correlation coefficients, and analysis of applications can be found in <span class="citation" data-cites="Josse:2016">Josse and Holmes (<a href="16-chap.html#ref-Josse:2016" role="doc-biblioref">2016</a>)</span>.</p>
</section>
<section id="exercises" class="level2 page-columns page-full" data-number="9.9">
<h2 data-number="9.9" class="anchored" data-anchor-id="exercises"><span class="header-section-number">9.9</span> Exercises</h2>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="exr-hetero-phylochip" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 9.1 </strong></span>&nbsp;</p>
</div>
</div>
</div>
<div class="page-columns page-full"><p>We are going to take another look at the Phylochip data, replacing the original expression values by presence/absence. We threshold the data to retain only those that have a value of at least 8.633 in at least 8 samples<a href="#fn7" class="footnote-ref" id="fnref7" role="doc-noteref"><sup>7</sup></a>.</p><div class="no-row-height column-margin column-container"><li id="fn7"><p><sup>7</sup>&nbsp;These values were chosen to give about retain about 3,000 taxa, similar to our previous choice of threshold.</p></li></div></div>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/Threesetscode1_2e41b52ae4aca28a454923f78318e5bd">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>ibd.pres <span class="ot">=</span> <span class="fu">ifelse</span>(assayIBD[, <span class="dv">1</span><span class="sc">:</span><span class="dv">28</span>] <span class="sc">&gt;</span> <span class="fl">8.633</span>, <span class="dv">1</span>, <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Perform a correspondence analysis on these binary data and compare the plot you obtain to what we saw in <a href="#fig-Threesetspca123" class="quarto-xref">Figure&nbsp;<span>9.15</span></a>.</p>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-35-contents" aria-controls="callout-35" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-35" class="callout-35-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>See <a href="#fig-Threesetscoa" class="quarto-xref">Figure&nbsp;<span>9.37</span></a>.</p>
<div>
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a>IBDca <span class="ot">=</span> <span class="fu">dudi.coa</span>(ibd.pres, <span class="at">scannf =</span> <span class="cn">FALSE</span>, <span class="at">nf =</span> <span class="dv">4</span>)</span>
<span id="cb103-2"><a href="#cb103-2" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_eig</span>(IBDca, <span class="at">geom =</span> <span class="st">"bar"</span>, <span class="at">bar_width =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb103-3"><a href="#cb103-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ylab</span>(<span class="st">"Percentage of chisquare"</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">""</span>)</span>
<span id="cb103-4"><a href="#cb103-4" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz</span>(IBDca, <span class="at">element =</span> <span class="st">"col"</span>, <span class="at">axes =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">geom =</span> <span class="st">"point"</span>,</span>
<span id="cb103-5"><a href="#cb103-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">habillage =</span> day, <span class="at">palette =</span> <span class="st">"Dark2"</span>, <span class="at">addEllipses =</span> <span class="cn">TRUE</span>, <span class="at">color =</span> day,</span>
<span id="cb103-6"><a href="#cb103-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">ellipse.type =</span> <span class="st">"convex"</span>, <span class="at">alpha =</span> <span class="dv">1</span>, <span class="at">col.row.sup =</span>  <span class="st">"blue"</span>,</span>
<span id="cb103-7"><a href="#cb103-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">select =</span> <span class="fu">list</span>(<span class="at">name =</span> <span class="cn">NULL</span>, <span class="at">cos2 =</span> <span class="cn">NULL</span>, <span class="at">contrib =</span> <span class="cn">NULL</span>),</span>
<span id="cb103-8"><a href="#cb103-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">repel =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-Threesetscoa" class="cell quarto-layout-panel" data-hash="09-chap_cache/html/fig-Threesetscoa_3373c0e167051b26c3fc81a818b43f8b">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-Threesetscoa-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-Threesetscoa-1.png" class="img-fluid figure-img" width="320"></p>
<figcaption class="figure-caption">Figure&nbsp;9.35: <span class="math inline">\(\text{}\)</span></figcaption>
</figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-Threesetscoa-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-Threesetscoa-2.png" class="img-fluid figure-img" width="320"></p>
<figcaption class="figure-caption">Figure&nbsp;9.36: <span class="math inline">\(\text{}\)</span></figcaption>
</figure>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;9.37: Correspondence analysis on binary data.</figcaption><p></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="exr-hetero-ColorTable" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 9.2 </strong></span>&nbsp;</p>
</div>
</div>
</div>
<p>Correspondence Analysis on color association tables:<br>
Here is an example of data collected by looking at the number of Google hits resulting from queries of pairs of words. The numbers in <a href="#tbl-colors" class="quarto-xref">Table&nbsp;<span>9.4</span></a> are to be multiplied by 1000. For instance, the combination of the words “quiet” and “blue” returned 2,150,000 hits.</p>
<div class="cell tbl-cap-location-top" data-layout-align="center" data-hash="09-chap_cache/html/tbl-colors_74205f75d728729d49a5804ab2a78254">
<div class="cell-output-display">
<div id="tbl-colors" class="anchored">
<table data-quarto-postprocess="true" class="table table-sm table-striped small">
<caption>Table&nbsp;9.4: Contingency table of co-occurring terms from search engine results.</caption>
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th"></th>
<th style="text-align: right;" data-quarto-table-cell-role="th">black</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">blue</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">green</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">grey</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">orange</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">purple</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">white</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">quiet</td>
<td style="text-align: right;">2770</td>
<td style="text-align: right;">2150</td>
<td style="text-align: right;">2140</td>
<td style="text-align: right;">875</td>
<td style="text-align: right;">1220</td>
<td style="text-align: right;">821</td>
<td style="text-align: right;">2510</td>
</tr>
<tr class="even">
<td style="text-align: left;">angry</td>
<td style="text-align: right;">2970</td>
<td style="text-align: right;">1530</td>
<td style="text-align: right;">1740</td>
<td style="text-align: right;">752</td>
<td style="text-align: right;">1040</td>
<td style="text-align: right;">710</td>
<td style="text-align: right;">1730</td>
</tr>
<tr class="odd">
<td style="text-align: left;">clever</td>
<td style="text-align: right;">1650</td>
<td style="text-align: right;">1270</td>
<td style="text-align: right;">1320</td>
<td style="text-align: right;">495</td>
<td style="text-align: right;">693</td>
<td style="text-align: right;">416</td>
<td style="text-align: right;">1420</td>
</tr>
<tr class="even">
<td style="text-align: left;">depressed</td>
<td style="text-align: right;">1480</td>
<td style="text-align: right;">957</td>
<td style="text-align: right;">983</td>
<td style="text-align: right;">147</td>
<td style="text-align: right;">330</td>
<td style="text-align: right;">102</td>
<td style="text-align: right;">1270</td>
</tr>
<tr class="odd">
<td style="text-align: left;">happy</td>
<td style="text-align: right;">19300</td>
<td style="text-align: right;">8310</td>
<td style="text-align: right;">8730</td>
<td style="text-align: right;">1920</td>
<td style="text-align: right;">4220</td>
<td style="text-align: right;">2610</td>
<td style="text-align: right;">9150</td>
</tr>
<tr class="even">
<td style="text-align: left;">lively</td>
<td style="text-align: right;">1840</td>
<td style="text-align: right;">1250</td>
<td style="text-align: right;">1350</td>
<td style="text-align: right;">659</td>
<td style="text-align: right;">621</td>
<td style="text-align: right;">488</td>
<td style="text-align: right;">1480</td>
</tr>
<tr class="odd">
<td style="text-align: left;">perplexed</td>
<td style="text-align: right;">110</td>
<td style="text-align: right;">71</td>
<td style="text-align: right;">80</td>
<td style="text-align: right;">19</td>
<td style="text-align: right;">23</td>
<td style="text-align: right;">15</td>
<td style="text-align: right;">109</td>
</tr>
<tr class="even">
<td style="text-align: left;">virtuous</td>
<td style="text-align: right;">179</td>
<td style="text-align: right;">80</td>
<td style="text-align: right;">102</td>
<td style="text-align: right;">20</td>
<td style="text-align: right;">25</td>
<td style="text-align: right;">17</td>
<td style="text-align: right;">165</td>
</tr>
</tbody>
</table>
</div>


</div>
</div>
<p>Perform a correspondence analysis of these data. What do you notice when you look at the two-dimensional biplot?</p>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-37-contents" aria-controls="callout-37" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-37" class="callout-37-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>See <a href="#fig-ColorBiplot-1" class="quarto-xref">Figure&nbsp;<span>9.38</span></a>. The code is not rendered here, but is shown in the document’s source file.</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/fig-ColorBiplot-1_a091c6128b0be337a07d840b94989290">
<div class="cell-output-display">
<div id="fig-ColorBiplot-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-ColorBiplot-1-1.png" class="img-fluid figure-img" style="width:50.0%"></p>
<figcaption class="figure-caption">Figure&nbsp;9.38: Correspondence Analysis allows for a symmetrical graphical representation of two categorical variables, in this case colors and emotions for a contingency table of co-occurrences such as <a href="#tbl-colors" class="quarto-xref">Table&nbsp;<span>9.4</span></a>.</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/PlatoTableImage_375c7013b14d008e1f6e396fa5a4596f">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="imgs/PlatoTableImage.png" class="img-fluid figure-img" style="width:50.0%"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="exr-hetero-PlatoSentence" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 9.3 </strong></span>&nbsp;</p>
</div>
</div>
</div>
<p>The dates Plato wrote his various books are not known. We take the sentence endings and use those pattern frequencies as the data.</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/platoca_c11b14ab80e34d09fef300a433a63558">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>platof <span class="ot">=</span> <span class="fu">read.table</span>(<span class="st">"../data/platof.txt"</span>, <span class="at">header =</span> <span class="cn">TRUE</span>)</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>platof[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Rep Laws Crit Phil Pol Soph Tim
uuuuu  42   91    5   24  13   26  18
-uuuu  60  144    3   27  19   33  30
u-uuu  64   72    3   20  24   31  46
uu-uu  72   98    2   25  20   24  14</code></pre>
</div>
</div>
<div class="page-columns page-full">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>resPlato <span class="ot">=</span> <span class="fu">dudi.coa</span>(platof, <span class="at">scannf =</span> <span class="cn">FALSE</span>, <span class="at">nf =</span> <span class="dv">2</span>)</span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_ca_biplot</span>(resPlato, <span class="at">axes=</span><span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">1</span>)) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">""</span>)</span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a><span class="fu">fviz_eig</span>(resPlato, <span class="at">geom =</span> <span class="st">"bar"</span>, <span class="at">width =</span> <span class="fl">0.6</span>) <span class="sc">+</span> <span class="fu">ggtitle</span>(<span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="no-row-height column-margin column-container"><div id="fig-platoca" class="cell quarto-layout-panel" data-hash="09-chap_cache/html/fig-platoca_fd3dcac0f47cdf40aeba53ee2f946266">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 100.0%;justify-content: center;">
<div id="fig-platoca-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-platoca-1.png" class="img-fluid figure-img" data-ref-parent="fig-platoca" width="320"></p>
<figcaption class="figure-caption">(a)</figcaption>
</figure>
</div>
</div>
</div>
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell quarto-layout-cell-subref" style="flex-basis: 100.0%;justify-content: center;">
<div id="fig-platoca-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-platoca-2.png" class="img-fluid figure-img" data-ref-parent="fig-platoca" width="320"></p>
<figcaption class="figure-caption">(b)</figcaption>
</figure>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;9.39: Biplot of Plato’s sentence endings.</figcaption><p></p>
</figure>
</div></div></div>
<ol type="a">
<li><p>From the biplot in <a href="#fig-platoca" class="quarto-xref">Figure&nbsp;<span>9.39</span></a> can you guess at the chronological order of Plato’s works?<br>
Hint: the first (earliest) is known to be <em>Republica</em>. The last (latest) is known to be <em>Laws</em>.</p></li>
<li><p>Which sentence ending did Plato use more frequently early in his life?</p></li>
<li><p>What percentage of the inertia (<span class="math inline">\(\chi^2\)</span>-distance) is explained by the map in <a href="#fig-platoca" class="quarto-xref">Figure&nbsp;<span>9.39</span></a>?</p></li>
</ol>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-39-contents" aria-controls="callout-39" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-39" class="callout-39-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<p>To compute the percentage of inertia explained by the first two axes we take the cumulative sum of the eigenvalues at the value 2:</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/PercentageInertia_69efff442cef4077c46d6a12224389f5">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(resPlato)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "tab"  "cw"   "lw"   "eig"  "rank" "nf"   "c1"   "li"   "co"   "l1"  
[11] "call" "N"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(resPlato<span class="sc">$</span>eig)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.132618</code></pre>
</div>
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>percentageInertia<span class="ot">=</span><span class="fu">round</span>(<span class="dv">100</span><span class="sc">*</span><span class="fu">cumsum</span>(resPlato<span class="sc">$</span>eig)<span class="sc">/</span><span class="fu">sum</span>(resPlato<span class="sc">$</span>eig))</span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a>percentageInertia</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  69  85  92  96  98 100</code></pre>
</div>
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a>percentageInertia[<span class="dv">2</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 85</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="exr-hetero-ComparisonCA" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 9.4 </strong></span>&nbsp;</p>
</div>
</div>
</div>
<p>We are going to look at two datasets, one is a perturbed version of the other and they both present gradients as often seen in ecological data. Read in the two species count matrices <code>lakelike</code> and <code>lakelikeh</code>, which are stored as the object <code>lakes.RData</code>. Compare the output of correspondence analysis and principal component analysis on each of the two data sets; restrict yourself two dimensions. In the plots and the eigenvalues, what do you notice?</p>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-41-contents" aria-controls="callout-41" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-41" class="callout-41-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/MakeMatrices_baa19c8af1bdd0e817ec82333916bfad">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"../data/lakes.RData"</span>)</span>
<span id="cb115-2"><a href="#cb115-2" aria-hidden="true" tabindex="-1"></a>lakelike[ <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     plant1 plant2 plant3 plant4 plant5 plant6 plant7 plant8
loc1      6      4      0      3      0      0      0      0
loc2      4      5      5      3      4      2      0      0
loc3      3      4      7      4      5      2      1      1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a>lakelikeh[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">8</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     plant1 plant2 plant3 plant4 plant5 plant6 plant7 plant8
loc1      6      4      0      3      0      0      0      0
loc2      4      5      5      3      4      2      0      0
loc3      3      4      7      4      5      2      1      1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a>e_coa  <span class="ot">=</span> <span class="fu">dudi.coa</span>(lakelike,  <span class="at">scannf =</span> <span class="cn">FALSE</span>, <span class="at">nf =</span> <span class="dv">2</span>)</span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a>e_pca  <span class="ot">=</span> <span class="fu">dudi.pca</span>(lakelike,  <span class="at">scannf =</span> <span class="cn">FALSE</span>, <span class="at">nf =</span> <span class="dv">2</span>)</span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a>eh_coa <span class="ot">=</span> <span class="fu">dudi.coa</span>(lakelikeh, <span class="at">scannf =</span> <span class="cn">FALSE</span>, <span class="at">nf =</span> <span class="dv">2</span>)</span>
<span id="cb119-4"><a href="#cb119-4" aria-hidden="true" tabindex="-1"></a>eh_pca <span class="ot">=</span> <span class="fu">dudi.pca</span>(lakelikeh, <span class="at">scannf =</span> <span class="cn">FALSE</span>, <span class="at">nf =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Comparison (output not shown):</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/adescatter-show_61b7adc2a108b5414fd95c228de13b94">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scatter</span>(e_pca)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scatter</span>(e_coa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="fu">s.label</span>(e_pca<span class="sc">$</span>li)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a><span class="fu">s.label</span>(e_coa<span class="sc">$</span>li)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="fu">s.label</span>(eh_pca<span class="sc">$</span>co)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="fu">s.label</span>(eh_pca<span class="sc">$</span>li)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="fu">s.label</span>(eh_coa<span class="sc">$</span>li)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="fu">s.label</span>(eh_coa<span class="sc">$</span>co)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="exr-hetero-RawMoignard" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 9.5 </strong></span>&nbsp;</p>
</div>
</div>
</div>
<p>We analyzed the normalized Moignard data in <a href="#sec-hetero-celldynamics" class="quarto-xref"><span>Section&nbsp;9.5.1</span></a>. Now redo the analysis with the <em>raw</em> data (in file <em>nbt.3154-S3-raw.csv</em>) and compare the output with that obtained using the normalized values.</p>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-43-contents" aria-controls="callout-43" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-43" class="callout-43-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/RawData_3c2c5534588c32b33d8b1c9b3f619c8b">
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a>moignard_raw <span class="ot">=</span> <span class="fu">as.matrix</span>(<span class="fu">read.csv</span>(<span class="st">"../data/nbt.3154-S3-raw.csv"</span>, <span class="at">row.names =</span> <span class="dv">1</span>))</span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a>dist2r.euclid <span class="ot">=</span> <span class="fu">dist</span>(moignard_raw)</span>
<span id="cb128-3"><a href="#cb128-3" aria-hidden="true" tabindex="-1"></a>dist1r.l1     <span class="ot">=</span> <span class="fu">dist</span>(moignard_raw, <span class="st">"manhattan"</span>)</span>
<span id="cb128-4"><a href="#cb128-4" aria-hidden="true" tabindex="-1"></a>cells1.cmds <span class="ot">=</span> <span class="fu">cmdscale</span>(dist1r.l1,     <span class="at">k =</span> <span class="dv">20</span>, <span class="at">eig =</span> <span class="cn">TRUE</span>)</span>
<span id="cb128-5"><a href="#cb128-5" aria-hidden="true" tabindex="-1"></a>cells2.cmds <span class="ot">=</span> <span class="fu">cmdscale</span>(dist2r.euclid, <span class="at">k =</span> <span class="dv">20</span>, <span class="at">eig =</span> <span class="cn">TRUE</span>)</span>
<span id="cb128-6"><a href="#cb128-6" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(cells1.cmds<span class="sc">$</span>eig[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]) <span class="sc">/</span> <span class="fu">sum</span>(cells1.cmds<span class="sc">$</span>eig)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.776075</code></pre>
</div>
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(cells2.cmds<span class="sc">$</span>eig[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]) <span class="sc">/</span> <span class="fu">sum</span>(cells2.cmds<span class="sc">$</span>eig)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.6297133</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="exr-hetero-KernelMethods" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 9.6 </strong></span>&nbsp;</p>
</div>
</div>
</div>
<p>We are going to explore the use of kernel methods.</p>
<ol type="a">
<li><p>Compute kernelized distances using the <strong><a href="https://cran.r-project.org/web/packages/kernlab/">kernlab</a></strong> for the Moignard data using various values for the sigma tuning parameter in the definition of the kernels. Then perform MDS on these kernelized distances. What difference is there in variability explained by the first four components of kernel multidimensional scaling?</p></li>
<li><p>Make interactive three dimensional representations of the components: is there a projection where you see a branch for the purple points?</p></li>
</ol>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-45-contents" aria-controls="callout-45" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-45" class="callout-45-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<ol type="a">
<li>kernelized distances</li>
</ol>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/KernelD1_700f55472cdb9b87e28e2e28703cd881">
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"kernlab"</span>)</span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a>laplacedot1 <span class="ot">=</span> <span class="fu">laplacedot</span>(<span class="at">sigma =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">3934</span>)</span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a>rbfdot1     <span class="ot">=</span> <span class="fu">rbfdot</span>(<span class="at">sigma =</span> (<span class="dv">1</span><span class="sc">/</span><span class="dv">3934</span>)<span class="sc">^</span><span class="dv">2</span> )</span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a>Klaplace_cellsn   <span class="ot">=</span> <span class="fu">kernelMatrix</span>(laplacedot1, blom)</span>
<span id="cb132-5"><a href="#cb132-5" aria-hidden="true" tabindex="-1"></a>KGauss_cellsn     <span class="ot">=</span> <span class="fu">kernelMatrix</span>(rbfdot1, blom)</span>
<span id="cb132-6"><a href="#cb132-6" aria-hidden="true" tabindex="-1"></a>Klaplace_rawcells <span class="ot">=</span> <span class="fu">kernelMatrix</span>(laplacedot1, moignard_raw)</span>
<span id="cb132-7"><a href="#cb132-7" aria-hidden="true" tabindex="-1"></a>KGauss_rawcells   <span class="ot">=</span> <span class="fu">kernelMatrix</span>(rbfdot1, moignard_raw)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Use kernelized distances to protect against outliers and allows discovery of non-linear components.</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/KernelD2_761a4d741cb0317ebab178f26f1ef2a8">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a>dist1kr <span class="ot">=</span> <span class="dv">1</span> <span class="sc">-</span> Klaplace_rawcells</span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a>dist2kr <span class="ot">=</span> <span class="dv">1</span> <span class="sc">-</span> KGauss_rawcells</span>
<span id="cb133-3"><a href="#cb133-3" aria-hidden="true" tabindex="-1"></a>dist1kn <span class="ot">=</span> <span class="dv">1</span> <span class="sc">-</span> Klaplace_cellsn</span>
<span id="cb133-4"><a href="#cb133-4" aria-hidden="true" tabindex="-1"></a>dist2kn <span class="ot">=</span> <span class="dv">1</span> <span class="sc">-</span> KGauss_cellsn</span>
<span id="cb133-5"><a href="#cb133-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-6"><a href="#cb133-6" aria-hidden="true" tabindex="-1"></a>cells1.kcmds <span class="ot">=</span> <span class="fu">cmdscale</span>(dist1kr, <span class="at">k =</span> <span class="dv">20</span>, <span class="at">eig =</span> <span class="cn">TRUE</span>) </span>
<span id="cb133-7"><a href="#cb133-7" aria-hidden="true" tabindex="-1"></a>cells2.kcmds <span class="ot">=</span> <span class="fu">cmdscale</span>(dist2kr, <span class="at">k =</span> <span class="dv">20</span>, <span class="at">eig =</span> <span class="cn">TRUE</span>) </span>
<span id="cb133-8"><a href="#cb133-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-9"><a href="#cb133-9" aria-hidden="true" tabindex="-1"></a>percentage <span class="ot">=</span> <span class="cf">function</span>(x, <span class="at">n =</span> <span class="dv">4</span>) <span class="fu">round</span>(<span class="dv">100</span> <span class="sc">*</span> <span class="fu">sum</span>(x[<span class="fu">seq_len</span>(n)]) <span class="sc">/</span> <span class="fu">sum</span>(x[x<span class="sc">&gt;</span><span class="dv">0</span>]))</span>
<span id="cb133-10"><a href="#cb133-10" aria-hidden="true" tabindex="-1"></a>kperc1 <span class="ot">=</span> <span class="fu">percentage</span>(cells1.kcmds<span class="sc">$</span>eig)</span>
<span id="cb133-11"><a href="#cb133-11" aria-hidden="true" tabindex="-1"></a>kperc2 <span class="ot">=</span> <span class="fu">percentage</span>(cells2.kcmds<span class="sc">$</span>eig)</span>
<span id="cb133-12"><a href="#cb133-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb133-13"><a href="#cb133-13" aria-hidden="true" tabindex="-1"></a>cellsn1.kcmds <span class="ot">=</span> <span class="fu">cmdscale</span>(dist1kn, <span class="at">k =</span> <span class="dv">20</span>, <span class="at">eig =</span> <span class="cn">TRUE</span>) </span>
<span id="cb133-14"><a href="#cb133-14" aria-hidden="true" tabindex="-1"></a>cellsn2.kcmds <span class="ot">=</span> <span class="fu">cmdscale</span>(dist2kn, <span class="at">k =</span> <span class="dv">20</span>, <span class="at">eig =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="2" type="a">
<li>using a 3d scatterplot interactively:</li>
</ol>
<div>
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a>colc <span class="ot">=</span> <span class="fu">rowData</span>(Moignard)<span class="sc">$</span>cellcol</span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"scatterplot3d"</span>)</span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a><span class="fu">scatterplot3d</span>(cellsn2.kcmds<span class="sc">$</span>points[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">color=</span>colc, <span class="at">pch =</span> <span class="dv">20</span>,</span>
<span id="cb134-4"><a href="#cb134-4" aria-hidden="true" tabindex="-1"></a>   <span class="at">xlab =</span> <span class="st">"Axis k1"</span>, <span class="at">ylab =</span> <span class="st">"Axis k2"</span>, <span class="at">zlab =</span> <span class="st">"Axis k3"</span>, <span class="at">angle=</span><span class="dv">15</span>)</span>
<span id="cb134-5"><a href="#cb134-5" aria-hidden="true" tabindex="-1"></a><span class="fu">scatterplot3d</span>(cellsn2.kcmds<span class="sc">$</span>points[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">color=</span>colc, <span class="at">pch =</span> <span class="dv">20</span>,</span>
<span id="cb134-6"><a href="#cb134-6" aria-hidden="true" tabindex="-1"></a>   <span class="at">xlab =</span> <span class="st">"Axis k1"</span>, <span class="at">ylab =</span> <span class="st">"Axis k2"</span>, <span class="at">zlab =</span> <span class="st">"Axis k3"</span>, <span class="at">angle =</span> <span class="sc">-</span><span class="dv">70</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div id="fig-KernelMDSplots" class="cell quarto-layout-panel" data-hash="09-chap_cache/html/fig-KernelMDSplots_453cda271a0cfeff745177ab4bc233ea">
<figure class="figure">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-KernelMDSplots-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-KernelMDSplots-1.png" class="img-fluid figure-img" width="320"></p>
<figcaption class="figure-caption">Figure&nbsp;9.40: <span class="math inline">\(\text{}\)</span></figcaption>
</figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-KernelMDSplots-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="09-chap_files/figure-html/fig-KernelMDSplots-2.png" class="img-fluid figure-img" width="320"></p>
<figcaption class="figure-caption">Figure&nbsp;9.41: <span class="math inline">\(\text{}\)</span></figcaption>
</figure>
</div>
</div>
</div>
<p></p><figcaption class="figure-caption">Figure&nbsp;9.42: Kernel multidimensional scaling.</figcaption><p></p>
</figure>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="exr-hetero-celldatahigh" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 9.7 </strong></span>&nbsp;</p>
</div>
</div>
</div>
<p><strong>Higher resolution study of cell data.</strong><br>
Take the original expression data <code>blom</code> we generated in <a href="#sec-hetero-celldynamics" class="quarto-xref"><span>Section&nbsp;9.5.1</span></a>. Map the intensity of expression of each of the top 10 most variable genes onto the 3d plot made with the diffusion mapping. Which dimension, or which one of the principal coordinates (1,2,3,4) can be seen as the one that clusters the <strong>4SG</strong> (red) points the most?</p>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-47-contents" aria-controls="callout-47" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-47" class="callout-47-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/CodeB_d10a075558bd0286de1bd6dd5a665aa6">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"rgl"</span>)</span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot3d</span>(cellsn2.kcmds<span class="sc">$</span>points[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">col =</span> colc, <span class="at">size =</span> <span class="dv">3</span>,</span>
<span id="cb135-3"><a href="#cb135-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">xlab =</span> <span class="st">"Axis1"</span>, <span class="at">ylab =</span> <span class="st">"Axis2"</span>, <span class="at">zlab =</span> <span class="st">"Axis3"</span>)</span>
<span id="cb135-4"><a href="#cb135-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot3d</span>(cellsn2.kcmds<span class="sc">$</span>points[, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>)], <span class="at">col =</span> colc, <span class="at">size =</span> <span class="dv">3</span>,</span>
<span id="cb135-5"><a href="#cb135-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">xlab =</span> <span class="st">"Axis1"</span>, <span class="at">ylab =</span> <span class="st">"Axis2"</span>, <span class="at">zlab =</span> <span class="st">"Axis4"</span>)</span>
<span id="cb135-6"><a href="#cb135-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Using an L1 distance instead.</span></span>
<span id="cb135-7"><a href="#cb135-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot3d</span>(cellsn1.kcmds<span class="sc">$</span>points[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">col =</span> colc, <span class="at">size =</span> <span class="dv">3</span>,</span>
<span id="cb135-8"><a href="#cb135-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">xlab =</span> <span class="st">"Axis1"</span>, <span class="at">ylab =</span> <span class="st">"Axis2"</span>, <span class="at">zlab =</span> <span class="st">"Axis3"</span>)</span>
<span id="cb135-9"><a href="#cb135-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot3d</span>(cellsn1.kcmds<span class="sc">$</span>points[, <span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">4</span>)], <span class="at">col =</span> colc, <span class="at">size =</span> <span class="dv">3</span>,</span>
<span id="cb135-10"><a href="#cb135-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">xlab =</span> <span class="st">"Axis1"</span>, <span class="at">ylab =</span> <span class="st">"Axis2"</span>, <span class="at">zlab =</span> <span class="st">"Axis4"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>An implementation in the package <strong><a href="https://cran.r-project.org/web/packages/LPCM/">LPCM</a></strong> provides the function <code>lpc</code>, which estimates principal curves. Here we constrain ourselves to three dimensions chosen from the output of the diffusion map and create smoothed curves.</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/lpc3d_734baa117f5540b06e6c40cb05238505">
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"LPCM"</span>)</span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"diffusionMap"</span>)</span>
<span id="cb136-3"><a href="#cb136-3" aria-hidden="true" tabindex="-1"></a>dmap1 <span class="ot">=</span> <span class="fu">diffuse</span>(dist1n.l1, <span class="at">neigen =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Performing eigendecomposition
Computing Diffusion Coordinates
Elapsed time: 9.804 seconds</code></pre>
</div>
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>combs <span class="ot">=</span> <span class="fu">combn</span>(<span class="dv">4</span>, <span class="dv">3</span>)</span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a>lpcplots <span class="ot">=</span> <span class="fu">apply</span>(combs, <span class="dv">2</span>, <span class="cf">function</span>(j) <span class="fu">lpc</span>(dmap1<span class="sc">$</span>X[, j], <span class="at">scale =</span> <span class="cn">FALSE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To get a feel for what the smoothed data are showing us, we take a look at the interactive graphics using the function <code>plot3d</code> from them <strong><a href="https://cran.r-project.org/web/packages/rgl/">rgl</a></strong> package.</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/PLOT3DLPC_0c99f435aaa3aa96a29c92bdba05722b">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"rgl"</span>)</span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_along</span>(lpcplots))</span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(lpcplots[[i]], <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lwd =</span> <span class="dv">3</span>,</span>
<span id="cb139-4"><a href="#cb139-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="fu">paste</span>(<span class="st">"Axis"</span>, combs[<span class="dv">1</span>, i]),</span>
<span id="cb139-5"><a href="#cb139-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="fu">paste</span>(<span class="st">"Axis"</span>, combs[<span class="dv">2</span>, i]),</span>
<span id="cb139-6"><a href="#cb139-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">zlab =</span> <span class="fu">paste</span>(<span class="st">"Axis"</span>, combs[<span class="dv">3</span>, i]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>

<p>One way of plotting both the smoothed line and the data points is to add the line using the <code>plot3d</code> function.</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/scatter3smoothedline_2e1058b7120ca8c7948e7cb07e9e2ce7">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a>outlpce134 <span class="ot">=</span> <span class="fu">lpc</span>(dmap1<span class="sc">$</span>X[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">4</span>)], <span class="at">scale=</span><span class="cn">FALSE</span>, <span class="at">h=</span><span class="fl">0.5</span>)</span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot3d</span>(dmap1<span class="sc">$</span>X[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">4</span>)], <span class="at">col=</span>colc, <span class="at">pch=</span><span class="dv">20</span>, </span>
<span id="cb140-3"><a href="#cb140-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">xlab=</span><span class="st">"Axis1"</span>, <span class="at">ylab=</span><span class="st">"Axis3"</span>, <span class="at">zlab=</span><span class="st">"Axis4"</span>)</span>
<span id="cb140-4"><a href="#cb140-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot3d</span>(outlpce134<span class="sc">$</span>LPC, <span class="at">type=</span><span class="st">"l"</span>, <span class="at">lwd=</span><span class="dv">7</span>, <span class="at">add=</span><span class="cn">TRUE</span>)</span>
<span id="cb140-5"><a href="#cb140-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb140-6"><a href="#cb140-6" aria-hidden="true" tabindex="-1"></a>outlpce134 <span class="ot">=</span> <span class="fu">lpc</span>(dmap1<span class="sc">$</span>X[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">4</span>)], <span class="at">scale=</span><span class="cn">FALSE</span>, <span class="at">h=</span><span class="fl">0.7</span>)</span>
<span id="cb140-7"><a href="#cb140-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot3d</span>(outlpce134<span class="sc">$</span>LPC, <span class="at">type=</span><span class="st">"l"</span>, <span class="at">lwd=</span><span class="dv">7</span>,</span>
<span id="cb140-8"><a href="#cb140-8" aria-hidden="true" tabindex="-1"></a>       <span class="at">xlab=</span><span class="st">"Axis1"</span>, <span class="at">ylab=</span><span class="st">"Axis3"</span>, <span class="at">zlab=</span><span class="st">"Axis4"</span>)</span>
<span id="cb140-9"><a href="#cb140-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot3d</span>(dmap1<span class="sc">$</span>X[,<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">3</span>,<span class="dv">4</span>)], <span class="at">col=</span>colc, </span>
<span id="cb140-10"><a href="#cb140-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">xlab=</span><span class="st">""</span>, <span class="at">ylab=</span><span class="st">""</span>, <span class="at">zlab=</span><span class="st">""</span>, <span class="at">add=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
<div class="no-row-height column-margin column-container"><div class="cell quarto-layout-panel callout-47-contents callout-collapse collapse callout-margin-content" data-hash="09-chap_cache/html/fig-diffusionmap3_7e8b7ab70878d71577e2d9d92b062c30">
<div class="quarto-layout-row quarto-layout-valign-top">
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-diffusionmap3-1" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="imgs/TripleArm.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;9.43: Diffusion map projection for Axes 1, 3 and 4. The lower figure shows the smoothed path followed by the cells in their development.</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display quarto-layout-cell" style="flex-basis: 50.0%;justify-content: center;">
<div id="fig-diffusionmap3-2" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="imgs/SmoothLineP134h7.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;9.44: Diffusion map projection for Axes 1, 3 and 4. The lower figure shows the smoothed path followed by the cells in their development.</figcaption>
</figure>
</div>
</div>
</div>
</div></div><div class="callout callout-style-default callout-important no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
</div>
<div class="callout-body-container callout-body">
<div id="exr-hetero-moredistancescell" class="theorem exercise">
<p><span class="theorem-title"><strong>Exercise 9.8 </strong></span>&nbsp;</p>
</div>
</div>
</div>
<p>Here we explore more refined distances and diffusion maps that can show cell development trajectories as in <a href="#fig-dmap134" class="quarto-xref">Figure&nbsp;<span>9.45</span></a>.</p>
<div class="cell page-columns page-full" data-layout-align="center" data-hash="09-chap_cache/html/fig-dmap134_d8042a8ad37f0c48b13f244757a62b1d">

<div class="no-row-height column-margin column-container"><div class="cell-output-display">
<div id="fig-dmap134" class="quarto-figure quarto-figure-center anchored">
<figure class="figure">
<p><img src="imgs/dmap134.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Figure&nbsp;9.45: Ouput from a three-dimensional diffusion map projection.</figcaption>
</figure>
</div>
</div></div></div>
<p>The diffusion map method restricts the estimation of distances to local points, thus further pursuing the idea that often only local distances should be represented precisely and as points become further apart they are not being measured with the same ‘reference’. This method also uses the distances as input but then creates local probabilistic transitions as indicators of similarity, these are combined into an affinity matrix for which the eigenvalues and eigenvectors are also computed much like in standard MDS.</p>
<p>Compare the output of the <code>diffuse</code> function from the <strong><a href="https://cran.r-project.org/web/packages/diffusionMap/">diffusionMap</a></strong> package on both the l1 and l2 distances computed between the cells available in the <code>dist2n.euclid</code> and <code>dist1n.l1</code> objects from <a href="#sec-hetero-celldynamics" class="quarto-xref"><span>Section&nbsp;9.5.1</span></a>.</p>
<div class="callout callout-style-default callout-tip no-icon callout-titled">
<div class="callout-header d-flex align-content-center" data-bs-toggle="collapse" data-bs-target=".callout-49-contents" aria-controls="callout-49" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">

</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-49" class="callout-49-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/Diffuse_1c3edffa969b90edf94c98e1ae6e1711">
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"diffusionMap"</span>)</span>
<span id="cb141-2"><a href="#cb141-2" aria-hidden="true" tabindex="-1"></a>dmap2 <span class="ot">=</span> <span class="fu">diffuse</span>(dist2n.euclid, <span class="at">neigen =</span> <span class="dv">11</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Performing eigendecomposition
Computing Diffusion Coordinates
Elapsed time: 12.86 seconds</code></pre>
</div>
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a>dmap1 <span class="ot">=</span> <span class="fu">diffuse</span>(dist1n.l1, <span class="at">neigen =</span> <span class="dv">11</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Performing eigendecomposition
Computing Diffusion Coordinates
Elapsed time: 11.58 seconds</code></pre>
</div>
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(dmap2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Notice that the vanilla plot for a <em>dmap</em> object does not allow the use of colors. As this essential to our understanding of cell development, we add the colors by hand. Of course, here we use static 3d plots but these should supplemented by the plot3d examples we give in the code.</p>
<p>We use a tailored wrapper function <code>scp3d</code>, so that we can easily insert relevant parameters:</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/scp3d_512a40f8aaa25139cf3f1cae1f1c51f3">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"scatterplot3d"</span>)</span>
<span id="cb146-2"><a href="#cb146-2" aria-hidden="true" tabindex="-1"></a>scp3d <span class="ot">=</span> <span class="cf">function</span>(<span class="at">axestop =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="at">dmapRes =</span> dmap1, <span class="at">color =</span> colc,</span>
<span id="cb146-3"><a href="#cb146-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">anglea =</span> <span class="dv">20</span>, <span class="at">pch =</span> <span class="dv">20</span>)</span>
<span id="cb146-4"><a href="#cb146-4" aria-hidden="true" tabindex="-1"></a><span class="fu">scatterplot3d</span>(dmapRes<span class="sc">$</span>X[, axestop], <span class="at">color =</span> colc,</span>
<span id="cb146-5"><a href="#cb146-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlab =</span> <span class="fu">paste</span>(<span class="st">"Axis"</span>,axestop[<span class="dv">1</span>]), <span class="at">ylab =</span> <span class="fu">paste</span>(<span class="st">"Axis"</span>, axestop[<span class="dv">2</span>]),</span>
<span id="cb146-6"><a href="#cb146-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">zlab =</span> <span class="fu">paste</span>(<span class="st">"Axis"</span>,axestop[<span class="dv">3</span>]), <span class="at">pch =</span> pch, <span class="at">angle =</span> anglea)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/dmap3dplots-show_4737605beddc1398281cfee55861b6bf">
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="fu">scp3d</span>()</span>
<span id="cb147-2"><a href="#cb147-2" aria-hidden="true" tabindex="-1"></a><span class="fu">scp3d</span>(<span class="at">anglea=</span><span class="dv">310</span>)</span>
<span id="cb147-3"><a href="#cb147-3" aria-hidden="true" tabindex="-1"></a><span class="fu">scp3d</span>(<span class="at">anglea=</span><span class="dv">210</span>)</span>
<span id="cb147-4"><a href="#cb147-4" aria-hidden="true" tabindex="-1"></a><span class="fu">scp3d</span>(<span class="at">anglea=</span><span class="dv">150</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The best way of visualizing the data is to make a rotatable interactive plot using the <strong><a href="https://cran.r-project.org/web/packages/rgl/">rgl</a></strong> package.</p>
<div class="cell" data-layout-align="center" data-hash="09-chap_cache/html/CodeD_57b34edd60546e2254adc2b010ea4c86">
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a><span class="co"># interactive plot</span></span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">"rgl"</span>)</span>
<span id="cb148-3"><a href="#cb148-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot3d</span>(dmap1<span class="sc">$</span>X[,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">col=</span>colc, <span class="at">size=</span><span class="dv">3</span>)</span>
<span id="cb148-4"><a href="#cb148-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot3d</span>(dmap1<span class="sc">$</span>X[,<span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>], <span class="at">col=</span>colc, <span class="at">size=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>


<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-Belkin2003" class="csl-entry" role="listitem">
Belkin, Mikhail, and Partha Niyogi. 2003. <span>“Laplacian Eigenmaps for Dimensionality Reduction and Data Representation.”</span> <em>Neural Computation</em> 15 (6): 1373–96.
</div>
<div id="ref-Bengio2004" class="csl-entry" role="listitem">
Bengio, Yoshua, Jean-François Paiement, Pascal Vincent, Olivier Delalleau, Nicolas Le Roux, and Marie Ouimet. 2004. <span>“Out-of-Sample Extensions for <span>LLE</span>, Isomap, <span>MDS</span>, Eigenmaps, and Spectral Clustering.”</span> <em>Advances in Neural Information Processing Systems</em> 16: 177–84.
</div>
<div id="ref-terBraak:1985" class="csl-entry" role="listitem">
Braak, Cajo ter. 1985. <span>“Correspondence Analysis of Incidence and Abundance Data: Properties in Terms of a Unimodal Respose.”</span> <em>Biometrics</em> 41 (January).
</div>
<div id="ref-Brodie:2006" class="csl-entry" role="listitem">
Brodie, Eoin L, Todd Z DeSantis, Dominique C Joyner, Seung M Baek, Joern T Larsen, Gary L Andersen, Terry C Hazen, et al. 2006. <span>“Application of a High-Density Oligonucleotide Microarray Approach to Study Bacterial Population Dynamics During Uranium Reduction and Reoxidation.”</span> <em>Applied and Environmental Microbiology</em> 72 (9): 6288–98.
</div>
<div id="ref-Callahan2016Bioc" class="csl-entry" role="listitem">
Callahan, Ben J, Kris Sankaran, Julia A Fukuyama, Paul J McMurdie, and Susan P Holmes. 2016. <span>“Bioconductor Workflow for Microbiome Data Analysis: From Raw Reads to Community Analyses.”</span> <em>F1000Research</em> 5.
</div>
<div id="ref-ade4" class="csl-entry" role="listitem">
Chessel, Daniel, Anne Dufour, and Jean Thioulouse. 2004. <span>“The <span class="nocase">ade4</span> Package - i: One-Table Methods.”</span> <em>R News</em> 4 (1): 5–10. <a href="http://CRAN.R-project.org/doc/Rnews/">http://CRAN.R-project.org/doc/Rnews/</a>.
</div>
<div id="ref-Goel:2007aa" class="csl-entry" role="listitem">
Diaconis, Persi, Sharad Goel, and Susan Holmes. 2008. <span>“Horseshoes in Multidimensional Scaling and Kernel Methods.”</span> <em>Annals of Applied Statistics</em> 2: 777. <a href="https://doi.org/DOI:10.1214/08-AOAS165">https://doi.org/DOI:10.1214/08-AOAS165</a>.
</div>
<div id="ref-Ekman:1954" class="csl-entry" role="listitem">
Ekman, Gosta. 1954. <span>“Dimensions of Color Vision.”</span> <em>The Journal of Psychology</em> 38 (2): 467–74.
</div>
<div id="ref-Goslee:2007" class="csl-entry" role="listitem">
Goslee, Sarah C, Dean L Urban, et al. 2007. <span>“The Ecodist Package for Dissimilarity-Based Analysis of Ecological Data.”</span> <em>Journal of Statistical Software</em> 22 (7): 1–19.
</div>
<div id="ref-greenacre2007" class="csl-entry" role="listitem">
Greenacre, Michael J. 2007. <em>Correspondence Analysis in Practice</em>. Chapman &amp; Hall.
</div>
<div id="ref-Guillot:2013" class="csl-entry" role="listitem">
Guillot, Gilles, and François Rousset. 2013. <span>“Dismantling the Mantel Tests.”</span> <em>Methods in Ecology and Evolution</em> 4 (4): 336–44.
</div>
<div id="ref-Hastie1989" class="csl-entry" role="listitem">
Hastie, Trevor, and Werner Stuetzle. 1989. <span>“Principal Curves.”</span> <em>Journal of the American Statistical Association</em> 84 (406): 502–16.
</div>
<div id="ref-HastieTibshiraniFriedman" class="csl-entry" role="listitem">
Hastie, Trevor, Robert Tibshirani, and Jerome Friedman. 2008. <em>The Elements of Statistical Learning</em>. 2^{\text{nd}} ed. Springer.
</div>
<div id="ref-frenchway" class="csl-entry" role="listitem">
Holmes, Susan. 2006. <span>“Multivariate Analysis: <span class="nocase">The French way</span>.”</span> In <em>Probability and Statistics: Essays in Honor of David a. Freedman</em>, edited by D. Nolan and T. P. Speed. Vol. 56. IMS Lecture Notes–Monograph Series. Beachwood, OH: IMS. <a href="http://www.imstat.org/publications/lecnotes.htm">http://www.imstat.org/publications/lecnotes.htm</a>.
</div>
<div id="ref-holmes2011psb" class="csl-entry" role="listitem">
Holmes, Susan, Alexander V Alekseyenko, Alden Timme, Tyrrell Nelson, Pankaj Jay Pasricha, and Alfred Spormann. 2011. <span>“Visualization and Statistical Comparisons of Microbial Communities Using r Packages on Phylochip Data.”</span> In <em>Pacific Symposium on Biocomputing</em>, 142–53. World Scientific.
</div>
<div id="ref-Izenman2008" class="csl-entry" role="listitem">
Izenman, Alan Julian. 2008. <span>“Nonlinear Dimensionality Reduction and Manifold Learning.”</span> In <em>Modern Multivariate Statistical Techniques: Regression, Classification, and Manifold Learning</em>, 597–632. New York, NY: Springer New York.
</div>
<div id="ref-Josse:2016" class="csl-entry" role="listitem">
Josse, Julie, and Susan Holmes. 2016. <span>“Measuring Multivariate Association and Beyond.”</span> <em>Statistics Surveys</em> 10: 132–67.
</div>
<div id="ref-Kashyap:2013" class="csl-entry" role="listitem">
Kashyap, Purna C, Angela Marcobal, Luke K Ursell, Samuel A Smits, Erica D Sonnenburg, Elizabeth K Costello, Steven K Higginbottom, et al. 2013. <span>“Genetically Dictated Change in Host Mucus Carbohydrate Landscape Exerts a Diet-Dependent Effect on the Gut Microbiota.”</span> <em>PNAS</em> 110 (42): 17059–64.
</div>
<div id="ref-Kendall:1969" class="csl-entry" role="listitem">
Kendall, David. 1969. <span>“Incidence Matrices, Interval Graphs and Seriation in Archeology.”</span> <em>Pacific Journal of Mathematics</em> 28 (3): 565–70.
</div>
<div id="ref-Leek:2010:batch" class="csl-entry" role="listitem">
Leek, Jeffrey T, Robert B Scharpf, Héctor Corrada Bravo, David Simcha, Benjamin Langmead, W Evan Johnson, Donald Geman, Keith Baggerly, and Rafael A Irizarry. 2010. <span>“Tackling the Widespread and Critical Impact of Batch Effects in High-Throughput Data.”</span> <em>Nature Reviews Genetics</em> 11 (10): 733–39.
</div>
<div id="ref-Moignard:2015" class="csl-entry" role="listitem">
Moignard, Victoria, Steven Woodhouse, Laleh Haghverdi, Andrew J Lilly, Yosuke Tanaka, Adam C Wilkinson, Florian Buettner, et al. 2015. <span>“Decoding the Regulatory Network of Early Blood Development from Single-Cell Gene Expression Measurements.”</span> <em>Nature Biotechnology</em>.
</div>
<div id="ref-Nelson:2011" class="csl-entry" role="listitem">
Nelson, Tyrell A, Susan Holmes, Alexander Alekseyenko, Masha Shenoy, Todd DeSantis, Cindy Wu, Gary Andersen, et al. 2010. <span>“<span>PhyloChip</span> Microarray Analysis Reveals Altered Gastrointestinal Microbial Communities in a Rat Model of Colonic Hypersensitivity.”</span> <em>Neurogastroenterology &amp; Motility</em>.
</div>
<div id="ref-Pages:2016" class="csl-entry" role="listitem">
Pagès, Jérôme. 2016. <em>Multiple Factor Analysis by Example Using <span>R</span></em>. CRC Press.
</div>
<div id="ref-Perraudeau:2017" class="csl-entry" role="listitem">
Perraudeau, Fanny, Davide Risso, Kelly Street, Elizabeth Purdom, and Sandrine Dudoit. 2017. <span>“Bioconductor Workflow for Single-Cell <span>RNA</span> Sequencing: Normalization, Dimensionality Reduction, Clustering, and Lineage Inference.”</span> <em>F1000Research</em> 6.
</div>
<div id="ref-Prentice:1977" class="csl-entry" role="listitem">
Prentice, IC. 1977. <span>“Non-Metric Ordination Methods in Ecology.”</span> <em>The Journal of Ecology</em>, 85–94.
</div>
<div id="ref-HIVdb" class="csl-entry" role="listitem">
Rhee, Soo-Yon, Matthew J Gonzales, Rami Kantor, Bradley J Betts, Jaideep Ravela, and Robert W Shafer. 2003. <span>“Human Immunodeficiency Virus Reverse Transcriptase and Protease Sequence Database.”</span> <em>Nucleic Acids Research</em> 31 (1): 298–303.
</div>
<div id="ref-Roweis2000" class="csl-entry" role="listitem">
Roweis, Sam T, and Lawrence K Saul. 2000. <span>“Nonlinear Dimensionality Reduction by Locally Linear Embedding.”</span> <em>Science</em> 290 (5500): 2323–26.
</div>
<div id="ref-Tenenbaum2000" class="csl-entry" role="listitem">
Tenenbaum, Joshua B, Vin De Silva, and John C Langford. 2000. <span>“A Global Geometric Framework for Nonlinear Dimensionality Reduction.”</span> <em>Science</em> 290 (5500): 2319–23.
</div>
<div id="ref-Trosset2008" class="csl-entry" role="listitem">
Trosset, Michael W, and Carey E Priebe. 2008. <span>“The Out-of-Sample Problem for Classical Multidimensional Scaling.”</span> <em>Computational Statistics &amp; Data Analysis</em> 52 (10): 4635–42.
</div>
</div>
</section>
<aside id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes">
<hr>
<ol>







</ol>
</aside>

<p>Page built on 2023-08-03 21:37:40.85015 using R version 4.3.0 (2023-04-21)</p></main> <!-- /main --></p><p class="build-date">
Support for maintaining the online version of this book is provided by <a href="http://www.denbi.de">de.NBI</a>
  <img src="imgs/denbi.png" style="vertical-align: text-top;height: 16px"><br>
For website support please contact msmith@embl.de
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS_eq = new window.AnchorJS();
  anchorJS_eq.options = {
    placement: 'left',
    class: 'eq-link',
    icon: icon
  };
  anchorJS_eq.add('.anchored-eq');
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>
