---
title: ""
author: "H. Emre Etlioglu"
format:
  html:
    self-contained: true
    toc: true
    code-fold: true
    code-summary: "Show the code"
    page-layout: full
editor: visual
---

# RNA-seq

The `AzureDevops` repo for this project is located [here](lin-to-the-git-repo) and contains all of the code as well as input and output data, with the exception of large raw data files, which are stored in `Stor9`.

## Info on this report

### Report version:

version-1.0-YYYY-MM-DD

### What is different from the previous version

### About the analyses performed

The heavy lifting is mainly done by `DESeq2` and this report closely resembles its following vignettes:

-   [RNA-seq workflow: gene-level exploratory analysis and differential expression](http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html)
-   [Analyzing RNA-seq data with DESeq2](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)

## Info on the project

### To-do

### Questions to the researchers

### Hypothesis/Questions from the researchers

Contrasts to be performed:

## Set-up

```{r set-up}

library(DESeq2)
library(ggplot2)
library(EnhancedVolcano)

# customize ggplot
source("https://raw.githubusercontent.com/etlioglu/bioinformatics/main/templates/customize-ggplot.r")
```

## 2. Input to DESeq2

### 2.1 Read count generation

### 2.2 Count data

```{r count-data}

count_df <- read.delim("preprocessing/nextflow-output/star_salmon/salmon.merged.gene_counts_length_scaled.tsv")
rownames(count_df) <- count_df[, 1]

# get rid of the first two columns: ENSEMBL identifiers and gene names
count_df[, 1:2] <- NULL

# integers are required for the downstream applications
count_df <- round(count_df)
```

### 2.3 Sample meta-data

### 2.4 Gene meta-data

## 3.4 DESeq2 object creation

```{r deseq2-object}

plotDispEsts(dds)
```

## 4. Exploratory data analysis

### 4.1 Pre-filtering

Pre-filtering is not necessary (ref here) but at one point I have seen a Biostar post (or sth like that) that Mike Love was arguing that lowly expressed genes could be effecting the model fit (ref to be added). The plot with the dispersion estimates above should be taken into account.

### 4.2 variance stabilizing transformation

```{r vst}

vsd <- vst(dds, blind = FALSE)
```

### 4.3 Sample distances

### 4.4 PCA Plots

#### 4.4.1 Non-interactive

#### 4.4.2 Interactive

## 5. Differential expression analysis

```{r results}

res <- results(dds, contrast = c("condition", "treated", "untreated"))

# to be used for volcano plots
res_lfc <- lfcShrink(dds, coef="condition_treated_vs_untreated", type="apeglm")
```

## 6 Session Info

```{r session-info}

sessionInfo()
```
